<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>The Principle of R Package GSClassifier</title>
  <meta name="description" content="The Principle of GSClassifier" />
  <meta name="generator" content="bookdown #bookdown:version# and GitBook 2.6.7" />

  <meta property="og:title" content="The Principle of R Package GSClassifier" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The Principle of GSClassifier" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="The Principle of R Package GSClassifier" />
  
  <meta name="twitter:description" content="The Principle of GSClassifier" />
  

<meta name="author" content="Weibin Huang" />


<meta name="date" content="2022-09-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
<script src="libs/header-attrs-2.14/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



<!--bookdown:title:start-->
<div id="header">
<h1 class="title">The Principle of R Package GSClassifier</h1>
<p class="author"><em>Weibin Huang</em></p>
<p class="date"><em>2022-09-20</em></p>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#basic-information" id="toc-basic-information">Basic information</a></li>
<li><a href="#license" id="toc-license">License</a></li>
<li><a href="#installation" id="toc-installation">Installation</a></li>
<li><a href="#mirror" id="toc-mirror"><span>Mirror</span></a></li>
<li><a href="#change-log" id="toc-change-log">Change log</a>
<ul>
<li><a href="#version-0.1.9" id="toc-version-0.1.9">Version 0.1.9</a></li>
<li><a href="#version-0.1.8" id="toc-version-0.1.8">Version 0.1.8</a></li>
</ul></li>
<li><a href="#todo" id="toc-todo">TODO</a></li>
<li><a href="#other-projects" id="toc-other-projects">Other Projects</a></li>
<li><a href="#the-principle-of-gsclassifier" id="toc-the-principle-of-gsclassifier"><span class="toc-section-number">1</span> The Principle of GSClassifier</a>
<ul>
<li><a href="#packages" id="toc-packages"><span class="toc-section-number">1.1</span> Packages</a></li>
<li><a href="#flowchart" id="toc-flowchart"><span class="toc-section-number">1.2</span> Flowchart</a>
<ul>
<li><a href="#data-processing" id="toc-data-processing"><span class="toc-section-number">1.2.1</span> Data Processing</a></li>
<li><a href="#model-establishment-and-validation" id="toc-model-establishment-and-validation"><span class="toc-section-number">1.2.2</span> Model Establishment and Validation</a></li>
<li><a href="#model-application" id="toc-model-application"><span class="toc-section-number">1.2.3</span> Model Application</a></li>
</ul></li>
<li><a href="#topicTSP" id="toc-topicTSP"><span class="toc-section-number">1.3</span> Top scoring pairs</a>
<ul>
<li><a href="#tsp-intro" id="toc-tsp-intro"><span class="toc-section-number">1.3.1</span> Introduction</a></li>
<li><a href="#simulated-dataset" id="toc-simulated-dataset"><span class="toc-section-number">1.3.2</span> Simulated Dataset</a></li>
<li><a href="#binned-expression" id="toc-binned-expression"><span class="toc-section-number">1.3.3</span> Binned expression</a></li>
<li><a href="#pair-difference" id="toc-pair-difference"><span class="toc-section-number">1.3.4</span> Pair difference</a></li>
<li><a href="#set-difference" id="toc-set-difference"><span class="toc-section-number">1.3.5</span> Set difference</a></li>
</ul></li>
</ul></li>
<li><a href="#discussion" id="toc-discussion"><span class="toc-section-number">2</span> Discussion</a>
<ul>
<li><a href="#packages-1" id="toc-packages-1"><span class="toc-section-number">2.1</span> Packages</a></li>
<li><a href="#missing-value-imputation-mvi" id="toc-missing-value-imputation-mvi"><span class="toc-section-number">2.2</span> Missing value imputation (MVI)</a></li>
<li><a href="#batch-effect" id="toc-batch-effect"><span class="toc-section-number">2.3</span> Batch effect</a></li>
<li><a href="#hyperparameters" id="toc-hyperparameters"><span class="toc-section-number">2.4</span> Hyperparameters</a></li>
</ul></li>
<li><a href="#quick-start" id="toc-quick-start"><span class="toc-section-number">3</span> Quick start</a>
<ul>
<li><a href="#about" id="toc-about"><span class="toc-section-number">3.1</span> About</a></li>
<li><a href="#data" id="toc-data"><span class="toc-section-number">3.2</span> Data</a></li>
<li><a href="#pad" id="toc-pad"><span class="toc-section-number">3.3</span> PAD</a>
<ul>
<li><a href="#preparation-of-the-test-data" id="toc-preparation-of-the-test-data"><span class="toc-section-number">3.3.1</span> Preparation of the test data</a></li>
<li><a href="#unsupervised-clustering" id="toc-unsupervised-clustering"><span class="toc-section-number">3.3.2</span> Unsupervised clustering</a></li>
<li><a href="#of-note" id="toc-of-note"><span class="toc-section-number">3.3.3</span> Of note</a></li>
</ul></li>
<li><a href="#padi" id="toc-padi"><span class="toc-section-number">3.4</span> PADi</a>
<ul>
<li><a href="#preparation-of-the-test-data-1" id="toc-preparation-of-the-test-data-1"><span class="toc-section-number">3.4.1</span> Preparation of the test data</a></li>
<li><a href="#use-a-specific-function-called-padi" id="toc-use-a-specific-function-called-padi"><span class="toc-section-number">3.4.2</span> Use a specific function called <code>PADi</code></a></li>
<li><a href="#use-the-callensemble-function" id="toc-use-the-callensemble-function"><span class="toc-section-number">3.4.3</span> Use the <code>callEnsemble</code> function</a></li>
<li><a href="#parallel-strategy-for-padi" id="toc-parallel-strategy-for-padi"><span class="toc-section-number">3.4.4</span> Parallel strategy for PADi</a></li>
<li><a href="#single-sample-subtype-calling" id="toc-single-sample-subtype-calling"><span class="toc-section-number">3.4.5</span> Single sample subtype calling</a></li>
<li><a href="#of-note-1" id="toc-of-note-1"><span class="toc-section-number">3.4.6</span> Of note</a></li>
</ul></li>
<li><a href="#use-external-models-from-luckymodel-package" id="toc-use-external-models-from-luckymodel-package"><span class="toc-section-number">3.5</span> Use external models from luckyModel package</a></li>
<li><a href="#pancancer-immune-subtypes" id="toc-pancancer-immune-subtypes"><span class="toc-section-number">3.6</span> PanCancer immune subtypes</a></li>
</ul></li>
<li><a href="#model-establishment-via-gsclassifier" id="toc-model-establishment-via-gsclassifier"><span class="toc-section-number">4</span> Model establishment via GSClassifier</a>
<ul>
<li><a href="#about-1" id="toc-about-1"><span class="toc-section-number">4.1</span> About</a></li>
<li><a href="#data-preparation" id="toc-data-preparation"><span class="toc-section-number">4.2</span> Data preparation</a></li>
<li><a href="#fitting-models" id="toc-fitting-models"><span class="toc-section-number">4.3</span> Fitting models</a>
<ul>
<li><a href="#gsclassifier-model-training" id="toc-gsclassifier-model-training"><span class="toc-section-number">4.3.1</span> GSClassifier model training</a></li>
<li><a href="#scaller-for-the-best-call" id="toc-scaller-for-the-best-call"><span class="toc-section-number">4.3.2</span> Scaller for the best call</a></li>
<li><a href="#assemble-your-model" id="toc-assemble-your-model"><span class="toc-section-number">4.3.3</span> Assemble your model</a></li>
<li><a href="#of-note-2" id="toc-of-note-2"><span class="toc-section-number">4.3.4</span> Of note</a></li>
</ul></li>
<li><a href="#calling-subtypes" id="toc-calling-subtypes"><span class="toc-section-number">4.4</span> Calling subtypes</a></li>
</ul></li>
<li><a href="#suggestions-for-gsclassifier-model-developers" id="toc-suggestions-for-gsclassifier-model-developers"><span class="toc-section-number">5</span> Suggestions for GSClassifier model developers</a>
<ul>
<li><a href="#about-2" id="toc-about-2"><span class="toc-section-number">5.1</span> About</a></li>
<li><a href="#available-models" id="toc-available-models"><span class="toc-section-number">5.2</span> Available models</a></li>
<li><a href="#components-of-a-gsclassifier-model" id="toc-components-of-a-gsclassifier-model"><span class="toc-section-number">5.3</span> Components of a GSClassifier model</a></li>
<li><a href="#submit-models-to-luckymodel-package" id="toc-submit-models-to-luckymodel-package"><span class="toc-section-number">5.4</span> Submit models to luckyModel package</a></li>
<li><a href="#repeatablility-of-models" id="toc-repeatablility-of-models"><span class="toc-section-number">5.5</span> Repeatablility of models</a></li>
<li><a href="#gene-annotation" id="toc-gene-annotation"><span class="toc-section-number">5.6</span> Gene Annotation</a></li>
</ul></li>
<li><a href="#references" id="toc-references">References</a></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Principle of R Package GSClassifier</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<!-- 

# Welcome {-}

<!--

[GSClassifier](https://github.com/huangwb8/GSClassifier) is an R package for modeling and identification of Gene Expression Profiles (GEPs) subtypes. The detail of **GSClassifier** package usage had been demonstrated in [Github WiKi](https://github.com/huangwb8/GSClassifier/wiki). Here, we propose to introduce the principle of GSClassifier, including flowchart, **top scoring pairs (TSP)** algorithm, and batch effect control. 

emoji: https://github.com/rstudio/blogdown/issues/171

-->
<div id="basic-information" class="section level2 unnumbered">
<h2 class="unnumbered">Basic information</h2>
<p><strong>The Priciple of GSClassifier</strong> is a book for users of R package <strong>GSClassifier</strong> who want to know the most details. If you’re looking for the PDF edition, you can find it at <a rel="pdf" href="https://github.com/huangwb8/GSClassifier.principle/blob/master/docs/GSClassifier.principle.pdf">here</a>.</p>
<ul>
<li><a href="https://github.com/huangwb8/GSClassifier"><strong>GSClassifier</strong></a> is an R-based comprehensive classification tool for subtypes modeling and personalized calling based on pure transcriptomics. It could be used for precision medicine, such as cancer diagnosis.</li>
<li>The inspiration of <strong>GSClassifier</strong> come from <a href="https://github.com/CRI-iAtlas/ImmuneSubtypeClassifier">ImmuneSubtypeClassifier</a>, an R package for classification of PanCancer immune subtypes based on the work of Gibbs et al <span class="citation">[<a href="#ref-RN160" role="doc-biblioref">1</a>,<a href="#ref-RN315" role="doc-biblioref">2</a>]</span>.</li>
<li>Lots of surprising features in <strong>GSClassifier</strong> as follows:
<ul>
<li>Optimized for just <code>one sample</code></li>
<li>Available for modeling and calling of brand-new <code>GEPs-based subtypes</code> in any diseases (cancers)</li>
<li>No limitation of the amount of <code>gene signatures</code>(≥1) or <code>subtypes</code>(≥2)</li>
<li><code>Normalization insensitive</code> due to the use of the individual <code>gene rank matrix</code></li>
<li>More ensemble and repeatable modeling process</li>
<li>More optimizations in the parallel computing</li>
<li>New useful functions as supplements</li>
</ul></li>
<li><strong>ATTENTION!</strong> In the future, there might be third-party contributors in <code>GSClassifier</code> platform, with some useful models for specific usages. If you use models provided by these people, <strong>you had better know more details as possible</strong>, including <strong>designs, data sources, destinations, training scripts and limitations</strong> of models, expecially those from studies under peer-review.</li>
<li><strong>MORE PROJECTS</strong>:
<ul>
<li><a href="https://huangwb8.github.io/GSClassifier.principle/"><strong>The Principle of GSClassifier</strong></a>: A eBook with more details about GSClassifier package</li>
<li><a href="https://github.com/huangwb8/luckyModel"><strong>luckyModel</strong></a>: Model ensemble for third-party lucky series, such GSClassifier</li>
</ul></li>
</ul>
</div>
<div id="license" class="section level2 unnumbered">
<h2 class="unnumbered">License</h2>
<p><strong>GSClassifier</strong> is released under the Apache-2.0 license. See <a href="https://github.com/huangwb8/GSClassifier/blob/master/license.txt">LICENSE</a> for details.</p>
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a></p>
<p>The technical documentation, as a whole, is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.</p>
<p>The code contained in this book is simultaneously available under the <a href="https://opensource.org/licenses/MIT">MIT license</a>; this means that you are free to use it in your own packages, as long as you cite the source.</p>
</div>
<div id="installation" class="section level2 unnumbered">
<h2 class="unnumbered">Installation</h2>
<p><code>RStudio</code> is one of the best Integrated Development Environments (IDE) in R programming. If you’re struggling in R-GUI, it is recommanded to turn to <a href="https://www.rstudio.com/">RStudio</a>.</p>
<p>For installation of <strong>GSClassifier</strong>, please run these commands in an R environment:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install &quot;devtools&quot; package</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;luckyBase&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/luckyBase&quot;</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;GSClassifier&quot; package</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;GSClassifier&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/GSClassifier&quot;</span>)</span></code></pre></div>
<p>In the future, a stable <code>GSClassifier</code> version might be sent to <a href="https://cran.r-project.org/"><code>CRAN</code></a>. Still beta.</p>
</div>
<div id="mirror" class="section level2 unnumbered">
<h2 class="unnumbered"><a href="https://gitee.com/huangwb8/GSClassifier">Mirror</a></h2>
<p>For some special countries or regions, users could also try:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;https://gitee.com/huangwb8/luckyBase/repository/archive/Primary?format=tar.gz&quot;</span>, <span class="at">repos=</span><span class="cn">NULL</span>, <span class="at">method=</span><span class="st">&quot;libcurl&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;GSClassifier&quot; package</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;https://gitee.com/huangwb8/GSClassifier/repository/archive/Primary?format=tar.gz&quot;</span>, <span class="at">repos=</span><span class="cn">NULL</span>, <span class="at">method=</span><span class="st">&quot;libcurl&quot;</span>)</span></code></pre></div>
</div>
<div id="change-log" class="section level2 unnumbered">
<h2 class="unnumbered">Change log</h2>
<div id="version-0.1.9" class="section level3 unnumbered">
<h3 class="unnumbered">Version 0.1.9</h3>
<ul>
<li>Optimize function verbose</li>
<li>Optimize for a routine scenario: one gene set and two subtypes</li>
<li>Optimize the strategy of automatic parameters selection for modeling training with R package <code>caret</code></li>
<li>Interact with external models from the <a href="https://github.com/huangwb8/luckyModel">luckyModel</a> package</li>
</ul>
</div>
<div id="version-0.1.8" class="section level3 unnumbered">
<h3 class="unnumbered">Version 0.1.8</h3>
<ul>
<li>Primary public version of <code>GSClassifier</code></li>
<li>Apache License, Version 2.0</li>
<li>Friendly wiki-based tutorial</li>
<li>Platform for developers</li>
</ul>
</div>
</div>
<div id="todo" class="section level2 unnumbered">
<h2 class="unnumbered">TODO</h2>
<ul>
<li>More medical fields included, such as in Pan-cancer uitility</li>
<li>Advanced methods (such as artificial intelligence) for enhanced robustness</li>
<li>Unsupervised learning for de-novo classification based on intrinsic frames of omics instead of human knowledges</li>
<li>Multi-omics exploration and support</li>
<li>More friendly characteristics for developers and contributors</li>
<li>Web application for newbies of R programing</li>
</ul>
</div>
<div id="other-projects" class="section level2 unnumbered">
<h2 class="unnumbered">Other Projects</h2>
<p>You may also be interested in:</p>
<ul>
<li><p><strong>“<a href="https://github.com/huangwb8/GSClassifier">GSClassifier</a>”</strong> A comprehensive classification tool based on pure transcriptomics for precision medicine.</p></li>
<li><p><strong>“<a href="https://github.com/huangwb8/luckyModel">luckyModel</a>”</strong> Model ensemble for third-party lucky series, such GSClassifier.</p></li>
</ul>
<p>–&gt;</p>
<!--chapter:end:index.Rmd-->
</div>
<div id="the-principle-of-gsclassifier" class="section level1" number="1">
<h1><span class="header-section-number">1</span> The Principle of GSClassifier</h1>
<div id="packages" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Packages</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Install &quot;devtools&quot; package</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;luckyBase&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/luckyBase&quot;</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;GSClassifier&quot; package</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;GSClassifier&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/GSClassifier&quot;</span>)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;pacman&quot; package</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;pacman&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;pacman&quot;</span>)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Load needed packages</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>packages_needed <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;readxl&quot;</span>,</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;ComplexHeatmap&quot;</span>,</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;GSClassifier&quot;</span>,</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;rpart&quot;</span>,</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;tidyr&quot;</span>,</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;reshape2&quot;</span>,</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> packages_needed){<span class="fu">p_load</span>(<span class="at">char=</span>i)}</span></code></pre></div>
<p>Here is the environment of R programming:</p>
<pre><code># R version 4.0.3 (2020-10-10)
# Platform: x86_64-w64-mingw32/x64 (64-bit)
# Running under: Windows 10 x64 (build 18363)
# 
# Matrix products: default
# 
# locale:
# [1] LC_COLLATE=Chinese (Simplified)_China.936 
# [2] LC_CTYPE=Chinese (Simplified)_China.936   
# [3] LC_MONETARY=Chinese (Simplified)_China.936
# [4] LC_NUMERIC=C                              
# [5] LC_TIME=Chinese (Simplified)_China.936    
# 
# attached base packages:
# [1] grid      stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
# [1] ggplot2_3.3.6        reshape2_1.4.4       tidyr_1.2.0         
# [4] rpart_4.1.16         GSClassifier_0.1.22  luckyBase_0.1.0     
# [7] ComplexHeatmap_2.4.3 readxl_1.4.0         pacman_0.5.1        
# 
# loaded via a namespace (and not attached):
#   [1] colorspace_2.0-3     ggsignif_0.6.3       rjson_0.2.21        
#   [4] ellipsis_0.3.2       class_7.3-20         rprojroot_2.0.3     
#   [7] circlize_0.4.15      GlobalOptions_0.1.2  fs_1.5.2            
#  [10] clue_0.3-57          rstudioapi_0.13      ggpubr_0.4.0        
#  [13] listenv_0.8.0        remotes_2.4.2        prodlim_2019.11.13  
#  [16] fansi_1.0.3          lubridate_1.8.0      codetools_0.2-18    
#  [19] splines_4.0.3        doParallel_1.0.17    cachem_1.0.6        
#  [22] knitr_1.30           pkgload_1.2.4        jsonlite_1.8.0      
#  [25] pROC_1.18.0          caret_6.0-92         broom_1.0.0         
#  [28] cluster_2.1.3        png_0.1-7            compiler_4.0.3      
#  [31] backports_1.4.1      assertthat_0.2.1     Matrix_1.2-18       
#  [34] fastmap_1.1.0        cli_3.3.0            htmltools_0.5.2     
#  [37] prettyunits_1.1.1    tools_4.0.3          gtable_0.3.0        
#  [40] glue_1.6.2           dplyr_1.0.9          Rcpp_1.0.8.3        
#  [43] carData_3.0-5        cellranger_1.1.0     vctrs_0.4.1         
#  [46] nlme_3.1-149         iterators_1.0.14     timeDate_3043.102   
#  [49] xfun_0.33            gower_1.0.0          stringr_1.4.0       
#  [52] globals_0.15.1       ps_1.4.0             testthat_3.1.0      
#  [55] lifecycle_1.0.1      devtools_2.4.3       rstatix_0.7.0       
#  [58] future_1.26.1        MASS_7.3-53          scales_1.2.0        
#  [61] ipred_0.9-12         parallel_4.0.3       RColorBrewer_1.1-3  
#  [64] yaml_2.3.5           memoise_2.0.1        stringi_1.7.6       
#  [67] desc_1.4.1           randomForest_4.6-14  foreach_1.5.2       
#  [70] hardhat_1.1.0        pkgbuild_1.3.1       lava_1.6.10         
#  [73] shape_1.4.6          tuneR_1.4.0          rlang_1.0.2         
#  [76] pkgconfig_2.0.3      evaluate_0.15        lattice_0.20-41     
#  [79] purrr_0.3.4          recipes_0.2.0        processx_3.7.0      
#  [82] tidyselect_1.1.2     parallelly_1.32.0    plyr_1.8.7          
#  [85] magrittr_2.0.3       bookdown_0.21        R6_2.5.1            
#  [88] generics_0.1.2       DBI_1.1.3            pillar_1.7.0        
#  [91] withr_2.5.0          survival_3.3-1       abind_1.4-5         
#  [94] nnet_7.3-17          tibble_3.1.7         future.apply_1.9.0  
#  [97] crayon_1.5.1         car_3.1-0            xgboost_1.6.0.1     
# [100] utf8_1.2.2           rmarkdown_2.14       GetoptLong_1.0.5    
# [103] usethis_2.1.3        data.table_1.14.2    callr_3.7.0         
# [106] ModelMetrics_1.2.2.2 digest_0.6.29        stats4_4.0.3        
# [109] signal_0.7-7         munsell_0.5.0        sessioninfo_1.2.2</code></pre>
</div>
<div id="flowchart" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Flowchart</h2>
<p>The flowchart of <strong>GSClassifier</strong> is showed in Figure @ref(fig:flowchart).</p>
<div id="data-processing" class="section level3" number="1.2.1">
<h3><span class="header-section-number">1.2.1</span> Data Processing</h3>
<p>For each dataset, RNA expression matrix would be normalized internally (<strong>Raw Matrix</strong>) so that the expression data of the samples in the dataset were comparable and suitable for subtype identification. As demonstrated in Figure @ref(fig:flowchart), the <strong>Subtype vector</strong> is identified based on independent cohorts instead of a merged matrix with batch effect control technologies. More details about batch effect control are discussed in @ref(batch-effect).</p>
<p>There is no standard method to figure out subtype vectors. It depends on the Gene Expression Profiles (GEPs) used, the biological problems or ideas of researchers. For <strong>Pan-immune Activation and Dysfunction (PAD)</strong> subtypes, the GEPs, <strong>Pan-Immune Activation Module (PIAM)</strong> and <strong>Pan-Immune Dysfunction Genes (PIDG)</strong>, are biologically associated and suitable for calling four sutbypes (PIAM<sup>high</sup>PIDG<sup>high</sup>, PIAM<sup>high</sup>PIDG<sup>low</sup>, PIAM<sup>low</sup>PIDG<sup>high</sup>, and PIAM<sup>low</sup>PIDG<sup>low</sup>).<!--这里描述有待改进--> Theoretically, we can also use a category strategy like low/medium/high, but more evidences or motivations are lacked for chasing such a complex model.</p>
<p>With subtype vectors and raw matrices, <strong>Top Scoring Pairs (TSP)</strong>, the core data format for model training and application in GSClassifier, would be calculated for the following process. The details of TSP normalization are summarized in @ref(topicTSP).</p>
</div>
<div id="model-establishment-and-validation" class="section level3" number="1.2.2">
<h3><span class="header-section-number">1.2.2</span> Model Establishment and Validation</h3>
<p>The TSP matrix would be devided into training cohort and internal validation cohort. In PAD project, the rate of samples (training vs. test) is <strong>7:3</strong>. Next, each <strong>SubSet</strong> (70% of the training cohort in PAD project) would be further selected randomly to build a <strong>SubModel</strong> via cross-validation Extreme Gradient Boosting algorithm (<strong>xgboost::xgb.cv</strong> function) <span class="citation">[<a href="#ref-RN345" role="doc-biblioref">3</a>]</span>. The number of submodels is suggested over 20 (more details in @ref(topicSubmodel)).</p>
<p>The internal validation cohort and external validation cohort (if any) would be used to test the performace of the trained model. By the way, <strong>the data of both internal and external validation cohort would not be used during model training</strong> so as to avoid over-fitting.</p>
</div>
<div id="model-application" class="section level3" number="1.2.3">
<h3><span class="header-section-number">1.2.3</span> Model Application</h3>
<p>In PAD project, <strong>Model for individual</strong>, the ensemble of submodels, is called “PAD for individual” (<strong>PADi</strong>). Supposed raw RNA expression of a sample was given. As showed in @ref(fig:flowchart) and @ref(fig:tsp), <strong>GSClassifier</strong> would turn raw RNA expression into a TSP vector, which would be as a input to <strong>Model for individual</strong>. Then, <strong>GSClassifier</strong> would output the possibility matrix and the subtype for this sample. No extra data (RNA expression of others, follow-up data, etc) would be needed but RNA expression of the patient for subtype identification, so we suggest <strong>Model for individual</strong> (<strong>PADi</strong>, etc) as personalized model.</p>
</div>
</div>
<div id="topicTSP" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Top scoring pairs</h2>
<div id="tsp-intro" class="section level3" number="1.3.1">
<h3><span class="header-section-number">1.3.1</span> Introduction</h3>
<p>Genes expression of an individual is normalized during the model training and the subtype identification via <strong>Top Scoring Pairs</strong> (<strong>TSP</strong>, also called <strong>Relative Expression Orderings</strong> (<strong>REOs</strong>)) algorithm, which was previously described by Geman et al <span class="citation">[<a href="#ref-RN267" role="doc-biblioref">4</a>]</span>. <strong>TSP</strong> normalization for an individual depends on its transcript data, implying that subtype calling would not be perturbed by data from other individuals or other extra information like follow-up data. <strong>TSP</strong> had been used in cancer research and effectively predicts cancer progression and ICIs response <span class="citation">[<a href="#ref-RN265" role="doc-biblioref">5</a>–<a href="#ref-RN261" role="doc-biblioref">7</a>]</span>.</p>
<p>As show in Figure @ref(fig:tsp), The TSP data in GSClassifier consists of three parts: <strong>binned expression</strong>, <strong>pair difference</strong>, and <strong>set difference</strong>. In this section, we would conduct some experiments to demonstrate the potential of TSP normalization for development of cross-dataset/platform GEP-based models.</p>
</div>
<div id="simulated-dataset" class="section level3" number="1.3.2">
<h3><span class="header-section-number">1.3.2</span> Simulated Dataset</h3>
<p>We simulated a dataset to demonstrate TSP normalization in GSClassifier:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Geneset</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>geneSet <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Set1 =</span> <span class="fu">paste</span>(<span class="st">&#39;Gene&#39;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="at">sep =</span> <span class="st">&#39;&#39;</span>),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Set2 =</span> <span class="fu">paste</span>(<span class="st">&#39;Gene&#39;</span>,<span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>,<span class="at">sep =</span> <span class="st">&#39;&#39;</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">&#39;./data/simulated-data.xlsx&#39;</span>, <span class="at">sheet =</span> <span class="st">&#39;RNA&#39;</span>)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>expr0 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(x[,<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr0) <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">as.matrix</span>(x[,<span class="dv">1</span>])); <span class="fu">rm</span>(x)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Missing value imputation (MVI)</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> <span class="fu">na_fill</span>(expr0, <span class="at">method =</span> <span class="st">&quot;quantile&quot;</span>, <span class="at">seed =</span> <span class="dv">447</span>)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co">#  Missing value imputation with quantile algorithm...</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Subtype information</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co"># It depends on the application scenarios of GEPs</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>subtype_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>)</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Binned data for subtype 1</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>Ybin <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(subtype_vector <span class="sc">==</span> <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Parameters</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>breakVec <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">1.0</span>)</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">c</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>, <span class="st">&#39;Gene sets:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>))</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(geneSet)</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr0); <span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression after MVI:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr)</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a><span class="co">#  Gene sets: </span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="co"># $Set1</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot;</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="co"># $Set2</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression: </span></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60    0.21    0.30    0.40</span></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58    0.22    0.31    0.35</span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61      NA    0.29    0.30</span></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40    0.51    0.52    0.60</span></span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35    0.52    0.54    0.58</span></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30    0.53      NA    0.61</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09    0.11    0.12    0.14</span></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression after MVI: </span></span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60  0.2100 0.30000    0.40</span></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58  0.2200 0.31000    0.35</span></span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61  0.2486 0.29000    0.30</span></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40  0.5100 0.52000    0.60</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35  0.5200 0.54000    0.58</span></span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30  0.5300 0.51774    0.61</span></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09  0.1100 0.12000    0.14</span></span></code></pre></div>
<p>Look at the matrix via heatmap:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(expr0))), <span class="at">name =</span> <span class="st">&quot;Z-score&quot;</span>, <span class="at">column_title =</span> <span class="st">&quot;Before MVI&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(expr))), <span class="at">name =</span> <span class="st">&quot;Z-score&quot;</span>, <span class="at">column_title =</span> <span class="st">&quot;After MVI&quot;</span>)</span></code></pre></div>
<p>This is an intersting dataset with features as following:</p>
<ul>
<li><p><strong>Distinguished gene sets</strong>: The expression profile between <strong>Gene 1-3</strong> and <strong>Gene 4-6</strong> is obviously different arross samples. Thus, these gene sets might represent different biology meaning.</p></li>
<li><p><strong>Stable gene</strong>: The expression level and rank of <strong>Gene 7</strong> seemed to be similar across samples. Thus, <strong>Gene 7</strong> might not be a robust marker for subtype modeling. Thus, it could help us to understand how filtering of <strong>GSClassifier</strong> works.</p></li>
<li><p><strong>Expression heterogeneity &amp; rank homogeneity</strong>: Take <strong>Sample1</strong> and <strong>Sample3</strong> as examples. The expression of <strong>Gene 1-6</strong> in <strong>Sample3</strong> seemed to be higher than those of <strong>Sample1</strong>. However, the expression of <strong>Gene 1-3</strong> is higher than <strong>Gene 4-6</strong> in both <strong>Sample1</strong> and <strong>Sample3</strong>, indicating similar bioprocess in these samples exists so that they should be classified as the same subtype.</p></li>
</ul>
</div>
<div id="binned-expression" class="section level3" number="1.3.3">
<h3><span class="header-section-number">1.3.3</span> Binned expression</h3>
<p>First, we binned genes with diffrent quantile intervals so that the distribution of rank information could be more consistent across samples.</p>
<p>Take <strong>Sample4</strong> as an example:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Data of Sample4</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> expr[,<span class="dv">4</span>]</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create quantiles  </span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>brks <span class="ot">&lt;-</span> <span class="fu">quantile</span>(<span class="fu">as.numeric</span>(x), </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">probs=</span>breakVec, </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">na.rm =</span> T)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Get interval orders</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>xbin <span class="ot">&lt;-</span> <span class="fu">.bincode</span>(<span class="at">x =</span> x, </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">breaks =</span> brks, </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">include.lowest =</span> T)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>xbin <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(xbin)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(xbin) <span class="ot">&lt;-</span> <span class="fu">names</span>(x)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Quantiles:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>); <span class="fu">print</span>(brks)</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Raw expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>);<span class="fu">print</span>(x)</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Binned expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>); <span class="fu">print</span>(xbin)</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantiles: </span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a><span class="co">#     0%    25%    50%    75%   100% </span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.1100 0.2150 0.2486 0.5150 0.5300 </span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Raw expression: </span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a><span class="co">#  Gene1  Gene2  Gene3  Gene4  Gene5  Gene6  Gene7 </span></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.2100 0.2200 0.2486 0.5100 0.5200 0.5300 0.1100 </span></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Binned expression: </span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1 Gene2 Gene3 Gene4 Gene5 Gene6 Gene7 </span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a><span class="co">#     1     2     2     3     4     4     1</span></span></code></pre></div>
<p>For example, <strong>0.110</strong> is the minimun of the raw expression vector, so its binned expression is <strong>1</strong>. Similarly, the binned expression of maximum <strong>0.530</strong> is <strong>4</strong>.</p>
<p>Generally, we calculate binned expression via function <strong>breakBin</strong> of <strong>GSClassifier</strong>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>expr_binned <span class="ot">&lt;-</span> <span class="fu">apply</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  expr, <span class="dv">2</span>, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  GSClassifier<span class="sc">:::</span>breakBin,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  breakVec)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr_binned) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(expr)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr_binned)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1       3       3       4       1       2       2</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2       4       4       3       2       2       2</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3       4       4       4       2       1       1</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4       1       2       2       3       4       4</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5       2       2       2       4       4       3</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6       2       1       1       4       3       4</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7       1       1       1       1       1       1</span></span></code></pre></div>
<p>In this simulated dataset, <strong>Gene7</strong> is a gene whose expression is always the lowest across all samples. In other words, the rank of <strong>Gene7</strong> is stable or invariable across samples so that it’s not robust for identification of differentail subtypes.</p>
<p>Except binned expression, we also calculated pair difference later. Due to the number of gene pair is <span class="math inline">\(C_{2 \atop n}\)</span>, the removement of genes like <strong>Gene7</strong> before modeling could really reduce the complexibility and save computing resources. In all, genes with low rank difference should be dropped out in some extent in <strong>GSClassifier</strong>.</p>
<p>First, We use <strong>base::rank</strong> to return the sample ranks of the values in a vector:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>expr_binned_rank <span class="ot">&lt;-</span> <span class="fu">apply</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  expr_binned, <span class="dv">2</span>, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x)<span class="fu">rank</span>(x, <span class="at">na.last =</span> <span class="cn">TRUE</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr_binned_rank)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1     5.0     5.0     6.5     1.5     3.5     3.5</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2     6.5     6.5     5.0     3.5     3.5     3.5</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3     6.5     6.5     6.5     3.5     1.5     1.5</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4     1.5     3.5     3.5     5.0     6.5     6.5</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5     3.5     3.5     3.5     6.5     6.5     5.0</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6     3.5     1.5     1.5     6.5     5.0     6.5</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7     1.5     1.5     1.5     1.5     1.5     1.5</span></span></code></pre></div>
<p>Then, get weighted average rank difference of each gene based on specified subtype distribution (<strong>Ybin</strong>):</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>testRes <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(expr_binned_rank), </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(gi){</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Rank vector of each gene</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    rankg <span class="ot">=</span> expr_binned_rank[gi,];</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Weighted average rank difference of a gene for specified subtype </span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Here is subtype 1 vs. others</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    (<span class="fu">sum</span>(rankg[Ybin <span class="sc">==</span> <span class="dv">0</span>], <span class="at">na.rm =</span> T) <span class="sc">/</span> <span class="fu">sum</span>(Ybin <span class="sc">==</span> <span class="dv">0</span>, <span class="at">na.rm =</span> T)) <span class="sc">-</span> </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    (<span class="fu">sum</span>(rankg[Ybin <span class="sc">==</span> <span class="dv">1</span>], <span class="at">na.rm =</span> T) <span class="sc">/</span> <span class="fu">sum</span>(Ybin <span class="sc">==</span> <span class="dv">1</span>, <span class="at">na.rm =</span> T))</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(testRes) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(expr_binned_rank)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(testRes)</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="co">#     Gene1     Gene2     Gene3     Gene4     Gene5     Gene6     Gene7 </span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="co"># -2.666667 -2.500000 -4.333333  3.166667  2.500000  3.833333  0.000000</span></span></code></pre></div>
<p><strong>Gene7</strong> is the one with the lowest absolute value (0) of rank diffrence. By the way, <strong>Gene 1-3</strong> have the same direction (&lt;0), so do <strong>Gene 4-6</strong> (&gt;0), which indicates the nature of clustering based on these two gene sets.</p>
<p>In practice, we use <strong>ptail</strong> to select differential genes based on rank diffrences. <strong>Smaller ptail is, less gene kept</strong>. Here, we just set <strong>ptail=0.4</strong>:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ptail is a numeber ranging (0,0.5].</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>ptail <span class="ot">=</span> <span class="fl">0.4</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Index of target genes with big rank differences</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">which</span>((testRes <span class="sc">&lt;</span> <span class="fu">quantile</span>(testRes, ptail, <span class="at">na.rm =</span> T)) <span class="sc">|</span> </span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>             (testRes <span class="sc">&gt;</span> <span class="fu">quantile</span>(testRes, <span class="fl">1.0</span><span class="sc">-</span>ptail, <span class="at">na.rm =</span> T)))</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Target genes</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>gene_bigRank <span class="ot">&lt;-</span> <span class="fu">names</span>(testRes)[idx]</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Index of target genes: &#39;</span>,<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>);<span class="fu">print</span>(idx); <span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Target genes:&#39;</span>,<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>);<span class="fu">print</span>(gene_bigRank)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Index of target genes:  </span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1 Gene2 Gene3 Gene4 Gene5 Gene6 </span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a><span class="co">#     1     2     3     4     5     6 </span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Target genes: </span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot; &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span></code></pre></div>
<p>Hence, <strong>Gene7</strong> was filtered and excluded in the following analysis. By the way, both <strong>ptail</strong> and <strong>breakVec</strong> are hyperparameters in GSClassifier modeling.</p>
</div>
<div id="pair-difference" class="section level3" number="1.3.4">
<h3><span class="header-section-number">1.3.4</span> Pair difference</h3>
<p>In GSClassifier, we use a ensemble function <strong>featureSelection</strong> to select data for pair difference scoring.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>expr_feat <span class="ot">&lt;-</span> <span class="fu">featureSelection</span>(expr, Ybin,</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">testRes =</span> testRes,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ptail =</span> <span class="fl">0.4</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>expr_sub <span class="ot">&lt;-</span> expr_feat<span class="sc">$</span>Xsub</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>gene_bigRank <span class="ot">&lt;-</span> expr_feat<span class="sc">$</span>Genes</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Raw xpression without NA:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr_sub)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Genes with large rank diff:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene_bigRank)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Raw xpression without NA: </span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60  0.2100 0.30000    0.40</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58  0.2200 0.31000    0.35</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61  0.2486 0.29000    0.30</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40  0.5100 0.52000    0.60</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35  0.5200 0.54000    0.58</span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30  0.5300 0.51774    0.61</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Genes with large rank diff: </span></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot; &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span></code></pre></div>
<p>In GSClassifier, we use function <strong>makeGenePairs</strong> to calculate s</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>gene_bigRank_pairs <span class="ot">&lt;-</span> GSClassifier<span class="sc">:::</span><span class="fu">makeGenePairs</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  gene_bigRank, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  expr[gene_bigRank,])</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene_bigRank_pairs)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co">#             Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene2       0       0       1       0       0       1</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene3       0       0       0       0       1       1</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene4       1       1       1       0       0       0</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene5       1       1       1       0       0       0</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene6       1       1       1       0       0       0</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene3       0       0       0       0       1       1</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene4       1       1       1       0       0       0</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene5       1       1       1       0       0       0</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene6       1       1       1       0       0       0</span></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene4       1       1       1       0       0       0</span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene5       1       1       1       0       0       0</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene6       1       1       1       0       0       0</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene5       0       0       1       0       0       1</span></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene6       0       1       1       0       1       0</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5:Gene6       0       1       1       0       1       0</span></span></code></pre></div>
<p>Take <strong>Gene1:Gene4</strong> of <strong>Sample1</strong> as an example. <span class="math inline">\(Expression_{Gene1} - Expression_{Gene4} = 0.51-0.21 = 0.3 &gt; 0\)</span>, so the pair score is 1. If the difference is less than or equal to 0, the pair score is 0. In addition, the difference of gene pair scoring between <strong>Sample 1-3</strong> and <strong>Sample 4-6</strong> is obivous, revealing the robustness of pair difference for subtype identification.</p>
</div>
<div id="set-difference" class="section level3" number="1.3.5">
<h3><span class="header-section-number">1.3.5</span> Set difference</h3>
<p>In <strong>GSClassifier</strong>, <strong>Set difference</strong> is defined as a weight average of gene-geneset rank difference.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># No. of gene sets</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>nGS <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Name of gene set comparision, which is like s1s2, s1s3 and so on.</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>featureNames <span class="ot">&lt;-</span> <span class="st">&#39;s1s2&#39;</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene set difference across samples</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>resultList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(expr_sub)) { <span class="co"># i=1</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  res0 <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="at">length=</span><span class="fu">length</span>(featureNames))</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j1 <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(nGS<span class="dv">-1</span>)) { <span class="co"># j1=1</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j2 <span class="cf">in</span> (j1<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span>nGS) { <span class="co"># j2=2</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If j1=1 and j2=2, gene sets s1/s2 would be selected</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Genes of different gene sets</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>      set1 <span class="ot">&lt;-</span> geneSet[[j1]] <span class="co"># &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot;</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>      set2 <span class="ot">&lt;-</span> geneSet[[j2]] <span class="co"># &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>      <span class="co"># RNA expression of Genes by different gene sets</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>      vals1 <span class="ot">&lt;-</span> expr_sub[<span class="fu">rownames</span>(expr_sub) <span class="sc">%in%</span> set1,i]</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Gene1 Gene2 Gene3</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>      <span class="co"># 0.51  0.52  0.53</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>      vals2 <span class="ot">&lt;-</span> expr_sub[<span class="fu">rownames</span>(expr_sub) <span class="sc">%in%</span> set2,i]</span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Gene4 Gene5 Gene6</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>      <span class="co"># 0.21  0.22  0.23</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Differences between one gene and gene sets</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Compare expression of each gene in Set1 with all genes in Set2. </span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>      <span class="co"># For example, 0.51&gt;0.21/0.22/0.23, so the value of Gene1:s2 is 3.</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>      res1 <span class="ot">&lt;-</span> <span class="fu">sapply</span>(vals1, <span class="cf">function</span>(v1) <span class="fu">sum</span>(v1 <span class="sc">&gt;</span> vals2, <span class="at">na.rm=</span>T))</span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Gene1:s2   Gene2:s2   Gene3:s2</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>      <span class="co"># 3          3           3</span></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Weight average of gene-geneset rank difference</span></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>      res0[idx] <span class="ot">&lt;-</span> <span class="fu">sum</span>(res1, <span class="at">na.rm =</span> T) <span class="sc">/</span> (<span class="fu">length</span>(vals1) <span class="sc">*</span> <span class="fu">length</span>(vals2))</span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Next gene set pair</span></span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>      idx <span class="ot">&lt;-</span> idx <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a>  resultList[[i]] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(res0)</span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a>resMat <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, resultList)</span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(resMat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(expr_sub)</span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(resMat) <span class="ot">&lt;-</span> featureNames</span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Report </span></span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Set difference across samples: &#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(resMat)</span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Set difference across samples:  </span></span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a><span class="co">#      Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a><span class="co"># s1s2       1       1       1       0       0       0</span></span></code></pre></div>
<p>In <strong>GSClassifier</strong>, we established <strong>makeSetData</strong> to evaluate set difference across samples:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene set difference across samples</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>geneset_interaction <span class="ot">&lt;-</span> GSClassifier<span class="sc">:::</span><span class="fu">makeSetData</span>(expr_sub, geneSet)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Report </span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Set difference across samples: &#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(resMat)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Set difference across samples:  </span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="co">#      Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="co"># s1s2       1       1       1       0       0       0</span></span></code></pre></div>
<p>We have known that the subtype of <strong>Sample 1-3</strong> differs from that of <strong>Sample 4-6</strong>, which revealed the robustness of set difference for subtype indentification.</p>
<p>Based on the structure of TSP in Figure @ref(fig:tsp), TSP matrix of the simulated dataset should be :</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># TSP matrix</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>tsp <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Binned expression</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  expr_binned[gene_bigRank,],</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Pair difference</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  gene_bigRank_pairs,</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Set difference</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>  resMat</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Report </span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;TSP matrix: &#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(tsp)</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="co"># TSP matrix:  </span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="co">#             Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1             3       3       4       1       2       2</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2             4       4       3       2       2       2</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3             4       4       4       2       1       1</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4             1       2       2       3       4       4</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5             2       2       2       4       4       3</span></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6             2       1       1       4       3       4</span></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene2       0       0       1       0       0       1</span></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene3       0       0       0       0       1       1</span></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene4       1       1       1       0       0       0</span></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene5       1       1       1       0       0       0</span></span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene6       1       1       1       0       0       0</span></span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene3       0       0       0       0       1       1</span></span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene4       1       1       1       0       0       0</span></span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene5       1       1       1       0       0       0</span></span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene6       1       1       1       0       0       0</span></span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene4       1       1       1       0       0       0</span></span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene5       1       1       1       0       0       0</span></span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene6       1       1       1       0       0       0</span></span>
<span id="cb17-38"><a href="#cb17-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene5       0       0       1       0       0       1</span></span>
<span id="cb17-39"><a href="#cb17-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene6       0       1       1       0       1       0</span></span>
<span id="cb17-40"><a href="#cb17-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5:Gene6       0       1       1       0       1       0</span></span>
<span id="cb17-41"><a href="#cb17-41" aria-hidden="true" tabindex="-1"></a><span class="co"># s1s2              1       1       1       0       0       0</span></span></code></pre></div>
<p>Have a look at the distribution:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>tsp_df <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(tsp)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(tsp_df,<span class="fu">aes</span>(<span class="at">x=</span>Var2,<span class="at">y=</span>value,<span class="at">fill=</span>Var2)) <span class="sc">+</span> </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_boxplot</span>(<span class="at">outlier.size =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Samples&#39;</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> <span class="st">&#39;Epression&#39;</span>,</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="cn">NULL</span>) </span></code></pre></div>
<!--chapter:end:Flowchart.Rmd-->
</div>
</div>
</div>
<div id="discussion" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Discussion</h1>
<p>In this section, we would discuss some key topics about <strong>GSClassifier</strong>, including <strong>Missing value imputation (MVI)</strong>, <strong>Batch effect</strong>, <strong>hyperparameters</strong>, and so on.</p>
<div id="packages-1" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Packages</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Install &quot;devtools&quot; package</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;luckyBase&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/luckyBase&quot;</span>)</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;**GSClassifier**&quot; package</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;GSClassifier&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/GSClassifier&quot;</span>)</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;pacman&quot; package</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;pacman&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;pacman&quot;</span>)</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Load needed packages</span></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>packages_needed <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;readxl&quot;</span>,</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;ComplexHeatmap&quot;</span>,</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;GSClassifier&quot;</span>,</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;rpart&quot;</span>,</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;tidyr&quot;</span>,</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;reshape2&quot;</span>,</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> packages_needed){<span class="fu">p_load</span>(<span class="at">char=</span>i)}</span></code></pre></div>
<p>Here is the environment of R programming:</p>
<pre><code># R version 4.0.3 (2020-10-10)
# Platform: x86_64-w64-mingw32/x64 (64-bit)
# Running under: Windows 10 x64 (build 18363)
# 
# Matrix products: default
# 
# locale:
# [1] LC_COLLATE=Chinese (Simplified)_China.936 
# [2] LC_CTYPE=Chinese (Simplified)_China.936   
# [3] LC_MONETARY=Chinese (Simplified)_China.936
# [4] LC_NUMERIC=C                              
# [5] LC_TIME=Chinese (Simplified)_China.936    
# 
# attached base packages:
# [1] grid      stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
# [1] ggplot2_3.3.6        reshape2_1.4.4       tidyr_1.2.0         
# [4] rpart_4.1.16         GSClassifier_0.1.22  luckyBase_0.1.0     
# [7] ComplexHeatmap_2.4.3 readxl_1.4.0         pacman_0.5.1        
# 
# loaded via a namespace (and not attached):
#   [1] colorspace_2.0-3     ggsignif_0.6.3       rjson_0.2.21        
#   [4] ellipsis_0.3.2       class_7.3-20         rprojroot_2.0.3     
#   [7] circlize_0.4.15      GlobalOptions_0.1.2  fs_1.5.2            
#  [10] clue_0.3-57          rstudioapi_0.13      ggpubr_0.4.0        
#  [13] listenv_0.8.0        remotes_2.4.2        prodlim_2019.11.13  
#  [16] fansi_1.0.3          lubridate_1.8.0      codetools_0.2-18    
#  [19] splines_4.0.3        doParallel_1.0.17    cachem_1.0.6        
#  [22] knitr_1.30           pkgload_1.2.4        jsonlite_1.8.0      
#  [25] pROC_1.18.0          caret_6.0-92         broom_1.0.0         
#  [28] cluster_2.1.3        png_0.1-7            compiler_4.0.3      
#  [31] backports_1.4.1      assertthat_0.2.1     Matrix_1.2-18       
#  [34] fastmap_1.1.0        cli_3.3.0            htmltools_0.5.2     
#  [37] prettyunits_1.1.1    tools_4.0.3          gtable_0.3.0        
#  [40] glue_1.6.2           dplyr_1.0.9          Rcpp_1.0.8.3        
#  [43] carData_3.0-5        cellranger_1.1.0     vctrs_0.4.1         
#  [46] nlme_3.1-149         iterators_1.0.14     timeDate_3043.102   
#  [49] xfun_0.33            gower_1.0.0          stringr_1.4.0       
#  [52] globals_0.15.1       ps_1.4.0             testthat_3.1.0      
#  [55] lifecycle_1.0.1      devtools_2.4.3       rstatix_0.7.0       
#  [58] future_1.26.1        MASS_7.3-53          scales_1.2.0        
#  [61] ipred_0.9-12         parallel_4.0.3       RColorBrewer_1.1-3  
#  [64] yaml_2.3.5           memoise_2.0.1        stringi_1.7.6       
#  [67] desc_1.4.1           randomForest_4.6-14  foreach_1.5.2       
#  [70] hardhat_1.1.0        pkgbuild_1.3.1       lava_1.6.10         
#  [73] shape_1.4.6          tuneR_1.4.0          rlang_1.0.2         
#  [76] pkgconfig_2.0.3      evaluate_0.15        lattice_0.20-41     
#  [79] purrr_0.3.4          recipes_0.2.0        processx_3.7.0      
#  [82] tidyselect_1.1.2     parallelly_1.32.0    plyr_1.8.7          
#  [85] magrittr_2.0.3       bookdown_0.21        R6_2.5.1            
#  [88] generics_0.1.2       DBI_1.1.3            pillar_1.7.0        
#  [91] withr_2.5.0          survival_3.3-1       abind_1.4-5         
#  [94] nnet_7.3-17          tibble_3.1.7         future.apply_1.9.0  
#  [97] crayon_1.5.1         car_3.1-0            xgboost_1.6.0.1     
# [100] utf8_1.2.2           rmarkdown_2.14       GetoptLong_1.0.5    
# [103] usethis_2.1.3        data.table_1.14.2    callr_3.7.0         
# [106] ModelMetrics_1.2.2.2 digest_0.6.29        stats4_4.0.3        
# [109] signal_0.7-7         munsell_0.5.0        sessioninfo_1.2.2</code></pre>
</div>
<div id="missing-value-imputation-mvi" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Missing value imputation (MVI)</h2>
<!--
Missing Value Imputation (MVI); 
+ Missing Value type: Missing Completely at Random (MCAR), Missing at Random (MAR), and Not Missing at Random (NMAR).
+ The main reason of missing value in the scenario of **GSClassifier**?
+ Why is missing value imputation needed in **GSClassifier** modeling?
+ Routine algorithms of missing value imputation for DNA array (mRNA expression)?
+ Missing value imputation in subtype calling?
+ The flowchart of quantile algorithm?
+ zero == Cutout; https://arxiv.org/abs/1708.04552; In this paper, we show that the simple regularization technique of randomly masking out square regions of input during training, which we call cutout, can be used to improve the robustness and overall performance of convolutional neural networks. Not only is this method extremely easy to implement, but we also demonstrate that it can be used in conjunction with existing forms of data augmentation and other regularizers to further improve model performance. We evaluate this method by applying it to current state-of-the-art architectures on the CIFAR-10, CIFAR-100, and SVHN datasets, yielding new state-of-the-art results of 2.56%, 15.20%, and 1.30% test error respectively.
+ Experiments: leave genes out of validation cohorts and then test the performance of **GSClassifier**-PADi.
-->
<p>Due to reasons like low expression/weak signal, contamination of microarray surfaces, inappropriate manual operations, insufficient resolution or systematic errors during the laboratory process <span class="citation">[<a href="#ref-RN387" role="doc-biblioref">8</a>–<a href="#ref-RN382" role="doc-biblioref">10</a>]</span>, <strong>missing value</strong> in high-imput genetic data is common. Generally, tiny missing value could be just dealed with case deletion, while the biological discovery might be damaged when the missing rate tops 15% <span class="citation">[<a href="#ref-RN392" role="doc-biblioref">11</a>,<a href="#ref-RN386" role="doc-biblioref">12</a>]</span>. Currently, lots of methods, including statistic-based or machine learning-based methods (Figure @ref(fig:mvi01)), had been developed for <strong>missing value imputation (MVI)</strong> <span class="citation">[<a href="#ref-RN386" role="doc-biblioref">12</a>]</span>. Wang et al <span class="citation">[<a href="#ref-RN384" role="doc-biblioref">13</a>]</span> categorized MVI methods into simple (zeros or average),biology knowledge-, global learning-, local learning-, hybrid-based methods. In order to satisfy the working conditions of xgboost <span class="citation">[<a href="#ref-xgboost" role="doc-biblioref">14</a>]</span> functions (<code>xgb.train</code>, <code>xgboost</code>, and <code>xgb.cv</code>) in GSClassifer, the missing value in expression matrix must be deleted or imputation.</p>
<p>In <strong>PAD</strong> project, several strategies were applied to reduce the impact of missing values as possible. First, both <strong>PIAM</strong> and <strong>PIDG</strong> in <strong>PAD</strong> project were curated GEPs that were not be missing in over 80% gastric cancer datasets. Here we showed the actual distribution of missing value across samples in gastric cancer datasets we used.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, </span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>              <span class="st">&quot;testData.rds&quot;</span>, </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Missing value</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>expr_pad_na <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr_pad, <span class="dv">2</span>, </span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>                     <span class="cf">function</span>(x) <span class="fu">sum</span>(<span class="fu">is.na</span>(x))<span class="sc">/</span><span class="fu">length</span>(x))</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>expr_pad_na_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">names</span>(expr_pad_na),</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">prob =</span> <span class="fu">as.numeric</span>(expr_pad_na),</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">stringsAsFactors =</span> F</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>As shown in Figure @ref(fig:mvi02), the percentage of all samples in gastric cancer datasets we used is lower than 8%.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> expr_pad_na_df, </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> prob)) <span class="sc">+</span> </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&#39;identity&#39;</span>, <span class="at">color =</span> mycolor[<span class="dv">3</span>]) <span class="sc">+</span> </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">labels=</span>scales<span class="sc">::</span>percent) <span class="sc">+</span> </span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Samples in gastric cancer cohorts&#39;</span>, </span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Percentage of missing value&#39;</span>) <span class="sc">+</span> </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Second, we did conduct some MVI strategy to deal with data before model training in <strong>GSClassifier</strong>. Due to the low missing rate of our experimental data, we just <strong>set missing value as zero</strong> during model training and subtype identification in the early version of PADi (<strong>PAD.train.v20200110</strong>). The model seemed to be robust in both the internal cohort and external cohorts, and greatly predicted the response to immune checkpoint inhibitors (ICIs) in advanced gastric cancer.</p>
<p>In the new version of PADi (<strong>PAD.train.v20220917</strong>), we designed the so-called <strong>quantile</strong> algorithm for random MVI during <strong>PADi</strong> model training, which also seemed to work well for PADi model training.</p>
<!--

We supposed that ...
Add a flowchart for quantile algorithm

-->
<p>Here, we demonstrated the principle of <strong>quantile</strong> algorithm in the simulated dataset:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulated data</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">&#39;./data/simulated-data.xlsx&#39;</span>, <span class="at">sheet =</span> <span class="st">&#39;RNA&#39;</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>expr0 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(x[,<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr0) <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">as.matrix</span>(x[,<span class="dv">1</span>])); <span class="fu">rm</span>(x)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co"># MVI with Quantile algorithm</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> expr0</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>na.pos <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr,<span class="dv">2</span>,is.one.na)</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">478</span>); seeds <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(expr)<span class="sc">*</span><span class="dv">10</span>, <span class="fu">sum</span>(na.pos), <span class="at">replace =</span> F)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>tSample <span class="ot">&lt;-</span> <span class="fu">names</span>(na.pos)[na.pos]</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>quantile_vector <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span>)<span class="sc">/</span><span class="dv">1000</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tSample)){ <span class="co"># i=1</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  sample.i <span class="ot">&lt;-</span> tSample[i]</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  expr.i <span class="ot">&lt;-</span> expr[, sample.i]</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>  expr.i.max <span class="ot">&lt;-</span> <span class="fu">max</span>(expr.i, <span class="at">na.rm =</span> T)</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>  expr.i.min <span class="ot">&lt;-</span> <span class="fu">min</span>(expr.i, <span class="at">na.rm =</span> T)</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(seeds[i]);</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Details of quantile algorithm</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  expr.i[<span class="fu">is.na</span>(expr.i)] <span class="ot">&lt;-</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>    expr.i.min <span class="sc">+</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>    (expr.i.max<span class="sc">-</span>expr.i.min) <span class="sc">*</span> <span class="fu">sample</span>(quantile_vector,</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>                                     <span class="fu">sum</span>(<span class="fu">is.na</span>(expr.i)),</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">replace =</span> T)</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>  expr[, sample.i] <span class="ot">&lt;-</span> expr.i</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr0)</span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression without NA value:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr)</span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression: </span></span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60    0.21    0.30    0.40</span></span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58    0.22    0.31    0.35</span></span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61      NA    0.29    0.30</span></span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40    0.51    0.52    0.60</span></span>
<span id="cb23-42"><a href="#cb23-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35    0.52    0.54    0.58</span></span>
<span id="cb23-43"><a href="#cb23-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30    0.53      NA    0.61</span></span>
<span id="cb23-44"><a href="#cb23-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09    0.11    0.12    0.14</span></span>
<span id="cb23-45"><a href="#cb23-45" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb23-46"><a href="#cb23-46" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression without NA value: </span></span>
<span id="cb23-47"><a href="#cb23-47" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb23-48"><a href="#cb23-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60 0.21000 0.30000    0.40</span></span>
<span id="cb23-49"><a href="#cb23-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58 0.22000 0.31000    0.35</span></span>
<span id="cb23-50"><a href="#cb23-50" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61 0.43256 0.29000    0.30</span></span>
<span id="cb23-51"><a href="#cb23-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40 0.51000 0.52000    0.60</span></span>
<span id="cb23-52"><a href="#cb23-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35 0.52000 0.54000    0.58</span></span>
<span id="cb23-53"><a href="#cb23-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30 0.53000 0.32622    0.61</span></span>
<span id="cb23-54"><a href="#cb23-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09 0.11000 0.12000    0.14</span></span></code></pre></div>
<p>Look at the new matrix via heatmap, where the clustering result is not obviously disturbed after MVI:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(expr))), <span class="at">name =</span> <span class="st">&quot;Z-score&quot;</span>, <span class="at">column_title =</span> <span class="st">&quot;After MVI&quot;</span>)</span></code></pre></div>
<p>Due to missing value might damage the integrity of biological information, we explored <strong>how much the number of missing value in one sample impacts subtype identification via PADi</strong>. The steps are as following: (i) we used quantile algorithm to do MVI in the internal validation cohort of gastric cancer; (ii) we randomly masked different proportion of genes as zero expression; (iii) we calculated the relative multi-ROC <span class="citation">[<a href="#ref-pROC" role="doc-biblioref">15</a>]</span> (masked data vs. MVI data). In <strong>GSClassifier</strong>, we developed a function called <strong>mv_tolerance</strong> to complete the task.</p>
<ol style="list-style-type: lower-roman">
<li>Load the internal validation cohort:</li>
</ol>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Internal validation cohort</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;testData.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>modelInfo <span class="ot">&lt;-</span> <span class="fu">modelData</span>(</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> testData<span class="sc">$</span>PanSTAD_phenotype_part,</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">id.col =</span> <span class="st">&quot;ID&quot;</span>,</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">variable =</span> <span class="fu">c</span>(<span class="st">&quot;platform&quot;</span>, <span class="st">&quot;PAD_subtype&quot;</span>),</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">Prop =</span> <span class="fl">0.7</span>,</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">19871</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>validInform <span class="ot">&lt;-</span> modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Valid</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>expr_pad_innervalid <span class="ot">&lt;-</span> expr_pad[,validInform<span class="sc">$</span>ID]</span></code></pre></div>
<ol start="2" style="list-style-type: lower-roman">
<li>Missing value tolerance analysis:</li>
</ol>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Time-consuming</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>mvt <span class="ot">&lt;-</span> <span class="fu">mv_tolerance</span>(</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">X =</span> expr_pad_innervalid,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.loss =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">12</span>),</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>),</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;PAD.train_20220916&quot;</span>,</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">487</span>,</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> T</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<ol start="3" style="list-style-type: lower-roman">
<li>multi-ROC analysis:</li>
</ol>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>mvt_auc <span class="ot">&lt;-</span> mvt<span class="sc">$</span>multiAUC</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>mvt_auc_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(mvt_auc)){ <span class="co"># i=1</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  df.i <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">as.integer</span>(<span class="fu">Fastextra</span>(<span class="fu">names</span>(mvt_auc)[i], <span class="st">&#39;=&#39;</span>, <span class="dv">2</span>)),</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">as.numeric</span>(mvt_auc[[i]]<span class="sc">$</span>auc),</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">stringsAsFactors =</span> F</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>  mvt_auc_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(mvt_auc_df, df.i)</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>  </span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mvt_auc_df, <span class="fu">aes</span>(x,y)) <span class="sc">+</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">12</span>)) <span class="sc">+</span> </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_smooth</span>(<span class="at">formula =</span> y <span class="sc">~</span> x,<span class="at">method =</span> <span class="st">&#39;glm&#39;</span>) <span class="sc">+</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;No. of missing value&#39;</span>, </span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Relative AUC in multi-ROC analysis&#39;</span>) <span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>As showed in Figure @ref(fig:mvi03), there is linear negative correlation between the number of missing value (mising rate ranges from 6.25% to 37.5%) and subtype identification performance of <strong>PADi</strong> model. One of the reasons might be that PIAM/PIDG were small GEPs, so little gene loss might significantly impact the performance of <strong>PADi</strong>. By the way, there is no mising value in PIAM/PIDG of the ‘Kim2018’ cohort, an external validation cohort for ICIs therapy response prediction via <strong>PADi</strong>. Nonetheless, we still used <strong>zero strategy</strong> during subtype identification of <strong>PADi</strong> if any missing value exist, because randomization might make the result unstable, which is not suitable for clinical decision.</p>
<!--
Nonetheless, we still used **zero strategy** during subtype identification of **PADi** if any missing value exist, because randomization might make the result unstable, which is not suitable for clinical decision. 
-->
<p>In conclusion, zero or quantile strategy could be applied for MVI before <strong>GSClassifier</strong> model training. However, missing value should be avoided as possible in subtype identification for missing value really damage the performance of <strong>GSClassifier</strong>. Nonetheless, due to low-input GEPs used in <strong>PADi</strong> model (No. of Gene=32), it’s easy to avoid missing value in clinical practice.</p>
</div>
<div id="batch-effect" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Batch effect</h2>
<!--

gene-pair batch effect in Google Scholar

relative expression orderings (REO)

single-sample enrichment analysis 

Batch effect control & gene number

-->
<p><strong>TSP</strong> was widely applied to control batch effects in transciptomic data <span class="citation">[<a href="#ref-RN369" role="doc-biblioref">16</a>–<a href="#ref-RN365" role="doc-biblioref">23</a>]</span>. Still, we tested whether <strong>TSP</strong> is a robust method for batch effect control in real-world data. As demonstrated in Figure @ref(fig:be01), the obvious batch effects across gastric cancer datsets were significantly reduced after <strong>TSP</strong> normalization.</p>
<p>In order to confirmed the association between <strong>gene counts</strong> in modeling and batch effect control via <strong>TSP</strong> normalization, we selected random genes with counts ranging 4, 8, 20, 40, and 80 for TSP matrix establishment. As shown in Figure @ref(fig:be02), <strong>TSP</strong> normalization works greatly in different gene counts for batch effect control compared with raw expression matrix.</p>
<!--
Batch effect reduction of TSP could be explained in three parts:

+ **Binned expression**. Regardless of what raw distribution of a numeric vector is, it would be converted into a binned vector with exactly the same distribution under a fixed quantile vector.

+ **Pair/set difference**.
-->
</div>
<div id="hyperparameters" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Hyperparameters</h2>
<!--
+ modelData(): Prop, seed

+ fitEnsembleModel():

  - n=100
  - sampSize=0.7
  - sampSeed = 2020
  - na.fill.seed = 443
  - breakVec=c(0, 0.25, 0.5, 0.75, 1.0)
  - ptail=0.8/2
  
  - https://www.youtube.com/watch?v=f3ryHJ05h5k
  
  - https://xgboost.readthedocs.io/en/latest/parameter.html
  - Parameters for Tree Booster
  - params = list(max_depth = 10,
                  eta = 0.5,
                  nrounds = 100,
                  nthread = 10,
                  nfold=5)
                  
  所以，在时间非常宝贵的情况下，你必须应用自 己的判断力，并使用小数据样本来找出合适的调优参数，否则硬盘空间可能不足
  nrounds ：最大迭代次数（最终模型中树的数量）。
  colsample_bytree ：建立树时随机抽取的特征数量，用一个比率表示，默认值为1（使
用100%的特征）。
  min_child_weight ：对树进行提升时使用的最小权重，默认为1。
  eta ：学习率，每棵树在最终解中的贡献，默认为0.3。
  gamma ：在树中新增一个叶子分区时所需的最小减损。
  subsample ：子样本数据占整个观测的比例，默认值为1（100%）。
  max_depth ：单个树的最大深度。
-->
<!--chapter:end:Discussion.Rmd-->
</div>
</div>
<div id="quick-start" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Quick start</h1>
<div id="about" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> About</h2>
<ul>
<li>Although with bright prospects in Pan-disease analysis, <a href="https://github.com/huangwb8/GSClassifier">GSClassifier</a> was primarily developed for clinic-friendly immune subtypes of gastric cancer (GC). Currently, only <code>PAD</code> subtypes and <code>PADi</code> for GC were supported. We would try to support more cancer types in the future as possible. More details in <a href="https://github.com/huangwb8/GSClassifier/wiki/Plans-in-the-future">Plans in the future</a> section.</li>
<li>Gibbs' <code>PanCancer immune subtypes</code> based on five gene signatures (485 genes) could also be called in <code>GSClassifier</code>, with a pre-trained model from the <a href="https://github.com/CRI-iAtlas/ImmuneSubtypeClassifier">ImmuneSubtypeClassifier</a> package. If you use their jobs, please cite: <a href="https://github.com/huangwb8/GSClassifier/wiki/Introduction#Reference">references</a>.</li>
<li>Particularly, all normal tissues should be eliminated before subtypes calling for cancer research.</li>
</ul>
</div>
<div id="data" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Data</h2>
<p>To lower the learning cost of <code>GSClassifier</code>, we provides some test data:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GSClassifier)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co"># 载入需要的程辑包：luckyBase</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;testData.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span></code></pre></div>
<p>Explore the <code>testData</code>:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(testData)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Kim2018_3&quot;              &quot;PanSTAD_phenotype_part&quot; &quot;PanSTAD_expr_part&quot;</span></span></code></pre></div>
</div>
<div id="pad" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> PAD</h2>
<div id="preparation-of-the-test-data" class="section level3" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Preparation of the test data</h3>
<p>load phenotype data:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_phenotype_part</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(design<span class="sc">$</span>Dataset)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  GSE13861  GSE13911  GSE15459  GSE22377  GSE26899  GSE26901  GSE29272  GSE51105 </span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">#        65        39       192        43        96       108       134        94 </span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  GSE54129  GSE57303  GSE62254  GSE65801  GSE84437 TCGA-STAD </span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co">#       111        70       300        32       433       372</span></span></code></pre></div>
<p>load target sample IDs in <code>GSE54129</code> cohort:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>target_ID <span class="ot">&lt;-</span> design<span class="sc">$</span>ID[design<span class="sc">$</span>Dataset <span class="sc">%in%</span> <span class="st">&#39;GSE54129&#39;</span>]</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part[,target_ID]</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(expr[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co">#                 GSM1308413 GSM1308414 GSM1308415 GSM1308416 GSM1308417</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000122122   7.888349   7.623663   6.873493   6.961102   7.150572</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000117091   7.051760   6.217445   5.651839   5.830996   5.908532</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000163219   6.056472   5.681844   5.411533   5.652684   5.555147</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000136167   9.322191   8.765794   8.502315   8.838166   8.845952</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000005844   7.119594   6.023631   5.400999   6.172863   6.059838</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000123338   7.204051   6.925328   6.259809   6.610681   6.595882</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                 GSM1308418 GSM1308419 GSM1308420 GSM1308421 GSM1308422</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000122122   7.871423   6.953329   8.334037   6.764335   6.522554</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000117091   6.526917   5.646446   6.617520   5.637693   5.742848</span></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000163219   5.962885   5.361763   5.975842   5.330428   5.172705</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000136167   9.366074   8.675718   9.118517   8.614068   8.114096</span></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000005844   6.523530   6.129181   7.331588   5.547059   5.867118</span></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000123338   6.699790   6.935390   7.050288   6.536710   6.200269</span></span></code></pre></div>
</div>
<div id="unsupervised-clustering" class="section level3" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Unsupervised clustering</h3>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>res_pad <span class="ot">&lt;-</span> <span class="fu">PAD</span>(</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">expr =</span> expr,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster.method =</span> <span class="st">&quot;ward.D2&quot;</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">extra.annot =</span> <span class="cn">NULL</span>,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot.title =</span> <span class="st">&#39;PAD test&#39;</span>,</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">subtype =</span> <span class="st">&quot;PAD.train_20220916&quot;</span>,</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> T</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="Quick-start_files/figure-html/unnamed-chunk-5-1.png" width="60%" style="display: block; margin: auto;" /></p>
<pre><code># Use default PIAM... 
# Use default PIDG... 
# Gene match: 100%. 
# Done!</code></pre>
</div>
<div id="of-note" class="section level3" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Of note</h3>
<ul>
<li><p>It’s strongly recommanded that the gene type of <code>expr</code> should be alway the same, such as ENSEMBL genes (ENSG00000111640 for GAPDH, for examples).</p></li>
<li><p><code>PAD</code> function is only for datasets with lots of samples for its classification depends on population-based unsupervised clusting. <code>PAD</code> is population-dependent and non-personalized.</p></li>
<li><p>Beta characteristics: You could try random forest classification based on the <code>randomForest</code> package or methods in <code>stats::hclust</code>.</p></li>
</ul>
</div>
</div>
<div id="padi" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> PADi</h2>
<ul>
<li><p>In <code>GSClassifier</code>, <code>PADi</code> is a pre-trained out-of-the-box model for GC personalized PAD subtypes calling.</p></li>
<li><p>During the sutype calling, the gene rank relations based on individuals instead of the relative values across samples would be used. Thus, <strong>you don’t have to do batch normalization</strong> even though the data (the <code>X</code> input) come from multiple cohorts or platform.</p></li>
<li><p>More limitations were discussed in our paper that you had better know.</p></li>
</ul>
<p>In this section, we would showed how to use <code>PADi</code> series: <code>PADi</code>, <code>callEnsemble</code>, and <code>parCallEnsemble</code> functions.</p>
<div id="preparation-of-the-test-data-1" class="section level3" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Preparation of the test data</h3>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a> X <span class="ot">&lt;-</span> testData<span class="sc">$</span>Kim2018_3</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">head</span>(X)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co">#                   PB-16-002   PB-16-003   PB-16-004</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000121410  0.07075272 -2.08976724 -1.43569557</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000148584 -1.49631022 -0.23917056  0.94827471</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000175899 -0.77315329  0.52163146  0.91264015</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000166535 -0.28860715 -0.45964255 -0.38401295</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000256069 -0.25034243  0.06863867  0.14429081</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000184389  0.08215945 -0.05966481  0.04937924</span></span></code></pre></div>
</div>
<div id="use-a-specific-function-called-padi" class="section level3" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> Use a specific function called <code>PADi</code></h3>
<p>Very simple, just:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>res_padi <span class="ot">&lt;-</span> <span class="fu">PADi</span>(<span class="at">X =</span> X, <span class="at">verbose =</span> F)</span></code></pre></div>
<p>Check the result:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_padi)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   SampleIDs BestCall BestCall_Max           1          2          3         4</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 PB-16-002        4            4 0.023372779 0.02631794 0.04864784 0.3336484</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 PB-16-003        4            4 0.007271698 0.08650742 0.01974812 0.9530730</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 PB-16-004        4            4 0.011559768 0.02922151 0.09018894 0.8649045</span></span></code></pre></div>
<p>Actually, <code>PADi</code> is exactly based on a general function called <code>callEnsemble</code>.</p>
</div>
<div id="use-the-callensemble-function" class="section level3" number="3.4.3">
<h3><span class="header-section-number">3.4.3</span> Use the <code>callEnsemble</code> function</h3>
<p>Also simple, just:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>res_padi <span class="ot">&lt;-</span> <span class="fu">callEnsemble</span>(</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> X,</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ens =</span> <span class="cn">NULL</span>,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneAnnotation =</span> <span class="cn">NULL</span>,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneSet =</span> <span class="cn">NULL</span>,</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaller =</span> <span class="cn">NULL</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneid =</span> <span class="st">&quot;ensembl&quot;</span>,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtype =</span> <span class="st">&quot;PAD.train_20220916&quot;</span>,</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> F</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Check the result:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_padi)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   SampleIDs BestCall BestCall_Max          1          2          3         4</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 PB-16-002        4            4 0.01338872 0.01624520 0.03965218 0.8052567</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 PB-16-003        4            4 0.04709511 0.08833681 0.03879361 0.6244038</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 PB-16-004        4            4 0.01389035 0.03638009 0.05852707 0.6980438</span></span></code></pre></div>
</div>
<div id="parallel-strategy-for-padi" class="section level3" number="3.4.4">
<h3><span class="header-section-number">3.4.4</span> Parallel strategy for PADi</h3>
<ul>
<li><p>Sometimes, the number of patients for subtype callings could be huge (hundreds or even tens of thousands). Thus, the parallel computing (Windows or Linux pass; not tested in Mac or other OS) was also developed in the current version of <code>GSClassifier</code> package.</p></li>
<li><p>The parameter <code>numCores</code> was used to control the No. of CPU for computing (which depends on your CPU capacity).</p></li>
</ul>
<pre><code># No run for the tiny test data. With errors.

# Method 1:
res_padi &lt;- PADi(X = X, verbose = F, numCores = 4)

# Method 2: 
res_padi &lt;- parCallEnsemble(
  X = X,
  ens = NULL,
  geneAnnotation = NULL,
  geneSet = NULL,
  scaller = NULL,
  geneids = &#39;ensembl&#39;,
  subtype = &#39;PAD.train_20220916&#39;,
  verbose = T,
  numCores = 4)</code></pre>
</div>
<div id="single-sample-subtype-calling" class="section level3" number="3.4.5">
<h3><span class="header-section-number">3.4.5</span> Single sample subtype calling</h3>
<p>In clinical practice, the single sample subtype calling might be one of the most common scenarios and is also supported by functions of <code>PADi</code> series.</p>
<p>Supposed that there is a GC patient, its information should be:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>X_ind <span class="ot">&lt;-</span> X[,<span class="dv">1</span>]; <span class="fu">names</span>(X_ind) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(X)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X_ind)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000121410 ENSG00000148584 ENSG00000175899 ENSG00000166535 ENSG00000256069 </span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co">#      0.07075272     -1.49631022     -0.77315329     -0.28860715     -0.25034243 </span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000184389 </span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co">#      0.08215945</span></span></code></pre></div>
<p>Or it can also be another format:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>X_ind <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(X[,<span class="dv">1</span>]); <span class="fu">rownames</span>(X_ind) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(X)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X_ind)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="co">#                        [,1]</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000121410  0.07075272</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000148584 -1.49631022</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000175899 -0.77315329</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000166535 -0.28860715</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000256069 -0.25034243</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="co"># ENSG00000184389  0.08215945</span></span></code></pre></div>
<p>Similar to multiples sample calling, just:</p>
<p>check the result:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_padi)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   SampleIDs BestCall BestCall_Max          1          2          3         4</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 1    target        4            4 0.02337278 0.02631794 0.04864784 0.3336484</span></span></code></pre></div>
<p>Similarly, there is alternative choice:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>res_padi <span class="ot">&lt;-</span> <span class="fu">callEnsemble</span>(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> X_ind,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ens =</span> <span class="cn">NULL</span>,</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneAnnotation =</span> <span class="cn">NULL</span>,</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneSet =</span> <span class="cn">NULL</span>,</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaller =</span> <span class="cn">NULL</span>,</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneid =</span> <span class="st">&quot;ensembl&quot;</span>,</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtype =</span> <span class="st">&quot;PAD.train_20220916&quot;</span>,</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> F</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Check the result:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_padi)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   SampleIDs BestCall BestCall_Max          1         2          3         4</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 1    target        4            4 0.01338872 0.0162452 0.03965218 0.8052567</span></span></code></pre></div>
</div>
<div id="of-note-1" class="section level3" number="3.4.6">
<h3><span class="header-section-number">3.4.6</span> Of note</h3>
<ul>
<li><p>In the results of <code>PADi</code>, two types of subtypes (<code>BestCall</code> and <code>BestCall_Max</code>) were integrated. <code>BestCall</code> was predicted based on a xgboost-trained model based on prior knowlege of <code>PAD</code> subtypes and the possibility matrix (columns 4 to 7 of four-subtype calling, for example), while <code>BestCall_Max</code> was predicted via maximum strategy. You should use THE SAME ONE in a specific practice no matter which one you use.</p></li>
<li><p><code>PADi</code> is individual-dependent and personalized, which means that the result of subtype calling would not be influenced by the data of others.</p></li>
</ul>
</div>
</div>
<div id="use-external-models-from-luckymodel-package" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Use external models from luckyModel package</h2>
<p>In the future, there might be lots of models available as a resource of <code>GSClassifier</code>, such as <a href="https://github.com/huangwb8/luckyModel">luckyModel</a>. Here we show how <code>luckyModel</code> support <code>GSClassifier</code>.</p>
<p>First, intall and load <code>luckyModel</code>:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install luckyModel</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;luckyModel&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/luckyModel&quot;</span>)</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(luckyModel)</span></code></pre></div>
<p>Check projects supported in current <code>luckyModel</code>:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list_project</span>()</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;GSClassifier&quot;</span></span></code></pre></div>
<p>Check available models in the project:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list_model</span>(<span class="at">project=</span><span class="st">&#39;GSClassifier&#39;</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Available models in GSClassifier:</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   *Gibbs_PanCancerImmuneSubtype_v20190731</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   *HWB_PAD_v20200110</span></span></code></pre></div>
<p>Here, <code>HWB_PAD_v20200110</code> is a standard name of <code>PADi</code>. They are exactly the same.</p>
<p>Taking <code>PADi</code> as an example, we here show how to use an external model from <code>luckyModel</code>. First, load a model:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">lucky_model</span>(<span class="at">project =</span> <span class="st">&#39;GSClassifier&#39;</span>,</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">developer=</span><span class="st">&#39;HWB&#39;</span>,</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">model =</span> <span class="st">&#39;PAD&#39;</span>,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">version =</span> <span class="st">&#39;v20200110&#39;</span>)</span></code></pre></div>
<p>Then, check the gene id type:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>model<span class="sc">$</span>geneSet</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="co"># $PIAM</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;ENSG00000122122&quot; &quot;ENSG00000117091&quot; &quot;ENSG00000163219&quot; &quot;ENSG00000136167&quot;</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co">#  [5] &quot;ENSG00000005844&quot; &quot;ENSG00000123338&quot; &quot;ENSG00000102879&quot; &quot;ENSG00000010671&quot;</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  [9] &quot;ENSG00000185862&quot; &quot;ENSG00000104814&quot; &quot;ENSG00000134516&quot; &quot;ENSG00000100055&quot;</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] &quot;ENSG00000082074&quot; &quot;ENSG00000113263&quot; &quot;ENSG00000153283&quot; &quot;ENSG00000198821&quot;</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="co"># [17] &quot;ENSG00000185811&quot; &quot;ENSG00000117090&quot; &quot;ENSG00000171608&quot;</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a><span class="co"># $PIDG</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;ENSG00000116667&quot; &quot;ENSG00000107771&quot; &quot;ENSG00000196782&quot; &quot;ENSG00000271447&quot;</span></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="co">#  [5] &quot;ENSG00000173517&quot; &quot;ENSG00000134686&quot; &quot;ENSG00000100614&quot; &quot;ENSG00000134247&quot;</span></span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a><span class="co">#  [9] &quot;ENSG00000109686&quot; &quot;ENSG00000197321&quot; &quot;ENSG00000179981&quot; &quot;ENSG00000187189&quot;</span></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] &quot;ENSG00000140836&quot;</span></span></code></pre></div>
<p>The model should use <code>ensembl</code> as the value of <code>geneid</code> parameter in <code>callEnsemble</code> series.</p>
<p>Next, you can use the model like:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>res_padi <span class="ot">&lt;-</span> <span class="fu">callEnsemble</span>(</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">X =</span> X,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ens =</span> model<span class="sc">$</span>ens<span class="sc">$</span>Model,</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">geneAnnotation =</span> model<span class="sc">$</span>geneAnnotation,</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">geneSet =</span> model<span class="sc">$</span>geneSet,</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">scaller =</span> model<span class="sc">$</span>scaller<span class="sc">$</span>Model,</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">geneid =</span> <span class="st">&quot;ensembl&quot;</span>,</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">subtype =</span> <span class="cn">NULL</span>,</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> F</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Or just:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>res_padi <span class="ot">&lt;-</span> <span class="fu">callEnsemble</span>(</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  X,</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ens =</span> <span class="cn">NULL</span>,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">geneAnnotation =</span> <span class="cn">NULL</span>,</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">geneSet =</span> <span class="cn">NULL</span>,</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">scaller =</span> <span class="cn">NULL</span>,</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">geneid =</span> <span class="st">&quot;ensembl&quot;</span>,</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">subtype =</span> model, </span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> F</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>They are exactly the same.</p>
<p>Finally, check the result:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_padi)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   SampleIDs BestCall BestCall_Max           1          2          3         4</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 PB-16-002        4            4 0.023372779 0.02631794 0.04864784 0.3336484</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 PB-16-003        4            4 0.007271698 0.08650742 0.01974812 0.9530730</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 PB-16-004        4            4 0.011559768 0.02922151 0.09018894 0.8649045</span></span></code></pre></div>
</div>
<div id="pancancer-immune-subtypes" class="section level2" number="3.6">
<h2><span class="header-section-number">3.6</span> PanCancer immune subtypes</h2>
<p><code>GSClassifier</code> could also call the <code>PanCancer immune subtypes</code> of Gibbs’.</p>
<p>First, see data available in current <code>GSClassifier</code>:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">GSClassifier_Data</span>()</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Available data:</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage example:</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype.rds </span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20200110.rds </span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20220916.rds </span></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;PAD.train_20200110.rds&quot;, package = &quot;GSClassifier&quot;)) </span></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;ImmuneSubtype.rds&quot;, package = &quot;GSClassifier&quot;))</span></span></code></pre></div>
<p>Let’s use our test data to do this:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> testData<span class="sc">$</span>Kim2018_3</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>symbol <span class="ot">&lt;-</span> <span class="fu">convert</span>(<span class="fu">rownames</span>(X))</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(X) <span class="ot">&lt;-</span> symbol</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> X[<span class="sc">!</span><span class="fu">is.na</span>(symbol),]</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(X)</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 19118     3</span></span></code></pre></div>
<p>Have a check</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="co">#                 PB-16-002   PB-16-003   PB-16-004</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co"># A1BG           0.07075272 -2.08976724 -1.43569557</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="co"># A1CF          -1.49631022 -0.23917056  0.94827471</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="co"># A2M           -0.77315329  0.52163146  0.91264015</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a><span class="co"># A2ML1         -0.28860715 -0.45964255 -0.38401295</span></span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a><span class="co"># RP11-118B22.6 -0.25034243  0.06863867  0.14429081</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a><span class="co"># A3GALT2        0.08215945 -0.05966481  0.04937924</span></span></code></pre></div>
<p>PanCan Immune Subtype callings:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>res_pis <span class="ot">&lt;-</span> <span class="fu">callEnsemble</span>(</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> X,</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ens =</span> <span class="cn">NULL</span>,</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneAnnotation =</span> <span class="cn">NULL</span>,</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneSet =</span> <span class="cn">NULL</span>,</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaller =</span> <span class="cn">NULL</span>,</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneid =</span> <span class="st">&quot;symbol&quot;</span>,</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtype =</span> <span class="st">&quot;ImmuneSubtype&quot;</span>,</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> F</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Check the result:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_pis)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   SampleIDs BestCall BestCall_Max            1           2            3</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 PB-16-002        2            2 1.078418e-03 0.561300665 6.356077e-06</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 PB-16-003        4            4 5.370724e-07 0.018167170 1.505137e-04</span></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 PB-16-004        3            3 3.805509e-06 0.001126488 3.622906e-01</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="co">#              4           5           6</span></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 0.2321599349 0.009215117 0.003811775</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 0.6093911827 0.002128245 0.002645677</span></span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 0.0002729274 0.005802790 0.007748680</span></span></code></pre></div>
<p>Also, you can try to use <code>luckyModel</code>:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>pci <span class="ot">&lt;-</span> <span class="fu">lucky_model</span>(</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">&quot;GSClassifier&quot;</span>,</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;PanCancerImmuneSubtype&quot;</span>,</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">developer =</span> <span class="st">&quot;Gibbs&quot;</span>,</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">version =</span> <span class="st">&quot;v20190731&quot;</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>PanCan Immune Subtype callings:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>res_pis <span class="ot">&lt;-</span> <span class="fu">callEnsemble</span>(</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> X,</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ens =</span> <span class="cn">NULL</span>,</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneAnnotation =</span> <span class="cn">NULL</span>,</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneSet =</span> <span class="cn">NULL</span>,</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">scaller =</span> <span class="cn">NULL</span>,</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">geneid =</span> <span class="st">&quot;symbol&quot;</span>,</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtype =</span> pci,</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> F</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Finally, we take a look at the <code>PanCancer immune subtypes</code> model:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>ImmuneSubtype <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;ImmuneSubtype.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(ImmuneSubtype)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;ens&quot;            &quot;scaller&quot;        &quot;geneAnnotation&quot; &quot;geneSet&quot;</span></span></code></pre></div>
<p>Its gene annotation:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ImmuneSubtype<span class="sc">$</span>geneAnnotation)</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="co">#      SYMBOL ENTREZID         ENSEMBL</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 235  ACTL6A       86 ENSG00000136518</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 294   ADAM9     8754 ENSG00000168615</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 305 ADAMTS1     9510 ENSG00000154734</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 322    ADAR      103 ENSG00000160710</span></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 340   ADCY7      113 ENSG00000121281</span></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 479   AIMP2     7965 ENSG00000106305</span></span></code></pre></div>
<p>Its gene sets:</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>ImmuneSubtype<span class="sc">$</span>geneSet</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co"># $LIexpression_score</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;CCL5&quot;  &quot;CD19&quot;  &quot;CD37&quot;  &quot;CD3D&quot;  &quot;CD3E&quot;  &quot;CD3G&quot;  &quot;CD3Z&quot;  &quot;CD79A&quot; &quot;CD79B&quot;</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [10] &quot;CD8A&quot;  &quot;CD8B1&quot; &quot;IGHG3&quot; &quot;IGJ&quot;   &quot;IGLC1&quot; &quot;CD14&quot;  &quot;LCK&quot;   &quot;LTB&quot;   &quot;MS4A1&quot;</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="co"># $CSF1_response</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   [1] &quot;CORO1A&quot;   &quot;MNDA&quot;     &quot;CCRL2&quot;    &quot;SLC7A7&quot;   &quot;HLA-DMA&quot;  &quot;FYB&quot;     </span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a><span class="co">#   [7] &quot;RNASE6&quot;   &quot;TLR2&quot;     &quot;CTSC&quot;     &quot;LILRB4&quot;   &quot;PSCDBP&quot;   &quot;CTSS&quot;    </span></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a><span class="co">#  [13] &quot;RASSF4&quot;   &quot;MSN&quot;      &quot;CYBB&quot;     &quot;LAPTM5&quot;   &quot;DOCK2&quot;    &quot;FCGR1A&quot;  </span></span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a><span class="co">#  [19] &quot;EVI2B&quot;    &quot;ADCY7&quot;    &quot;CD48&quot;     &quot;ARHGAP15&quot; &quot;ARRB2&quot;    &quot;SYK&quot;     </span></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a><span class="co">#  [25] &quot;BTK&quot;      &quot;TNFAIP3&quot;  &quot;FCGR2A&quot;   &quot;VSIG4&quot;    &quot;FPRL2&quot;    &quot;IL10RA&quot;  </span></span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a><span class="co">#  [31] &quot;IFI16&quot;    &quot;ITGB2&quot;    &quot;IL7R&quot;     &quot;TBXAS1&quot;   &quot;FMNL1&quot;    &quot;FLI1&quot;    </span></span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a><span class="co">#  [37] &quot;RASSF2&quot;   &quot;LYZ&quot;      &quot;CD163&quot;    &quot;CD97&quot;     &quot;CCL2&quot;     &quot;FCGR2B&quot;  </span></span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a><span class="co">#  [43] &quot;MERTK&quot;    &quot;CD84&quot;     &quot;CD53&quot;     &quot;CD86&quot;     &quot;HMHA1&quot;    &quot;CTSL&quot;    </span></span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a><span class="co">#  [49] &quot;EVI2A&quot;    &quot;TNFRSF1B&quot; &quot;CXCR4&quot;    &quot;LCP1&quot;     &quot;SAMHD1&quot;   &quot;CPVL&quot;    </span></span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a><span class="co">#  [55] &quot;HLA-DRB1&quot; &quot;C13orf18&quot; &quot;GIMAP4&quot;   &quot;SAMSN1&quot;   &quot;PLCG2&quot;    &quot;OSBPL3&quot;  </span></span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a><span class="co">#  [61] &quot;CD8A&quot;     &quot;RUNX3&quot;    &quot;FCGR3A&quot;   &quot;AMPD3&quot;    &quot;MYO1F&quot;    &quot;CECR1&quot;   </span></span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a><span class="co">#  [67] &quot;LYN&quot;      &quot;MPP1&quot;     &quot;LRMP&quot;     &quot;FGL2&quot;     &quot;NCKAP1L&quot;  &quot;HCLS1&quot;   </span></span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a><span class="co">#  [73] &quot;SELL&quot;     &quot;CASP1&quot;    &quot;SELPLG&quot;   &quot;CD33&quot;     &quot;GPNMB&quot;    &quot;NCF2&quot;    </span></span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a><span class="co">#  [79] &quot;FNBP1&quot;    &quot;IL18&quot;     &quot;B2M&quot;      &quot;SP140&quot;    &quot;FCER1G&quot;   &quot;LCP2&quot;    </span></span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a><span class="co">#  [85] &quot;LY86&quot;     &quot;LAIR1&quot;    &quot;IFI30&quot;    &quot;TNFSF13B&quot; &quot;LST1&quot;     &quot;FGR&quot;     </span></span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a><span class="co">#  [91] &quot;NPL&quot;      &quot;PLEK&quot;     &quot;CCL5&quot;     &quot;PTPRC&quot;    &quot;GNPTAB&quot;   &quot;SLC1A3&quot;  </span></span>
<span id="cb63-23"><a href="#cb63-23" aria-hidden="true" tabindex="-1"></a><span class="co">#  [97] &quot;HCK&quot;      &quot;NPC2&quot;     &quot;C3AR1&quot;    &quot;PIK3CG&quot;   &quot;DAPK1&quot;    &quot;ALOX5AP&quot; </span></span>
<span id="cb63-24"><a href="#cb63-24" aria-hidden="true" tabindex="-1"></a><span class="co"># [103] &quot;CSF1R&quot;    &quot;CUGBP2&quot;   &quot;APOE&quot;     &quot;APOC1&quot;    &quot;CD52&quot;     &quot;LHFPL2&quot;  </span></span>
<span id="cb63-25"><a href="#cb63-25" aria-hidden="true" tabindex="-1"></a><span class="co"># [109] &quot;C1orf54&quot;  &quot;IKZF1&quot;    &quot;SH2B3&quot;    &quot;WIPF1&quot;   </span></span>
<span id="cb63-26"><a href="#cb63-26" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb63-27"><a href="#cb63-27" aria-hidden="true" tabindex="-1"></a><span class="co"># $Module3_IFN_score</span></span>
<span id="cb63-28"><a href="#cb63-28" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;IFI44&quot;  &quot;IFI44L&quot; &quot;DDX58&quot;  &quot;IFI6&quot;   &quot;IFI27&quot;  &quot;IFIT2&quot;  &quot;IFIT1&quot;  &quot;IFIT3&quot; </span></span>
<span id="cb63-29"><a href="#cb63-29" aria-hidden="true" tabindex="-1"></a><span class="co">#  [9] &quot;CXCL10&quot; &quot;MX1&quot;    &quot;OAS1&quot;   &quot;OAS2&quot;   &quot;OAS3&quot;   &quot;HERC5&quot;  &quot;SAMD9&quot;  &quot;HERC6&quot; </span></span>
<span id="cb63-30"><a href="#cb63-30" aria-hidden="true" tabindex="-1"></a><span class="co"># [17] &quot;DDX60&quot;  &quot;RTP4&quot;   &quot;IFIH1&quot;  &quot;STAT1&quot;  &quot;TAP1&quot;   &quot;OASL&quot;   &quot;RSAD2&quot;  &quot;ISG15&quot; </span></span>
<span id="cb63-31"><a href="#cb63-31" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb63-32"><a href="#cb63-32" aria-hidden="true" tabindex="-1"></a><span class="co"># $TGFB_score_21050467</span></span>
<span id="cb63-33"><a href="#cb63-33" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;MMP3&quot;     &quot;MARCKSL1&quot; &quot;IGF2R&quot;    &quot;LAMB1&quot;    &quot;SPARC&quot;    &quot;FN1&quot;     </span></span>
<span id="cb63-34"><a href="#cb63-34" aria-hidden="true" tabindex="-1"></a><span class="co">#  [7] &quot;ITGA4&quot;    &quot;SMO&quot;      &quot;MMP19&quot;    &quot;ITGB8&quot;    &quot;ITGA5&quot;    &quot;NID1&quot;    </span></span>
<span id="cb63-35"><a href="#cb63-35" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] &quot;TIMP1&quot;    &quot;SEMA3F&quot;   &quot;RHOQ&quot;     &quot;CTNNB1&quot;   &quot;MMP2&quot;     &quot;SERPINE1&quot;</span></span>
<span id="cb63-36"><a href="#cb63-36" aria-hidden="true" tabindex="-1"></a><span class="co"># [19] &quot;EPHB2&quot;    &quot;COL16A1&quot;  &quot;EPHA2&quot;    &quot;TNC&quot;      &quot;JUP&quot;      &quot;ITGA3&quot;   </span></span>
<span id="cb63-37"><a href="#cb63-37" aria-hidden="true" tabindex="-1"></a><span class="co"># [25] &quot;TCF7L2&quot;   &quot;COL3A1&quot;   &quot;CDH6&quot;     &quot;WNT2B&quot;    &quot;ADAM9&quot;    &quot;DSP&quot;     </span></span>
<span id="cb63-38"><a href="#cb63-38" aria-hidden="true" tabindex="-1"></a><span class="co"># [31] &quot;HSPG2&quot;    &quot;ARHGAP1&quot;  &quot;ITGB5&quot;    &quot;IGFBP5&quot;   &quot;ARHGDIA&quot;  &quot;LRP1&quot;    </span></span>
<span id="cb63-39"><a href="#cb63-39" aria-hidden="true" tabindex="-1"></a><span class="co"># [37] &quot;IGFBP2&quot;   &quot;CTNNA1&quot;   &quot;LRRC17&quot;   &quot;MMP14&quot;    &quot;NEO1&quot;     &quot;EFNA5&quot;   </span></span>
<span id="cb63-40"><a href="#cb63-40" aria-hidden="true" tabindex="-1"></a><span class="co"># [43] &quot;ITGB3&quot;    &quot;EPHB3&quot;    &quot;CD44&quot;     &quot;IGFBP4&quot;   &quot;TNFRSF1A&quot; &quot;RAC1&quot;    </span></span>
<span id="cb63-41"><a href="#cb63-41" aria-hidden="true" tabindex="-1"></a><span class="co"># [49] &quot;PXN&quot;      &quot;PLAT&quot;     &quot;COL8A1&quot;   &quot;WNT8B&quot;    &quot;IGFBP3&quot;   &quot;RHOA&quot;    </span></span>
<span id="cb63-42"><a href="#cb63-42" aria-hidden="true" tabindex="-1"></a><span class="co"># [55] &quot;EPHB4&quot;    &quot;MMP1&quot;     &quot;PAK1&quot;     &quot;MTA1&quot;     &quot;THBS2&quot;    &quot;CSPG2&quot;   </span></span>
<span id="cb63-43"><a href="#cb63-43" aria-hidden="true" tabindex="-1"></a><span class="co"># [61] &quot;MMP17&quot;    &quot;CD59&quot;     &quot;DVL3&quot;     &quot;RHOB&quot;     &quot;COL6A3&quot;   &quot;NOTCH2&quot;  </span></span>
<span id="cb63-44"><a href="#cb63-44" aria-hidden="true" tabindex="-1"></a><span class="co"># [67] &quot;BSG&quot;      &quot;MMP11&quot;    &quot;COL1A2&quot;   &quot;ZYX&quot;      &quot;RND3&quot;     &quot;THBS1&quot;   </span></span>
<span id="cb63-45"><a href="#cb63-45" aria-hidden="true" tabindex="-1"></a><span class="co"># [73] &quot;RHOG&quot;     &quot;ICAM1&quot;    &quot;LAMA4&quot;    &quot;DVL1&quot;     &quot;PAK2&quot;     &quot;ITGB2&quot;   </span></span>
<span id="cb63-46"><a href="#cb63-46" aria-hidden="true" tabindex="-1"></a><span class="co"># [79] &quot;COL6A1&quot;   &quot;FGD1&quot;    </span></span>
<span id="cb63-47"><a href="#cb63-47" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb63-48"><a href="#cb63-48" aria-hidden="true" tabindex="-1"></a><span class="co"># $CHANG_CORE_SERUM_RESPONSE_UP</span></span>
<span id="cb63-49"><a href="#cb63-49" aria-hidden="true" tabindex="-1"></a><span class="co">#   [1] &quot;CEP78&quot;     &quot;LSM3&quot;      &quot;LRRC40&quot;    &quot;STK17A&quot;    &quot;RPN1&quot;      &quot;JUNB&quot;     </span></span>
<span id="cb63-50"><a href="#cb63-50" aria-hidden="true" tabindex="-1"></a><span class="co">#   [7] &quot;NUP85&quot;     &quot;FLNC&quot;      &quot;HMGN2&quot;     &quot;RPP40&quot;     &quot;UQCR10&quot;    &quot;AIMP2&quot;    </span></span>
<span id="cb63-51"><a href="#cb63-51" aria-hidden="true" tabindex="-1"></a><span class="co">#  [13] &quot;CHEK1&quot;     &quot;VTA1&quot;      &quot;EXOSC8&quot;    &quot;CENPO&quot;     &quot;PNO1&quot;      &quot;SLC16A1&quot;  </span></span>
<span id="cb63-52"><a href="#cb63-52" aria-hidden="true" tabindex="-1"></a><span class="co">#  [19] &quot;WDR77&quot;     &quot;UBE2J1&quot;    &quot;NOP16&quot;     &quot;NUDT1&quot;     &quot;SMC2&quot;      &quot;SLC25A5&quot;  </span></span>
<span id="cb63-53"><a href="#cb63-53" aria-hidden="true" tabindex="-1"></a><span class="co">#  [25] &quot;NUPL1&quot;     &quot;DLEU2&quot;     &quot;PDAP1&quot;     &quot;CCBL2&quot;     &quot;COX17&quot;     &quot;BCCIP&quot;    </span></span>
<span id="cb63-54"><a href="#cb63-54" aria-hidden="true" tabindex="-1"></a><span class="co">#  [31] &quot;PLG&quot;       &quot;RGS8&quot;      &quot;SNRPC&quot;     &quot;PLK4&quot;      &quot;NUTF2&quot;     &quot;LSM4&quot;     </span></span>
<span id="cb63-55"><a href="#cb63-55" aria-hidden="true" tabindex="-1"></a><span class="co">#  [37] &quot;SMS&quot;       &quot;EBNA1BP2&quot;  &quot;C13orf27&quot;  &quot;VDAC1&quot;     &quot;PSMD14&quot;    &quot;MYCBP&quot;    </span></span>
<span id="cb63-56"><a href="#cb63-56" aria-hidden="true" tabindex="-1"></a><span class="co">#  [43] &quot;SMURF2&quot;    &quot;GNG11&quot;     &quot;F3&quot;        &quot;IL7R&quot;      &quot;BRIP1&quot;     &quot;HNRNPA2B1&quot;</span></span>
<span id="cb63-57"><a href="#cb63-57" aria-hidden="true" tabindex="-1"></a><span class="co">#  [49] &quot;DCK&quot;       &quot;ALKBH7&quot;    &quot;HN1L&quot;      &quot;MSN&quot;       &quot;TPM1&quot;      &quot;HYLS1&quot;    </span></span>
<span id="cb63-58"><a href="#cb63-58" aria-hidden="true" tabindex="-1"></a><span class="co">#  [55] &quot;HAUS1&quot;     &quot;NUP93&quot;     &quot;SNRPE&quot;     &quot;ITGA6&quot;     &quot;CENPN&quot;     &quot;C11orf24&quot; </span></span>
<span id="cb63-59"><a href="#cb63-59" aria-hidden="true" tabindex="-1"></a><span class="co">#  [61] &quot;GGH&quot;       &quot;PFKP&quot;      &quot;FARSA&quot;     &quot;EIF2AK1&quot;   &quot;CENPW&quot;     &quot;TUBA4A&quot;   </span></span>
<span id="cb63-60"><a href="#cb63-60" aria-hidden="true" tabindex="-1"></a><span class="co">#  [67] &quot;TRA2B&quot;     &quot;UMPS&quot;      &quot;MRTO4&quot;     &quot;NUDT15&quot;    &quot;PGM2&quot;      &quot;DBNDD1&quot;   </span></span>
<span id="cb63-61"><a href="#cb63-61" aria-hidden="true" tabindex="-1"></a><span class="co">#  [73] &quot;SNRPB&quot;     &quot;MNAT1&quot;     &quot;NUP35&quot;     &quot;TCEB1&quot;     &quot;HSPB11&quot;    &quot;C19orf48&quot; </span></span>
<span id="cb63-62"><a href="#cb63-62" aria-hidden="true" tabindex="-1"></a><span class="co">#  [79] &quot;ID3&quot;       &quot;IPO4&quot;      &quot;FARSB&quot;     &quot;EIF4G1&quot;    &quot;SKA1&quot;      &quot;MFSD11&quot;   </span></span>
<span id="cb63-63"><a href="#cb63-63" aria-hidden="true" tabindex="-1"></a><span class="co">#  [85] &quot;PLAUR&quot;     &quot;MARVELD2&quot;  &quot;MCM3&quot;      &quot;DHFR&quot;      &quot;RNF41&quot;     &quot;ID2&quot;      </span></span>
<span id="cb63-64"><a href="#cb63-64" aria-hidden="true" tabindex="-1"></a><span class="co">#  [91] &quot;H2AFZ&quot;     &quot;CDK2&quot;      &quot;NCLN&quot;      &quot;ZWILCH&quot;    &quot;DYNLT1&quot;    &quot;C16orf61&quot; </span></span>
<span id="cb63-65"><a href="#cb63-65" aria-hidden="true" tabindex="-1"></a><span class="co">#  [97] &quot;SLC25A40&quot;  &quot;RHOC&quot;      &quot;CCT5&quot;      &quot;PDIA4&quot;     &quot;SNRPA&quot;     &quot;RBM14&quot;    </span></span>
<span id="cb63-66"><a href="#cb63-66" aria-hidden="true" tabindex="-1"></a><span class="co"># [103] &quot;PDLIM7&quot;    &quot;PITPNC1&quot;   &quot;TPM3&quot;      &quot;CORO1C&quot;    &quot;ERLIN1&quot;    &quot;PAICS&quot;    </span></span>
<span id="cb63-67"><a href="#cb63-67" aria-hidden="true" tabindex="-1"></a><span class="co"># [109] &quot;TPRKB&quot;     &quot;SKA2&quot;      &quot;MYBL1&quot;     &quot;SH3BP5L&quot;   &quot;BRCA2&quot;     &quot;SAR1A&quot;    </span></span>
<span id="cb63-68"><a href="#cb63-68" aria-hidden="true" tabindex="-1"></a><span class="co"># [115] &quot;POLR3K&quot;    &quot;MRPS28&quot;    &quot;NUP107&quot;    &quot;TUBG1&quot;     &quot;PNN&quot;       &quot;FAM167A&quot;  </span></span>
<span id="cb63-69"><a href="#cb63-69" aria-hidden="true" tabindex="-1"></a><span class="co"># [121] &quot;RFC3&quot;      &quot;MYL6&quot;      &quot;MCM7&quot;      &quot;MAGOHB&quot;    &quot;FAM89B&quot;    &quot;TOMM40&quot;   </span></span>
<span id="cb63-70"><a href="#cb63-70" aria-hidden="true" tabindex="-1"></a><span class="co"># [127] &quot;CDCA4&quot;     &quot;MT3&quot;       &quot;MTHFD1&quot;    &quot;PSMD12&quot;    &quot;MYBL2&quot;     &quot;CKLF&quot;     </span></span>
<span id="cb63-71"><a href="#cb63-71" aria-hidden="true" tabindex="-1"></a><span class="co"># [133] &quot;NRIP3&quot;     &quot;EZR&quot;       &quot;C12orf24&quot;  &quot;GPLD1&quot;     &quot;SRM&quot;       &quot;RAB3B&quot;    </span></span>
<span id="cb63-72"><a href="#cb63-72" aria-hidden="true" tabindex="-1"></a><span class="co"># [139] &quot;NLN&quot;       &quot;MT1F&quot;      &quot;TNFRSF12A&quot; &quot;TPI1&quot;      &quot;HAS2&quot;      &quot;APOO&quot;     </span></span>
<span id="cb63-73"><a href="#cb63-73" aria-hidden="true" tabindex="-1"></a><span class="co"># [145] &quot;FBXO41&quot;    &quot;MRPL37&quot;    &quot;GSTCD&quot;     &quot;SDC1&quot;      &quot;WDR54&quot;     &quot;RNF138&quot;   </span></span>
<span id="cb63-74"><a href="#cb63-74" aria-hidden="true" tabindex="-1"></a><span class="co"># [151] &quot;APITD1&quot;    &quot;RMND5B&quot;    &quot;ENO1&quot;      &quot;MAP3K8&quot;    &quot;TMEM130&quot;   &quot;SNX17&quot;    </span></span>
<span id="cb63-75"><a href="#cb63-75" aria-hidden="true" tabindex="-1"></a><span class="co"># [157] &quot;KRR1&quot;      &quot;TAGLN&quot;     &quot;PA2G4&quot;     &quot;RUVBL1&quot;    &quot;SNRPD1&quot;    &quot;LOXL2&quot;    </span></span>
<span id="cb63-76"><a href="#cb63-76" aria-hidden="true" tabindex="-1"></a><span class="co"># [163] &quot;POLE2&quot;     &quot;MAPRE1&quot;    &quot;IMP4&quot;      &quot;EMP2&quot;      &quot;PSMD2&quot;     &quot;MET&quot;      </span></span>
<span id="cb63-77"><a href="#cb63-77" aria-hidden="true" tabindex="-1"></a><span class="co"># [169] &quot;IFRD2&quot;     &quot;LMNB2&quot;     &quot;PLOD2&quot;     &quot;NCEH1&quot;     &quot;NME1&quot;      &quot;STRA13&quot;   </span></span>
<span id="cb63-78"><a href="#cb63-78" aria-hidden="true" tabindex="-1"></a><span class="co"># [175] &quot;ACTL6A&quot;    &quot;DLEU1&quot;     &quot;SNRPA1&quot;    &quot;CBX1&quot;      &quot;LYAR&quot;      &quot;PTPLB&quot;    </span></span>
<span id="cb63-79"><a href="#cb63-79" aria-hidden="true" tabindex="-1"></a><span class="co"># [181] &quot;PFN1&quot;      &quot;CENPJ&quot;     &quot;COTL1&quot;     &quot;SPRYD7&quot;    &quot;USPL1&quot;     &quot;MRPL12&quot;   </span></span>
<span id="cb63-80"><a href="#cb63-80" aria-hidden="true" tabindex="-1"></a><span class="co"># [187] &quot;ADAMTS1&quot;   &quot;GLRX3&quot;     &quot;WSB2&quot;      &quot;MRPS16&quot;    &quot;DCLRE1B&quot;   &quot;MKKS&quot;     </span></span>
<span id="cb63-81"><a href="#cb63-81" aria-hidden="true" tabindex="-1"></a><span class="co"># [193] &quot;C3orf26&quot;   &quot;CPEB4&quot;     &quot;SPAG17&quot;    &quot;MLF1IP&quot;    &quot;UAP1&quot;      &quot;COQ2&quot;     </span></span>
<span id="cb63-82"><a href="#cb63-82" aria-hidden="true" tabindex="-1"></a><span class="co"># [199] &quot;WDHD1&quot;     &quot;DCBLD2&quot;    &quot;KIAA0090&quot;  &quot;SAR1B&quot;     &quot;PSMA7&quot;     &quot;PSMC3&quot;    </span></span>
<span id="cb63-83"><a href="#cb63-83" aria-hidden="true" tabindex="-1"></a><span class="co"># [205] &quot;COPS6&quot;     &quot;DUT&quot;       &quot;PPIH&quot;      &quot;PHF19&quot;     &quot;TPM2&quot;      &quot;MCTS1&quot;    </span></span>
<span id="cb63-84"><a href="#cb63-84" aria-hidden="true" tabindex="-1"></a><span class="co"># [211] &quot;EIF4EBP1&quot;  &quot;HNRNPR&quot;</span></span></code></pre></div>
<p>Enjoy <code>GSClassifier</code>!</p>
<!--chapter:end:Quick-start.Rmd-->
</div>
</div>
<div id="model-establishment-via-gsclassifier" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Model establishment via GSClassifier</h1>
<div id="about-1" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> About</h2>
<p>Sometimes, researchers might have their own gene signatures and know how many subtypes they want to call before (based on some knowledges). Gratifyingly, comprehensive functions were also provided in <code>GSClassifier</code>. In this section, we would show how to build a <code>GSClassifier</code> model like <code>PADi</code>.</p>
</div>
<div id="data-preparation" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Data preparation</h2>
<blockquote>
<p>Note: The test data is only for the demonstration of the modeling</p>
</blockquote>
<p>Load packages:</p>
<p>Load data:</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;testData.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_phenotype_part</span></code></pre></div>
<p>Select training and testing cohorts across different platforms and <code>PAD subtypes</code> from <code>PAD</code> function:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>modelInfo <span class="ot">&lt;-</span> <span class="fu">modelData</span>(</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>    design,</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">id.col =</span> <span class="st">&quot;ID&quot;</span>,</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">variable =</span> <span class="fu">c</span>(<span class="st">&quot;platform&quot;</span>, <span class="st">&quot;PAD_subtype&quot;</span>),</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">Prop =</span> <span class="fl">0.7</span>,</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">145</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<p>Check the result <code>modelInfo</code>:</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(modelInfo)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Train)</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Valid)</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Repeat&quot; &quot;Data&quot;  </span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="co">#                    ID  Dataset PAD_subtype PIAM_subtype PIDG_subtype platform</span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM1606509 GSM1606509 GSE65801       PAD-I         high         high GPL14550</span></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM1606517 GSM1606517 GSE65801       PAD-I         high         high GPL14550</span></span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM1606503 GSM1606503 GSE65801       PAD-I         high         high GPL14550</span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM1606525 GSM1606525 GSE65801       PAD-I         high         high GPL14550</span></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM1606511 GSM1606511 GSE65801       PAD-I         high         high GPL14550</span></span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM1606527 GSM1606527 GSE65801       PAD-I         high         high GPL14550</span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a><span class="co">#                    ID  Dataset PAD_subtype PIAM_subtype PIDG_subtype platform</span></span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM2235558 GSM2235558 GSE84437       PAD-I         high         high  GPL6947</span></span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM2235561 GSM2235561 GSE84437      PAD-II         high          low  GPL6947</span></span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM2235562 GSM2235562 GSE84437      PAD-IV          low          low  GPL6947</span></span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM2235563 GSM2235563 GSE84437      PAD-IV          low          low  GPL6947</span></span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM2235564 GSM2235564 GSE84437      PAD-IV          low          low  GPL6947</span></span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a><span class="co"># GSM2235567 GSM2235567 GSE84437      PAD-IV          low          low  GPL6947</span></span></code></pre></div>
<p>Get training data <code>Xs</code> and <code>Ys</code>:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Training data</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  Xs <span class="ot">&lt;-</span> expr[,modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Train<span class="sc">$</span>ID]</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  y <span class="ot">&lt;-</span> modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Train</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  y <span class="ot">&lt;-</span> y[<span class="fu">colnames</span>(Xs),]</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  Ys <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-I&#39;</span>,<span class="dv">1</span>,<span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-II&#39;</span>,<span class="dv">2</span>,<span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-III&#39;</span>,<span class="dv">3</span>,<span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-IV&#39;</span>,<span class="dv">4</span>,<span class="cn">NA</span>)))); <span class="fu">table</span>(Ys)<span class="sc">/</span><span class="fu">length</span>(Ys)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Ys</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="co">#         1         2         3         4 </span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.1010169 0.2474576 0.1694915 0.4820339</span></span></code></pre></div>
<p>Get the number of subtype:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># No. of subtypes</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>nSubtype <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(Ys))</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(nSubtype)</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 4</span></span></code></pre></div>
<p>Also, you can take a look at the validation data:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Validating data</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>Xs_valid <span class="ot">&lt;-</span> expr[,modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Valid<span class="sc">$</span>ID]</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Valid</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> y[<span class="fu">colnames</span>(Xs_valid),]</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>Ys_valid <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-I&#39;</span>,<span class="dv">1</span>,<span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-II&#39;</span>,<span class="dv">2</span>,<span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-III&#39;</span>,<span class="dv">3</span>,<span class="fu">ifelse</span>(y<span class="sc">$</span>PAD_subtype <span class="sc">==</span> <span class="st">&#39;PAD-IV&#39;</span>,<span class="dv">4</span>,<span class="cn">NA</span>))))</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(Ys_valid)<span class="sc">/</span><span class="fu">length</span>(Ys_valid)</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Ys_valid</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a><span class="co">#          1          2          3          4 </span></span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.09609121 0.24592834 0.16612378 0.49185668</span></span></code></pre></div>
<p>Note: When you convert your phenotype into numeric, <strong>You CANNOT USE A ZERO VALUE</strong>, which is not supported by the xGboost.</p>
<p>Other parameteres for modeling:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Build 20 models</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>n<span class="ot">=</span><span class="dv">20</span> </span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a><span class="co"># In every model, 70% samples in the training cohort would be selected. </span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>sampSize<span class="ot">=</span><span class="fl">0.7</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Seed for sampling</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>sampSeed <span class="ot">=</span> <span class="dv">2020</span></span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>na.fill.seed <span class="ot">=</span> <span class="dv">2022</span></span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a><span class="co"># A vector for approximate gene rank estimation</span></span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a>breakVec<span class="ot">=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">1.0</span>)</span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Use 80% most variable gene &amp; gene-pairs for modeling</span></span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a>ptail<span class="ot">=</span><span class="fl">0.8</span><span class="sc">/</span><span class="dv">2</span></span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Automatical selection of parameters for xGboost</span></span>
<span id="cb70-19"><a href="#cb70-19" aria-hidden="true" tabindex="-1"></a>auto <span class="ot">=</span> F</span>
<span id="cb70-20"><a href="#cb70-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-21"><a href="#cb70-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span>auto){</span>
<span id="cb70-22"><a href="#cb70-22" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb70-23"><a href="#cb70-23" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Self-defined params. Fast.</span></span>
<span id="cb70-24"><a href="#cb70-24" aria-hidden="true" tabindex="-1"></a>      params <span class="ot">=</span> <span class="fu">list</span>(<span class="at">max_depth =</span> <span class="dv">10</span>,</span>
<span id="cb70-25"><a href="#cb70-25" aria-hidden="true" tabindex="-1"></a>                    <span class="at">eta =</span> <span class="fl">0.5</span>,</span>
<span id="cb70-26"><a href="#cb70-26" aria-hidden="true" tabindex="-1"></a>                    <span class="at">nrounds =</span> <span class="dv">100</span>,</span>
<span id="cb70-27"><a href="#cb70-27" aria-hidden="true" tabindex="-1"></a>                    <span class="at">nthread =</span> <span class="dv">10</span>,</span>
<span id="cb70-28"><a href="#cb70-28" aria-hidden="true" tabindex="-1"></a>                    <span class="at">nfold=</span><span class="dv">5</span>)</span>
<span id="cb70-29"><a href="#cb70-29" aria-hidden="true" tabindex="-1"></a>      caret.seed <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb70-30"><a href="#cb70-30" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb70-31"><a href="#cb70-31" aria-hidden="true" tabindex="-1"></a>      <span class="co"># No. of CPU for parallel computing. The optimized value depends on your CPU and RAM</span></span>
<span id="cb70-32"><a href="#cb70-32" aria-hidden="true" tabindex="-1"></a>      numCores <span class="ot">=</span> <span class="dv">4</span></span>
<span id="cb70-33"><a href="#cb70-33" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb70-34"><a href="#cb70-34" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb70-35"><a href="#cb70-35" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb70-36"><a href="#cb70-36" aria-hidden="true" tabindex="-1"></a>      <span class="co"># caret::train strategy by GSClassifier:::cvFitOneModel2. Time consuming</span></span>
<span id="cb70-37"><a href="#cb70-37" aria-hidden="true" tabindex="-1"></a>      params <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb70-38"><a href="#cb70-38" aria-hidden="true" tabindex="-1"></a>      caret.seed <span class="ot">=</span> <span class="dv">105</span></span>
<span id="cb70-39"><a href="#cb70-39" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb70-40"><a href="#cb70-40" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Self-defined. For this exmaple training grid, there are 2×1×1×3×2×1×2=24 grids. Make sure that you have a computer with a powerfull CPU.</span></span>
<span id="cb70-41"><a href="#cb70-41" aria-hidden="true" tabindex="-1"></a>      grid <span class="ot">=</span> <span class="fu">expand.grid</span>(</span>
<span id="cb70-42"><a href="#cb70-42" aria-hidden="true" tabindex="-1"></a>        <span class="at">nrounds =</span> <span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">200</span>),</span>
<span id="cb70-43"><a href="#cb70-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">colsample_bytree =</span> <span class="dv">1</span>,</span>
<span id="cb70-44"><a href="#cb70-44" aria-hidden="true" tabindex="-1"></a>        <span class="at">min_child_weight =</span> <span class="dv">1</span>,</span>
<span id="cb70-45"><a href="#cb70-45" aria-hidden="true" tabindex="-1"></a>        <span class="at">eta =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">0.3</span>),</span>
<span id="cb70-46"><a href="#cb70-46" aria-hidden="true" tabindex="-1"></a>        <span class="at">gamma =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.3</span>),</span>
<span id="cb70-47"><a href="#cb70-47" aria-hidden="true" tabindex="-1"></a>        <span class="at">subsample =</span> <span class="fl">0.7</span>,</span>
<span id="cb70-48"><a href="#cb70-48" aria-hidden="true" tabindex="-1"></a>        <span class="at">max_depth =</span> <span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">8</span>)</span>
<span id="cb70-49"><a href="#cb70-49" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb70-50"><a href="#cb70-50" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb70-51"><a href="#cb70-51" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If you don&#39;t know how to set, just use the same number of your subtypes</span></span>
<span id="cb70-52"><a href="#cb70-52" aria-hidden="true" tabindex="-1"></a>      numCores <span class="ot">=</span> <span class="dv">4</span> </span>
<span id="cb70-53"><a href="#cb70-53" aria-hidden="true" tabindex="-1"></a>    }</span></code></pre></div>
<p>Finaly, you have to provide your gene sets as a <code>list</code> object:</p>
<pre><code>geneSet = &lt;Your gene sets&gt;
</code></pre>
<p>Let’s take <code>PAD</code> as an example:</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>PAD <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;PAD.train_20220916.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>geneSet <span class="ot">&lt;-</span> PAD<span class="sc">$</span>geneSet</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(geneSet)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a><span class="co"># $PIAM</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;ENSG00000122122&quot; &quot;ENSG00000117091&quot; &quot;ENSG00000163219&quot; &quot;ENSG00000136167&quot;</span></span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a><span class="co">#  [5] &quot;ENSG00000005844&quot; &quot;ENSG00000123338&quot; &quot;ENSG00000102879&quot; &quot;ENSG00000010671&quot;</span></span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a><span class="co">#  [9] &quot;ENSG00000185862&quot; &quot;ENSG00000104814&quot; &quot;ENSG00000134516&quot; &quot;ENSG00000100055&quot;</span></span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] &quot;ENSG00000082074&quot; &quot;ENSG00000113263&quot; &quot;ENSG00000153283&quot; &quot;ENSG00000198821&quot;</span></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a><span class="co"># [17] &quot;ENSG00000185811&quot; &quot;ENSG00000117090&quot; &quot;ENSG00000171608&quot;</span></span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a><span class="co"># $PIDG</span></span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] &quot;ENSG00000116667&quot; &quot;ENSG00000107771&quot; &quot;ENSG00000196782&quot; &quot;ENSG00000271447&quot;</span></span>
<span id="cb72-13"><a href="#cb72-13" aria-hidden="true" tabindex="-1"></a><span class="co">#  [5] &quot;ENSG00000173517&quot; &quot;ENSG00000134686&quot; &quot;ENSG00000100614&quot; &quot;ENSG00000134247&quot;</span></span>
<span id="cb72-14"><a href="#cb72-14" aria-hidden="true" tabindex="-1"></a><span class="co">#  [9] &quot;ENSG00000109686&quot; &quot;ENSG00000197321&quot; &quot;ENSG00000179981&quot; &quot;ENSG00000187189&quot;</span></span>
<span id="cb72-15"><a href="#cb72-15" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] &quot;ENSG00000140836&quot;</span></span></code></pre></div>
</div>
<div id="fitting-models" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Fitting models</h2>
<div id="gsclassifier-model-training" class="section level3" number="4.3.1">
<h3><span class="header-section-number">4.3.1</span> GSClassifier model training</h3>
<p>Just fit the model like:</p>
<pre><code>
if(!auto){
  
  # Self-defined
  
  system.time(
    res &lt;- fitEnsembleModel(Xs,
                            Ys,
                            geneSet = geneSet,
                            na.fill.method = c(&#39;quantile&#39;,&#39;rpart&#39;,NULL)[1],
                            na.fill.seed = na.fill.seed,
                            n = n,
                            sampSize = sampSize,
                            sampSeed = sampSeed ,
                            breakVec = breakVec,
                            params = params,
                            ptail = ptail,
                            caret.grid = NULL,
                            caret.seed =  caret.seed,
                            verbose = verbose,
                            numCores = numCores)
  )
  
  # user   system  elapsed
  # 0.08s  0.18s   92.70s
  
} else {
  
  # caret::train-defined and time-consuming
  system.time(
    res &lt;- fitEnsembleModel(Xs,
                            Ys,
                            geneSet = geneSet,
                            na.fill.method = c(&#39;quantile&#39;,&#39;rpart&#39;,NULL)[1],
                            na.fill.seed = na.fill.seed,
                            n = n,
                            sampSize = sampSize,
                            sampSeed = sampSeed ,
                            breakVec = breakVec,
                            params = NULL,  # This must be NULL
                            ptail = ptail,
                            caret.grid = grid,
                            caret.seed =  caret.seed,
                            verbose = verbose,
                            numCores = numCores)
  )
  
  # user   system   elapsed
  # 1.10s  2.60s    2311.55s
}

mymusic() # Remind me with a music when the process completed
</code></pre>
<p>You should save it for convenience:</p>
<pre><code>saveRDS(res,&#39;&lt;your path&gt;/train_ens.rds&#39;)</code></pre>
<p>Although a <code>auto-parameter strategy</code> was provided in<code>GSClassifier</code>, it’s unknown for this method to improve your model performance to what extent. You can just try. It’s not a prior recommendation. In generall, setting <code>auto=F</code> in this script could be more cost-effective. Empirically, the speed of <code>caret::train</code> depends on single-core performance of the CPU instead of the core number.</p>
</div>
<div id="scaller-for-the-best-call" class="section level3" number="4.3.2">
<h3><span class="header-section-number">4.3.2</span> Scaller for the best call</h3>
<p>Next, we model the <code>scaller</code> for the training cohort, which would be used for <code>BestCall</code> based on the <code>probability matrix</code> in <code>callEnsemble</code> series. Here, <code>scaller=NULL</code> would cause an <code>NA</code> value of <code>BestCall</code> col. It’s not a big deal, because the <code>probability matrix</code> is the information we need.</p>
<pre><code># Time-consuming modeling
resTrain &lt;- parCallEnsemble(X = Xs,
                            ens = res$Model,
                            geneAnnotation = res$geneAnnotation,
                            geneSet = geneSet,
                            scaller = NULL,
                            geneids = &quot;ensembl&quot;,
                            subtype = NULL,
                            numCores = numCores)

# xgboost via best interation
library(xgboost)
dtrain &lt;- xgb.DMatrix(as.matrix(resTrain[4:(3 + nSubtype)]), label = Ys-1)

cvRes &lt;- xgb.cv(data = dtrain,
                nrounds=100,
                nthread=10,
                nfold=5,
                max_depth=5,
                eta=0.5,
                early_stopping_rounds=100,
                num_class = 4,
                objective = &quot;multi:softmax&quot;)

# xgboost via best interation
bst &lt;- xgboost(data = dtrain,
               max_depth=5,
               eta=0.5,
               nrounds = cvRes$best_iteration,
               nthread=10,
               num_class = 4,
               objective = &quot;multi:softmax&quot;)
Ys_pred &lt;- predict(bst, as.matrix(resTrain[4:7])) + 1
mean(Ys_pred == Ys) # Prediction rates

# Ensemble results
scaller.train &lt;- list(
      Repeat = list(
        data = dtrain,
        max_depth=5,
        eta=0.5,
        nrounds = cvRes$best_iteration,
        nthread=10,
        num_class = 4,
        objective = &quot;multi:softmax&quot;
      ), 
      Model = bst
    )
</code></pre>
</div>
<div id="assemble-your-model" class="section level3" number="4.3.3">
<h3><span class="header-section-number">4.3.3</span> Assemble your model</h3>
<blockquote>
<p>For more information of <code>geneAnnotation</code>, you could see <a href="https://github.com/huangwb8/GSClassifier/wiki/Advanced-development#Gene-Annotation">Advanced development: Gene Annotation</a> section for assistance.</p>
</blockquote>
<p>Here we give an example:</p>
<pre><code>l.train &lt;- list()

# bootstrap models based on the training cohort
l.train[[&#39;ens&#39;]] &lt;- res 

# Scaller model
l.train[[&#39;scaller&#39;]] &lt;- scaller.train 

# a data frame contarining gene annotation for IDs convertion
l.train[[&#39;geneAnnotation&#39;]] &lt;- &lt;Your gene annotation&gt;

# Your gene sets
l.train[[&#39;geneSet&#39;]] &lt;- geneSet
</code></pre>
<p>Finally, save it for downstream analysis</p>
<pre><code>saveRDS(l.train,&#39;&lt;Your path&gt;/train.rds&#39;)
</code></pre>
<p>About model constributions, you can go <code>Advanced development</code> in <a href="https://github.com/huangwb8/GSClassifier/wiki/Advanced-development">here</a> or <a href="http://htmlpreview.github.io/?https://raw.githubusercontent.com/wiki/huangwb8/GSClassifier/Advanced-development.html">here</a> for more information.</p>
</div>
<div id="of-note-2" class="section level3" number="4.3.4">
<h3><span class="header-section-number">4.3.4</span> Of note</h3>
<p>You can take a look at the <code>PAD.train_20220916</code> model (<code>PADi</code>). You have to make your model frame similar to the <code>PAD.train_20220916</code> model.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>l.train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;PAD.train_20220916.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(l.train)</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;ens&quot;            &quot;scaller&quot;        &quot;geneAnnotation&quot; &quot;geneSet&quot;</span></span></code></pre></div>
<p>The time of <code>GSClassifier</code> modeling depends on the number of individual models (controlled by <code>n</code>)/called subtypes/gene signatures, automatic parameter selection, and your CPU capacity.</p>
</div>
</div>
<div id="calling-subtypes" class="section level2" number="4.4">
<h2><span class="header-section-number">4.4</span> Calling subtypes</h2>
<p>Supposed that you had got a <code>GSClassifier</code> model, next you want to use it for personalized subtype calling.</p>
<p>Just:</p>
<pre><code># Load your model
l &lt;- readRDS(&#39;&lt;Your path&gt;/train.rds&#39;)

# subtype calling
res_i = callEnsemble(
  X,
  ens = l$ens$Model,
  geneAnnotation = l$geneAnnotation,
  geneSet = l$geneSet,
  scaller = l$scaller$Model,
  geneid = &lt;ID type of your training data&gt;,
  subtype = NULL,
  verbose = T
)
</code></pre>
<p>The usage of <code>parCallEnsemble</code> (for huge amount of samples) is similar.</p>
<!--chapter:end:Model-establishment.Rmd-->
</div>
</div>
<div id="suggestions-for-gsclassifier-model-developers" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Suggestions for GSClassifier model developers</h1>
<div id="about-2" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> About</h2>
<ul>
<li><p>The book <code>R packages</code> is a straightaway and useful reference book for R developers. The free-access website of <code>R packages</code> is <a href="https://r-pkgs.org/">https://r-pkgs.org/</a>. As a developer of R, if you haven’t hear about it, it’s strongly recommanded to just read it. Hadley Wickham, the main author of the book, is an active R developer and have led some master works like <code>ggplot2</code> and <code>plyr</code>.</p></li>
<li><p>With <code>GSClassifier</code> package, it could be easy for users to build a model only with certain gene sets and transcriptomics data. If you are interesting in sharing your model, <code>GSClassifier</code> also provides a simple methodology for this vision. In this section, let’s see how to achieve it!</p></li>
</ul>
<p>First, load the package</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GSClassifier)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a><span class="co"># 载入需要的程辑包：luckyBase</span></span></code></pre></div>
</div>
<div id="available-models" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Available models</h2>
<p>With <code>GSClassifier_Data()</code>, all models supported in the current <code>GSClassifier</code> package would showed.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">GSClassifier_Data</span>()</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Available data:</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage example:</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype.rds </span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20200110.rds </span></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20220916.rds </span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;PAD.train_20200110.rds&quot;, package = &quot;GSClassifier&quot;)) </span></span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;ImmuneSubtype.rds&quot;, package = &quot;GSClassifier&quot;))</span></span></code></pre></div>
<p>For more details of <code>GSClassifier_Data()</code>, just:</p>
<pre><code>?GSClassifier_Data()</code></pre>
<p>Set <code>model=F</code>, all <code>.rds</code> data would be showed:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">GSClassifier_Data</span>(<span class="at">model =</span> F)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Available data:</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage example:</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   general-gene-annotation.rds </span></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype.rds </span></span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20200110.rds </span></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20220916.rds </span></span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a><span class="co">#   testData.rds </span></span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;PAD.train_20200110.rds&quot;, package = &quot;GSClassifier&quot;)) </span></span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;ImmuneSubtype.rds&quot;, package = &quot;GSClassifier&quot;))</span></span></code></pre></div>
</div>
<div id="components-of-a-gsclassifier-model" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Components of a GSClassifier model</h2>
<p>Currently, a GSClassifier model and related product environments is designed as a <code>list</code> object. Let’s take <code>PAD.train_20210110</code>(also called <code>PADi</code>) as an example.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>PADi <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;PAD.train_20220916.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)) </span></code></pre></div>
<p>This picture shows the components of <code>PADi</code>:</p>
<div class="figure" style="text-align: center">
<img src="fig/GSClassifier-model.jpg" alt="Details of a GSClassifier model" width="90%" />
<p class="caption">
(#fig:GSClassifierModel)Details of a GSClassifier model
</p>
</div>
<p>As shown, a typical <code>GSClassifier</code> model is consist of four parts (with different colors in the picture):</p>
<ul>
<li><code>1. ens</code>:
<ul>
<li><code>Repeat</code>: productive parameters of <code>GSClassifier</code> models</li>
<li><code>Model</code>: <code>GSClassifier</code> models. Here, <code>PADi</code> had 20 models from different subs of the training cohorts</li>
</ul></li>
<li><code>2. scaller</code>:
<ul>
<li><code>Repeat</code>: productive parameters of the <code>scaller</code> model, which was used for <code>BestCall</code> calling</li>
<li><code>Model</code>: the <code>scaller</code> model</li>
</ul></li>
<li><code>3. geneAnnotation</code>: a data frame containing gene annotation information</li>
<li><code>4. geneSet</code>: a list contains several gene sets</li>
</ul>
<p>Thus, you can assemble your model like:</p>
<pre><code>model &lt;- list()

# bootstrap models based on the training cohort
model[[&#39;ens&#39;]] &lt;- &lt;Your model for subtypes calling&gt;

# Scaller model
model[[&#39;scaller&#39;]] &lt;- &lt;Your scaller for BestCall calling&gt;

# a data frame contarining gene annotation for IDs convertion
model[[&#39;geneAnnotation&#39;]] &lt;- &lt;Your gene annotation&gt;

# Your gene sets
model[[&#39;geneSet&#39;]] &lt;- &lt;Your gene sets&gt;

saveRDS(model, &#39;your-model.rds&#39;)
</code></pre>
<p>More tutorials for model establishment, please go to <a href="https://github.com/huangwb8/GSClassifier/wiki/Model-establishment">markdown tutorial</a> or <a href="http://htmlpreview.github.io/?https://raw.githubusercontent.com/wiki/huangwb8/GSClassifier/Model-establishment.html">html tutorial</a>.</p>
</div>
<div id="submit-models-to-luckymodel-package" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Submit models to luckyModel package</h2>
<p>Considering most users of <code>GSClassifier</code> might have no need for lots of models, We divided the model storage feature into an new ensembl package called <a href="https://github.com/huangwb8/luckyModel"><strong>luckyModel</strong></a>. Don’t worry, the usage is very easy!</p>
<p>If you want to summit your model, you should apply for a contributor of <code>luckyModel</code> first. Then, just send the model (<code>.rds</code>) into the <code>inst/extdata/&lt;project&gt;</code> path of <code>luckyModel</code>. After audit, your branch would be accepted and available for the users.</p>
<p>The name of your model must be the format as following:</p>
<pre><code># &lt;project&gt;
GSClassifier

# &lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;:
HWB_PAD_v20211201.rds</code></pre>
</div>
<div id="repeatablility-of-models" class="section level2" number="5.5">
<h2><span class="header-section-number">5.5</span> Repeatablility of models</h2>
<p>For repeatablility, you had better submit a <code>.zip</code> or <code>.tar.gz</code> file that containing the information of your model. Here are some suggestions:</p>
<ul>
<li><p><code>&lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;.md</code></p>
<ul>
<li><p><strong>Destinations</strong>: Why you develop the model</p></li>
<li><p><strong>Design</strong>: The evidence for gene sigatures, et al</p></li>
<li><p><strong>Data sources</strong>: The data for model training and validating, et al</p></li>
<li><p><strong>Applications</strong>: Where to use your model</p></li>
<li><p><strong>Limintations</strong>: Limitation or improvement direction of your model</p></li>
</ul></li>
<li><p><code>&lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;.R</code>: The code you used for model training and validating.</p></li>
<li><p><code>Data-of-&lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;.rds</code> (Optional): Due to huge size of omics data, it’s OK for you not to submit the raw data.</p></li>
</ul>
<p>:cupid: Welcome your contributions!</p>
</div>
<div id="gene-annotation" class="section level2" number="5.6">
<h2><span class="header-section-number">5.6</span> Gene Annotation</h2>
<p>For convenience, we provided a general gene annotation dataset for different genomics:</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>gga <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;general-gene-annotation.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gga)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;hg38&quot; &quot;hg19&quot; &quot;mm10&quot;</span></span></code></pre></div>
<p>I believe they’re enough for routine medicine studies.</p>
<p>Here, take a look at <code>hg38</code>:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>hg38 <span class="ot">&lt;-</span> gga<span class="sc">$</span>hg38</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hg38)</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="co">#           ENSEMBL       SYMBOL  ENTREZID</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 ENSG00000223972      DDX11L1 100287102</span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 ENSG00000227232       WASH7P      &lt;NA&gt;</span></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 4 ENSG00000278267    MIR6859-1 102466751</span></span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 5 ENSG00000243485 RP11-34P13.3      &lt;NA&gt;</span></span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 6 ENSG00000284332    MIR1302-2 100302278</span></span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 7 ENSG00000237613      FAM138A    645520</span></span></code></pre></div>
<p>With this kind of data, it’s simple to customize your own gene annotation (take <code>PADi</code> as examples):</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>tGene <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unlist</span>(PADi<span class="sc">$</span>geneSet))</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>geneAnnotation <span class="ot">&lt;-</span> hg38[hg38<span class="sc">$</span>ENSEMBL <span class="sc">%in%</span> tGene, ]</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(geneAnnotation)</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 32  3</span></span></code></pre></div>
<p>Have a check:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geneAnnotation)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="co">#              ENSEMBL SYMBOL ENTREZID</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 353  ENSG00000171608 PIK3CD     5293</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1169 ENSG00000134686   PHC2     1912</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 2892 ENSG00000134247 PTGFRN     5738</span></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 3855 ENSG00000117090 SLAMF1     6504</span></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 3858 ENSG00000117091   CD48      962</span></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 4043 ENSG00000198821  CD247      919</span></span></code></pre></div>
<p>This <code>geneAnnotation</code> could be the <code>model[['geneAnnotation']]</code>.</p>
<p>Also, we use a function called <code>convert</code> to do gene ID convertion.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>luckyBase<span class="sc">::</span><span class="fu">convert</span>(<span class="fu">c</span>(<span class="st">&#39;GAPDH&#39;</span>,<span class="st">&#39;TP53&#39;</span>), <span class="st">&#39;SYMBOL&#39;</span>, <span class="st">&#39;ENSEMBL&#39;</span>, hg38)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;ENSG00000111640&quot; &quot;ENSG00000141510&quot;</span></span></code></pre></div>
<p>Note: the <code>luckyBase</code> package integrates lots of useful tiny functions, you could explore it sometimes.</p>
<!--chapter:end:Advanced-development.Rmd-->
<!--  -->
<!-- # References {-} -->
<!--  -->
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<!--chapter:end:References.Rmd-->
<div id="refs" class="references csl-bib-body">
<div id="ref-RN160" class="csl-entry">
1. Thorsson V, Gibbs DL, Brown SD, et al. <a href="https://doi.org/10.1016/j.immuni.2018.03.023">The immune landscape of cancer</a>. Immunity 2018; 48:812–830 e14
</div>
<div id="ref-RN315" class="csl-entry">
2. Gibbs DL. Robust classification of immune subtypes in cancer. 2020;
</div>
<div id="ref-RN345" class="csl-entry">
3. Chen T, He T, Benesty M, et al. Xgboost: Extreme gradient boosting. 2015; 1:1–4
</div>
<div id="ref-RN267" class="csl-entry">
4. Geman D, d’Avignon C, Naiman DQ, et al. <a href="https://doi.org/10.2202/1544-6115.1071">Classifying gene expression profiles from pairwise mRNA comparisons</a>. Stat Appl Genet Mol Biol 2004; 3:Article19
</div>
<div id="ref-RN265" class="csl-entry">
5. Zhao H, Logothetis CJ, Gorlov IP. <a href="https://doi.org/10.1038/pcan.2010.9">Usefulness of the top-scoring pairs of genes for prediction of prostate cancer progression</a>. Prostate Cancer Prostatic Dis 2010; 13:252–9
</div>
<div id="ref-RN266" class="csl-entry">
6. Youssef YM, White NM, Grigull J, et al. <a href="https://doi.org/10.1016/j.eururo.2011.01.004">Accurate molecular classification of kidney cancer subtypes using microRNA signature</a>. Eur Urol 2011; 59:721–30
</div>
<div id="ref-RN261" class="csl-entry">
7. Auslander N, Zhang G, Lee JS, et al. <a href="https://doi.org/10.1038/s41591-018-0157-9">Robust prediction of response to immune checkpoint blockade therapy in metastatic melanoma</a>. Nat Med 2018; 24:1545–1549
</div>
<div id="ref-RN387" class="csl-entry">
8. Troyanskaya O, Cantor M, Sherlock G, et al. <a href="https://doi.org/10.1093/bioinformatics/17.6.520">Missing value estimation methods for DNA microarrays</a>. Bioinformatics 2001; 17:520–5
</div>
<div id="ref-RN389" class="csl-entry">
9. Arbeitman MN, Furlong EE, Imam F, et al. <a href="https://doi.org/10.1126/science.1072152">Gene expression during the life cycle of drosophila melanogaster</a>. Science 2002; 297:2270–5
</div>
<div id="ref-RN382" class="csl-entry">
10. Zhu X, Wang J, Sun B, et al. <a href="https://doi.org/10.1186/s12859-021-04109-4">An efficient ensemble method for missing value imputation in microarray gene expression data</a>. BMC Bioinformatics 2021; 22:188
</div>
<div id="ref-RN392" class="csl-entry">
11. Lin W-C, Tsai C-F. Missing value imputation: A review and analysis of the literature (2006–2017). 2020; 53:1487–1509
</div>
<div id="ref-RN386" class="csl-entry">
12. Hasan MK, Alam MA, Roy S, et al. Missing value imputation affects the performance of machine learning: A review and analysis of the literature (2010–2021). Informatics in Medicine Unlocked 2021; 27:100799
</div>
<div id="ref-RN384" class="csl-entry">
13. Wang A, Yang J, An N. Regularized sparse modelling for microarray missing value estimation. 2021; 9:16899–16913
</div>
<div id="ref-xgboost" class="csl-entry">
14. Chen T, He T, Benesty M, et al. <a href="https://CRAN.R-project.org/package=xgboost">Xgboost: Extreme gradient boosting</a>. 2022;
</div>
<div id="ref-pROC" class="csl-entry">
15. Robin X, Turck N, Hainard A, et al. pROC: An open-source package for r and s+ to analyze and compare ROC curves. BMC Bioinformatics 2011; 12:77
</div>
<div id="ref-RN369" class="csl-entry">
16. Zhu S, Kong W, Zhu J, et al. <a href="https://doi.org/10.1093/bib/bbac344">The genetic algorithm-aided three-stage ensemble learning method identified a robust survival risk score in patients with glioma</a>. Brief Bioinform 2022;
</div>
<div id="ref-RN367" class="csl-entry">
17. Tong M, Zheng W, Li H, et al. <a href="https://doi.org/10.1038/oncsis.2016.51">Multi-omics landscapes of colorectal cancer subtypes discriminated by an individualized prognostic signature for 5-fluorouracil-based chemotherapy</a>. Oncogenesis 2016; 5:e242
</div>
<div id="ref-RN368" class="csl-entry">
18. Qi L, Li Y, Qin Y, et al. <a href="https://doi.org/10.1038/bjc.2016.370">An individualised signature for predicting response with concordant survival benefit for lung adenocarcinoma patients receiving platinum-based chemotherapy</a>. Br J Cancer 2016; 115:1513–1519
</div>
<div id="ref-RN364" class="csl-entry">
19. Huang H, Zou Y, Zhang H, et al. <a href="https://doi.org/10.1016/j.trsl.2020.02.004">A qualitative transcriptional prognostic signature for patients with stage i-II pancreatic ductal adenocarcinoma</a>. Transl Res 2020; 219:30–44
</div>
<div id="ref-RN363" class="csl-entry">
20. Zheng H, Song K, Fu Y, et al. <a href="https://doi.org/10.1093/bib/bbz174">An absolute human stemness index associated with oncogenic dedifferentiation</a>. Brief Bioinform 2021; 22:2151–2160
</div>
<div id="ref-RN362" class="csl-entry">
21. Kong W, He L, Zhu J, et al. <a href="https://doi.org/10.1038/s41375-022-01662-6">An immunity and pyroptosis gene-pair signature predicts overall survival in acute myeloid leukemia</a>. Leukemia 2022;
</div>
<div id="ref-RN366" class="csl-entry">
22. Liu K, Geng Y, Wang L, et al. <a href="https://doi.org/10.1002/1878-0261.13279">Systematic exploration of the underlying mechanism of gemcitabine resistance in pancreatic adenocarcinoma</a>. Mol Oncol 2022; 16:3034–3051
</div>
<div id="ref-RN365" class="csl-entry">
23. Zheng H, Xie J, Song K, et al. <a href="https://doi.org/10.1186/s13287-022-02803-5">StemSC: A cross-dataset human stemness index for single-cell samples</a>. Stem Cell Res Ther 2022; 13:115
</div>
</div>
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
    </div>
  </div>
<!--bookdown:config-->

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
