--- 
title: "The Principle of R Package GSClassifier"
author: "Weibin Huang"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
csl: briefings-in-bioinformatics.csl
link-citations: yes
geometry:
  - top=1in
  - left=1in
  - right=1in
  - bottom=1in
fontsize: 12pt
linestretch: 1.5
linkcolor: "blue"
description: "The Principle of GSClassifier"
---

# Welcome {-}

<!--

[GSClassifier](https://github.com/huangwb8/GSClassifier) is an R package for modeling and identification of Gene Expression Profiles (GEPs) subtypes. The detail of **GSClassifier** package usage had been demonstrated in [Github WiKi](https://github.com/huangwb8/GSClassifier/wiki). Here, we propose to introduce the principle of GSClassifier, including flowchart, **top scoring pairs (TSP)** algorithm, and batch effect control. 

emoji: https://github.com/rstudio/blogdown/issues/171

-->


## `r if(knitr::is_html_output()) emo::ji("sweat_drops")`About {-}

**The Priciple of GSClassifier** is a book for users of the R package **GSClassifier** who want to know the most details. If you're looking for the PDF edition, you can find it  [here](https://github.com/huangwb8/GSClassifier.principle/blob/master/docs/GSClassifier.principle.pdf).

[**GSClassifier**](https://github.com/huangwb8/GSClassifier) is an R-based comprehensive classification tool for subtypes modeling and personalized calling based on pure transcriptomics. It could be used for precision medicine, such as cancer diagnosis. The inspiration for **GSClassifier** comes from [ImmuneSubtypeClassifier](https://github.com/CRI-iAtlas/ImmuneSubtypeClassifier), an R package for classification of PanCancer immune subtypes based on the work of Gibbs et al [@RN160; @RN315].

Lots of surprising features in **GSClassifier** are as follows: 

+ Optimized for just `one sample`

+ Available for modeling and calling of brand-new `GEPs-based subtypes` in any diseases (cancers)

+ No limitation of the number of `gene signatures`(≥1) or `subtypes`(≥2)

+ `Insensitive normalization` due to the use of  the individual `gene rank matrix`

+ More ensemble and repeatable modeling process

+ More optimizations in the parallel computing

+ New useful functions as supplements

**ATTENTION!** In the future, there might be third-party contributors in `GSClassifier` platform, with some useful models for specific usages. If you use models provided by these people, **you had better know more details as possible**, including **designs, data sources, destinations, training scripts, and limitations** of models, especially those from studies under peer review.

## `r if(knitr::is_html_output()) emo::ji("cd")`License {-}

+ **GSClassifier** is released under the Apache-2.0 license. See [LICENSE](https://github.com/huangwb8/GSClassifier/blob/master/license.txt) for details.

+ The technical documentation, as a whole, is licensed under a [Creative Commons Attribution-
NonCommercial-ShareAlike 4.0 International License](http://creativecommons.org/licenses/by-nc-sa/4.0/). The code contained in this book is simultaneously available under the [MIT license](https://opensource.org/licenses/MIT); this means that you are free to use it in your packages, as long as you cite the source.

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>

## `r if(knitr::is_html_output()) emo::ji("+1")`Installation {-}

**RStudio/Posit** is one of the best Integrated Development Environments (IDE) in R programming. If you're struggling in R-GUI, it is recommended to turn to [RStudio/Posit](https://www.rstudio.com/).

For installation of **GSClassifier**, please run these commands in an R environment: 

```R
# Install "devtools" package
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")

# Install dependencies
if (!requireNamespace("luckyBase", quietly = TRUE))
    devtools::install_github("huangwb8/luckyBase")

# Install the "GSClassifier" package
if (!requireNamespace("GSClassifier", quietly = TRUE))
    devtools::install_github("huangwb8/GSClassifier")
```

In the future, a stable `GSClassifier` version might be sent to [`CRAN`](https://cran.r-project.org/). Still beta.


## `r if(knitr::is_html_output()) emo::ji("eyes")`[Mirror](https://gitee.com/huangwb8/GSClassifier) {-}

For some special countries or regions, users could also try:

```R
# Install dependencies
install.packages("https://gitee.com/huangwb8/luckyBase/repository/archive/Primary?format=tar.gz", repos=NULL, method="libcurl")

# Install the "GSClassifier" package
install.packages("https://gitee.com/huangwb8/GSClassifier/repository/archive/Primary?format=tar.gz", repos=NULL, method="libcurl")
```

## `r if(knitr::is_html_output()) emo::ji("books")`Change log {-}

+ Version 0.1.9

  + Optimize function verbose
  
  + Optimize for a routine scenario: one gene set and two subtypes
  
  + Optimize the strategy of automatic parameters selection for modeling training with R package `caret`
  
  + Interact with external models from the [luckyModel](https://github.com/huangwb8/luckyModel) package

+ Version 0.1.8

  + Primary public version of `GSClassifier`
  
  + Apache License, Version 2.0
  
  + Friendly wiki-based tutorial
  
  + Platform for developers

## `r if(knitr::is_html_output()) emo::ji("calendar")`TODO {-}

+ More medical fields included, such as in the pan-cancer utility

+ Advanced methods (such as artificial intelligence) for enhanced robustness

+ Unsupervised learning for de-novo classification based on intrinsic frames of omics instead of human knowledge

+ Multi-omics exploration and support

+ More friendly characteristics for developers and contributors

+ Web application for newbies to R programming

## `r if(knitr::is_html_output()) emo::ji("palm_tree")`Other Projects {-}

You may also be interested in:

* __"[luckyBase](https://github.com/huangwb8/luckyBase)"__ The base functions of lucky series.

* __"[luckyModel](https://github.com/huangwb8/luckyModel)"__ Model ensemble for third-party lucky series, such GSClassifier.
