<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Discussion | The Principle of R Package GSClassifier</title>
  <meta name="description" content="The Principle of GSClassifier" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Discussion | The Principle of R Package GSClassifier" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The Principle of GSClassifier" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Discussion | The Principle of R Package GSClassifier" />
  
  <meta name="twitter:description" content="The Principle of GSClassifier" />
  

<meta name="author" content="Weibin Huang" />


<meta name="date" content="2025-05-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="the-principle-of-gsclassifier.html"/>
<link rel="next" href="quick-start.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">GSClassifier</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about"><i class="fa fa-check"></i>üí¶About</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>üíøLicense</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i>üëçInstallation</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#mirror"><i class="fa fa-check"></i>üëÄMirror</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#change-log"><i class="fa fa-check"></i>üìöChange log</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#todo"><i class="fa fa-check"></i>üìÜTODO</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#other-projects"><i class="fa fa-check"></i>üå¥Other Projects</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html"><i class="fa fa-check"></i><b>1</b> The Principle of GSClassifier</a>
<ul>
<li class="chapter" data-level="1.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#packages"><i class="fa fa-check"></i><b>1.1</b> Packages</a></li>
<li class="chapter" data-level="1.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#flowchart"><i class="fa fa-check"></i><b>1.2</b> Flowchart</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#data-processing"><i class="fa fa-check"></i><b>1.2.1</b> Data Processing</a></li>
<li class="chapter" data-level="1.2.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#model-establishment-and-validation"><i class="fa fa-check"></i><b>1.2.2</b> Model Establishment and Validation</a></li>
<li class="chapter" data-level="1.2.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#model-application"><i class="fa fa-check"></i><b>1.2.3</b> Model Application</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#topicTSP"><i class="fa fa-check"></i><b>1.3</b> Top scoring pairs</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#tsp-intro"><i class="fa fa-check"></i><b>1.3.1</b> Introduction</a></li>
<li class="chapter" data-level="1.3.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#simulated-dataset"><i class="fa fa-check"></i><b>1.3.2</b> Simulated Dataset</a></li>
<li class="chapter" data-level="1.3.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#binned-expression"><i class="fa fa-check"></i><b>1.3.3</b> Binned expression</a></li>
<li class="chapter" data-level="1.3.4" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#pair-difference"><i class="fa fa-check"></i><b>1.3.4</b> Pair difference</a></li>
<li class="chapter" data-level="1.3.5" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#set-difference"><i class="fa fa-check"></i><b>1.3.5</b> Set difference</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="discussion.html"><a href="discussion.html"><i class="fa fa-check"></i><b>2</b> Discussion</a>
<ul>
<li class="chapter" data-level="2.1" data-path="discussion.html"><a href="discussion.html#packages-1"><i class="fa fa-check"></i><b>2.1</b> Packages</a></li>
<li class="chapter" data-level="2.2" data-path="discussion.html"><a href="discussion.html#subtype-vector"><i class="fa fa-check"></i><b>2.2</b> Subtype Vector</a></li>
<li class="chapter" data-level="2.3" data-path="discussion.html"><a href="discussion.html#missing-value-imputation-mvi"><i class="fa fa-check"></i><b>2.3</b> Missing value imputation (MVI)</a></li>
<li class="chapter" data-level="2.4" data-path="discussion.html"><a href="discussion.html#batch-effect"><i class="fa fa-check"></i><b>2.4</b> Batch effect</a></li>
<li class="chapter" data-level="2.5" data-path="discussion.html"><a href="discussion.html#subtype-number"><i class="fa fa-check"></i><b>2.5</b> Subtype number</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>3</b> Quick start</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quick-start.html"><a href="quick-start.html#about-1"><i class="fa fa-check"></i><b>3.1</b> About</a></li>
<li class="chapter" data-level="3.2" data-path="quick-start.html"><a href="quick-start.html#package"><i class="fa fa-check"></i><b>3.2</b> Package</a></li>
<li class="chapter" data-level="3.3" data-path="quick-start.html"><a href="quick-start.html#data"><i class="fa fa-check"></i><b>3.3</b> Data</a></li>
<li class="chapter" data-level="3.4" data-path="quick-start.html"><a href="quick-start.html#pad"><i class="fa fa-check"></i><b>3.4</b> PAD</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="quick-start.html"><a href="quick-start.html#the-work-flow-of-pad-exploration"><i class="fa fa-check"></i><b>3.4.1</b> The work flow of PAD exploration</a></li>
<li class="chapter" data-level="3.4.2" data-path="quick-start.html"><a href="quick-start.html#preparation-of-the-test-data"><i class="fa fa-check"></i><b>3.4.2</b> Preparation of the test data</a></li>
<li class="chapter" data-level="3.4.3" data-path="quick-start.html"><a href="quick-start.html#unsupervised-clustering"><i class="fa fa-check"></i><b>3.4.3</b> Unsupervised clustering</a></li>
<li class="chapter" data-level="3.4.4" data-path="quick-start.html"><a href="quick-start.html#of-note"><i class="fa fa-check"></i><b>3.4.4</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="quick-start.html"><a href="quick-start.html#padi"><i class="fa fa-check"></i><b>3.5</b> PADi</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="quick-start.html"><a href="quick-start.html#preparation-of-the-test-data-1"><i class="fa fa-check"></i><b>3.5.1</b> Preparation of the test data</a></li>
<li class="chapter" data-level="3.5.2" data-path="quick-start.html"><a href="quick-start.html#use-a-specific-function-called-padi"><i class="fa fa-check"></i><b>3.5.2</b> Use a specific function called <code>PADi</code></a></li>
<li class="chapter" data-level="3.5.3" data-path="quick-start.html"><a href="quick-start.html#use-the-callensemble-function"><i class="fa fa-check"></i><b>3.5.3</b> Use the <code>callEnsemble</code> function</a></li>
<li class="chapter" data-level="3.5.4" data-path="quick-start.html"><a href="quick-start.html#parallel-strategy-for-padi"><i class="fa fa-check"></i><b>3.5.4</b> Parallel strategy for PADi</a></li>
<li class="chapter" data-level="3.5.5" data-path="quick-start.html"><a href="quick-start.html#single-sample-subtype-calling"><i class="fa fa-check"></i><b>3.5.5</b> Single sample subtype calling</a></li>
<li class="chapter" data-level="3.5.6" data-path="quick-start.html"><a href="quick-start.html#of-note-1"><i class="fa fa-check"></i><b>3.5.6</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="quick-start.html"><a href="quick-start.html#use-external-models-from-luckymodel-package"><i class="fa fa-check"></i><b>3.6</b> Use external models from luckyModel package</a></li>
<li class="chapter" data-level="3.7" data-path="quick-start.html"><a href="quick-start.html#pancancer-immune-subtypes"><i class="fa fa-check"></i><b>3.7</b> PanCancer immune subtypes</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html"><i class="fa fa-check"></i><b>4</b> Model establishment via GSClassifier</a>
<ul>
<li class="chapter" data-level="4.1" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#about-2"><i class="fa fa-check"></i><b>4.1</b> About</a></li>
<li class="chapter" data-level="4.2" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#data-preparation"><i class="fa fa-check"></i><b>4.2</b> Data preparation</a></li>
<li class="chapter" data-level="4.3" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#fitting-models"><i class="fa fa-check"></i><b>4.3</b> Fitting models</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#gsclassifier-model-training"><i class="fa fa-check"></i><b>4.3.1</b> GSClassifier model training</a></li>
<li class="chapter" data-level="4.3.2" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#scaller-for-the-best-call"><i class="fa fa-check"></i><b>4.3.2</b> Scaller for the best call</a></li>
<li class="chapter" data-level="4.3.3" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#assemble-your-model"><i class="fa fa-check"></i><b>4.3.3</b> Assemble your model</a></li>
<li class="chapter" data-level="4.3.4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#of-note-2"><i class="fa fa-check"></i><b>4.3.4</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#calling-subtypes"><i class="fa fa-check"></i><b>4.4</b> Calling subtypes</a></li>
<li class="chapter" data-level="4.5" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#topicSubmodel"><i class="fa fa-check"></i><b>4.5</b> Number of SubModel</a></li>
<li class="chapter" data-level="4.6" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#padiPara"><i class="fa fa-check"></i><b>4.6</b> Parameters for PADi training</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html"><i class="fa fa-check"></i><b>5</b> Suggestions for GSClassifier model developers</a>
<ul>
<li class="chapter" data-level="5.1" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#about-3"><i class="fa fa-check"></i><b>5.1</b> About</a></li>
<li class="chapter" data-level="5.2" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#available-models"><i class="fa fa-check"></i><b>5.2</b> Available models</a></li>
<li class="chapter" data-level="5.3" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#components-of-a-gsclassifier-model"><i class="fa fa-check"></i><b>5.3</b> Components of a GSClassifier model</a></li>
<li class="chapter" data-level="5.4" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#submit-models-to-luckymodel-package"><i class="fa fa-check"></i><b>5.4</b> Submit models to luckyModel package</a></li>
<li class="chapter" data-level="5.5" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#repeatablility-of-models"><i class="fa fa-check"></i><b>5.5</b> Repeatablility of models</a></li>
<li class="chapter" data-level="5.6" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#gene-annotation"><i class="fa fa-check"></i><b>5.6</b> Gene Annotation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://blognas.hwb0307.com/" target="blank">Welcome to My Blog!</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Principle of R Package GSClassifier</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="discussion" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Discussion<a href="discussion.html#discussion" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this section, we would discuss some key topics about <strong>GSClassifier</strong>, including <strong>Missing value imputation (MVI)</strong>, <strong>Batch effect</strong>, <strong>hyperparameters</strong>, and so on.</p>
<div id="packages-1" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Packages<a href="discussion.html#packages-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="discussion.html#cb19-1" tabindex="-1"></a><span class="co"># Install &quot;devtools&quot; package</span></span>
<span id="cb19-2"><a href="discussion.html#cb19-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb19-3"><a href="discussion.html#cb19-3" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb19-4"><a href="discussion.html#cb19-4" tabindex="-1"></a></span>
<span id="cb19-5"><a href="discussion.html#cb19-5" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb19-6"><a href="discussion.html#cb19-6" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;luckyBase&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb19-7"><a href="discussion.html#cb19-7" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/luckyBase&quot;</span>)</span>
<span id="cb19-8"><a href="discussion.html#cb19-8" tabindex="-1"></a></span>
<span id="cb19-9"><a href="discussion.html#cb19-9" tabindex="-1"></a><span class="co"># Install the &quot;**GSClassifier**&quot; package</span></span>
<span id="cb19-10"><a href="discussion.html#cb19-10" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;GSClassifier&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb19-11"><a href="discussion.html#cb19-11" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/GSClassifier&quot;</span>)</span>
<span id="cb19-12"><a href="discussion.html#cb19-12" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb19-13"><a href="discussion.html#cb19-13" tabindex="-1"></a></span>
<span id="cb19-14"><a href="discussion.html#cb19-14" tabindex="-1"></a><span class="co"># Install the &quot;pacman&quot; package</span></span>
<span id="cb19-15"><a href="discussion.html#cb19-15" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;pacman&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb19-16"><a href="discussion.html#cb19-16" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;pacman&quot;</span>)</span>
<span id="cb19-17"><a href="discussion.html#cb19-17" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb19-18"><a href="discussion.html#cb19-18" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb19-19"><a href="discussion.html#cb19-19" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb19-20"><a href="discussion.html#cb19-20" tabindex="-1"></a>}</span>
<span id="cb19-21"><a href="discussion.html#cb19-21" tabindex="-1"></a></span>
<span id="cb19-22"><a href="discussion.html#cb19-22" tabindex="-1"></a><span class="co"># Load needed packages</span></span>
<span id="cb19-23"><a href="discussion.html#cb19-23" tabindex="-1"></a>packages_needed <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb19-24"><a href="discussion.html#cb19-24" tabindex="-1"></a>  <span class="st">&quot;readxl&quot;</span>,</span>
<span id="cb19-25"><a href="discussion.html#cb19-25" tabindex="-1"></a>  <span class="st">&quot;ComplexHeatmap&quot;</span>,</span>
<span id="cb19-26"><a href="discussion.html#cb19-26" tabindex="-1"></a>  <span class="st">&quot;GSClassifier&quot;</span>,</span>
<span id="cb19-27"><a href="discussion.html#cb19-27" tabindex="-1"></a>  <span class="st">&quot;rpart&quot;</span>,</span>
<span id="cb19-28"><a href="discussion.html#cb19-28" tabindex="-1"></a>  <span class="st">&quot;tidyr&quot;</span>,</span>
<span id="cb19-29"><a href="discussion.html#cb19-29" tabindex="-1"></a>  <span class="st">&quot;reshape2&quot;</span>,</span>
<span id="cb19-30"><a href="discussion.html#cb19-30" tabindex="-1"></a>  <span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb19-31"><a href="discussion.html#cb19-31" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> packages_needed){<span class="fu">p_load</span>(<span class="at">char=</span>i)}</span></code></pre></div>
<p>Here is the environment of R programming:</p>
<pre><code># R version 4.3.1 (2023-06-16 ucrt)
# Platform: x86_64-w64-mingw32/x64 (64-bit)
# Running under: Windows 11 x64 (build 26100)
# 
# Matrix products: default
# 
# 
# locale:
# [1] LC_COLLATE=Chinese (Simplified)_China.utf8 
# [2] LC_CTYPE=Chinese (Simplified)_China.utf8   
# [3] LC_MONETARY=Chinese (Simplified)_China.utf8
# [4] LC_NUMERIC=C                               
# [5] LC_TIME=Chinese (Simplified)_China.utf8    
# 
# time zone: Asia/Shanghai
# tzcode source: internal
# 
# attached base packages:
# [1] grid      stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
#  [1] ggplot2_3.5.0         reshape2_1.4.4        tidyr_1.3.1          
#  [4] rpart_4.1.19          GSClassifier_0.4.0    xgboost_2.0.3.1      
#  [7] luckyBase_0.2.0       ComplexHeatmap_2.18.0 readxl_1.4.3         
# [10] pacman_0.5.1         
# 
# loaded via a namespace (and not attached):
#   [1] splines_4.3.1           later_1.3.2             bitops_1.0-7           
#   [4] cellranger_1.1.0        tibble_3.2.1            hardhat_1.3.1          
#   [7] preprocessCore_1.64.0   pROC_1.18.5             lifecycle_1.0.4        
#  [10] rstatix_0.7.2           fastcluster_1.2.6       doParallel_1.0.17      
#  [13] globals_0.16.3          lattice_0.21-8          MASS_7.3-60            
#  [16] backports_1.4.1         magrittr_2.0.3          Hmisc_5.1-2            
#  [19] sass_0.4.8              rmarkdown_2.26          jquerylib_0.1.4        
#  [22] yaml_2.3.8              remotes_2.4.2.1         httpuv_1.6.14          
#  [25] sessioninfo_1.2.2       pkgbuild_1.4.3          DBI_1.2.2              
#  [28] RColorBrewer_1.1-3      lubridate_1.9.3         abind_1.4-5            
#  [31] pkgload_1.3.4           zlibbioc_1.48.0         purrr_1.0.2            
#  [34] BiocGenerics_0.48.1     RCurl_1.98-1.16         nnet_7.3-19            
#  [37] ipred_0.9-14            circlize_0.4.16         lava_1.8.0             
#  [40] GenomeInfoDbData_1.2.11 IRanges_2.36.0          S4Vectors_0.40.2       
#  [43] listenv_0.9.1           parallelly_1.37.1       codetools_0.2-19       
#  [46] tidyselect_1.2.1        shape_1.4.6.1           matrixStats_1.2.0      
#  [49] stats4_4.3.1            dynamicTreeCut_1.63-1   base64enc_0.1-3        
#  [52] jsonlite_1.8.8          caret_6.0-94            GetoptLong_1.0.5       
#  [55] ellipsis_0.3.2          Formula_1.2-5           survival_3.5-5         
#  [58] iterators_1.0.14        signal_1.8-0            foreach_1.5.2          
#  [61] tools_4.3.1             Rcpp_1.0.12             glue_1.7.0             
#  [64] prodlim_2023.08.28      gridExtra_2.3           xfun_0.52              
#  [67] usethis_2.2.3           GenomeInfoDb_1.38.7     dplyr_1.1.4            
#  [70] withr_3.0.0             fastmap_1.1.1           fansi_1.0.6            
#  [73] digest_0.6.34           timechange_0.3.0        R6_2.5.1               
#  [76] mime_0.12               colorspace_2.1-0        GO.db_3.18.0           
#  [79] RSQLite_2.3.5           utf8_1.2.4              generics_0.1.3         
#  [82] tuneR_1.4.6             data.table_1.15.2       recipes_1.0.10         
#  [85] class_7.3-22            httr_1.4.7              htmlwidgets_1.6.4      
#  [88] ModelMetrics_1.2.2.2    pkgconfig_2.0.3         gtable_0.3.4           
#  [91] timeDate_4032.109       blob_1.2.4              impute_1.76.0          
#  [94] XVector_0.42.0          htmltools_0.5.7         carData_3.0-5          
#  [97] profvis_0.3.8           bookdown_0.43           clue_0.3-65            
# [100] scales_1.3.0            Biobase_2.62.0          png_0.1-8              
# [103] gower_1.0.1             knitr_1.50              rstudioapi_0.15.0      
# [106] rjson_0.2.21            checkmate_2.3.1         nlme_3.1-162           
# [109] cachem_1.0.8            GlobalOptions_0.1.2     stringr_1.5.1          
# [112] parallel_4.3.1          miniUI_0.1.1.1          foreign_0.8-84         
# [115] AnnotationDbi_1.64.1    pillar_1.9.0            vctrs_0.6.5            
# [118] urlchecker_1.0.1        promises_1.2.1          randomForest_4.7-1.1   
# [121] ggpubr_0.6.0            car_3.1-2               xtable_1.8-4           
# [124] cluster_2.1.4           htmlTable_2.4.2         evaluate_0.23          
# [127] cli_3.6.2               compiler_4.3.1          rlang_1.1.3            
# [130] crayon_1.5.2            future.apply_1.11.1     ggsignif_0.6.4         
# [133] plyr_1.8.9              fs_1.6.3                stringi_1.8.3          
# [136] WGCNA_1.72-5            munsell_0.5.0           Biostrings_2.70.2      
# [139] devtools_2.4.5          Matrix_1.6-5            bit64_4.0.5            
# [142] future_1.33.1           KEGGREST_1.42.0         shiny_1.8.0            
# [145] broom_1.0.5             memoise_2.0.1           bslib_0.6.1            
# [148] bit_4.0.5</code></pre>
</div>
<div id="subtype-vector" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Subtype Vector<a href="discussion.html#subtype-vector" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the PAD project, the <strong>Subtype Vector</strong> were identified based on independent cohorts under unsupervised hierarchical clustering, instead of an merged expression matrix after batch-effect control of the <strong>sva::ComBat</strong> function.</p>
<p>Here were some considerations:</p>
<ol style="list-style-type: decimal">
<li><strong>Batch control would damage the raw rank differences</strong></li>
</ol>
<p>Here we just showed how could this happen.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="discussion.html#cb21-1" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb21-2"><a href="discussion.html#cb21-2" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(</span>
<span id="cb21-3"><a href="discussion.html#cb21-3" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, </span>
<span id="cb21-4"><a href="discussion.html#cb21-4" tabindex="-1"></a>              <span class="st">&quot;testData.rds&quot;</span>, </span>
<span id="cb21-5"><a href="discussion.html#cb21-5" tabindex="-1"></a>              <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)</span>
<span id="cb21-6"><a href="discussion.html#cb21-6" tabindex="-1"></a>  )</span>
<span id="cb21-7"><a href="discussion.html#cb21-7" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb21-8"><a href="discussion.html#cb21-8" tabindex="-1"></a></span>
<span id="cb21-9"><a href="discussion.html#cb21-9" tabindex="-1"></a><span class="co"># Missing value imputation</span></span>
<span id="cb21-10"><a href="discussion.html#cb21-10" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> <span class="fu">na_fill</span>(expr_pad, </span>
<span id="cb21-11"><a href="discussion.html#cb21-11" tabindex="-1"></a>                    <span class="at">method =</span> <span class="st">&#39;quantile&#39;</span>, </span>
<span id="cb21-12"><a href="discussion.html#cb21-12" tabindex="-1"></a>                    <span class="at">seed =</span> <span class="dv">698</span>,</span>
<span id="cb21-13"><a href="discussion.html#cb21-13" tabindex="-1"></a>                    <span class="at">verbose =</span> F)</span>
<span id="cb21-14"><a href="discussion.html#cb21-14" tabindex="-1"></a></span>
<span id="cb21-15"><a href="discussion.html#cb21-15" tabindex="-1"></a><span class="co"># PADi</span></span>
<span id="cb21-16"><a href="discussion.html#cb21-16" tabindex="-1"></a>padi <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;PAD.train_20220916.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)) </span></code></pre></div>
<p>The raw rank differences were as follows:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="discussion.html#cb22-1" tabindex="-1"></a><span class="co"># Time-consuming</span></span>
<span id="cb22-2"><a href="discussion.html#cb22-2" tabindex="-1"></a>rank_pad <span class="ot">&lt;-</span> GSClassifier<span class="sc">:::</span><span class="fu">trainDataProc_X</span>(</span>
<span id="cb22-3"><a href="discussion.html#cb22-3" tabindex="-1"></a>  expr_pad, </span>
<span id="cb22-4"><a href="discussion.html#cb22-4" tabindex="-1"></a>  <span class="at">geneSet =</span> padi<span class="sc">$</span>geneSet,</span>
<span id="cb22-5"><a href="discussion.html#cb22-5" tabindex="-1"></a>  <span class="at">breakVec=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">1.0</span>)</span>
<span id="cb22-6"><a href="discussion.html#cb22-6" tabindex="-1"></a>)</span>
<span id="cb22-7"><a href="discussion.html#cb22-7" tabindex="-1"></a><span class="fu">print</span>(rank_pad<span class="sc">$</span>dat<span class="sc">$</span>Xbin[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span>
<span id="cb22-8"><a href="discussion.html#cb22-8" tabindex="-1"></a><span class="co">#            ENSG00000122122 ENSG00000117091 ENSG00000163219 ENSG00000136167</span></span>
<span id="cb22-9"><a href="discussion.html#cb22-9" tabindex="-1"></a><span class="co"># GSM2235556               3               2               1               4</span></span>
<span id="cb22-10"><a href="discussion.html#cb22-10" tabindex="-1"></a><span class="co"># GSM2235557               3               3               1               4</span></span>
<span id="cb22-11"><a href="discussion.html#cb22-11" tabindex="-1"></a><span class="co"># GSM2235558               3               2               1               4</span></span>
<span id="cb22-12"><a href="discussion.html#cb22-12" tabindex="-1"></a><span class="co"># GSM2235559               3               4               1               4</span></span>
<span id="cb22-13"><a href="discussion.html#cb22-13" tabindex="-1"></a><span class="co"># GSM2235560               3               3               1               4</span></span>
<span id="cb22-14"><a href="discussion.html#cb22-14" tabindex="-1"></a><span class="co">#            ENSG00000005844</span></span>
<span id="cb22-15"><a href="discussion.html#cb22-15" tabindex="-1"></a><span class="co"># GSM2235556               3</span></span>
<span id="cb22-16"><a href="discussion.html#cb22-16" tabindex="-1"></a><span class="co"># GSM2235557               2</span></span>
<span id="cb22-17"><a href="discussion.html#cb22-17" tabindex="-1"></a><span class="co"># GSM2235558               3</span></span>
<span id="cb22-18"><a href="discussion.html#cb22-18" tabindex="-1"></a><span class="co"># GSM2235559               3</span></span>
<span id="cb22-19"><a href="discussion.html#cb22-19" tabindex="-1"></a><span class="co"># GSM2235560               3</span></span></code></pre></div>
<p>Here, the expression matrix after removing batch effect were calculated:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="discussion.html#cb23-1" tabindex="-1"></a><span class="co"># Cleaned and normalized data for sva::Combat</span></span>
<span id="cb23-2"><a href="discussion.html#cb23-2" tabindex="-1"></a>expr_pad2 <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr_pad, <span class="dv">2</span>, scale)</span>
<span id="cb23-3"><a href="discussion.html#cb23-3" tabindex="-1"></a><span class="fu">rownames</span>(expr_pad2) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(expr_pad)</span></code></pre></div>
<p>Remove batch effects:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="discussion.html#cb24-1" tabindex="-1"></a><span class="co"># BatchQC data</span></span>
<span id="cb24-2"><a href="discussion.html#cb24-2" tabindex="-1"></a><span class="fu">library</span>(sva)</span>
<span id="cb24-3"><a href="discussion.html#cb24-3" tabindex="-1"></a><span class="co"># Loading required package: mgcv</span></span>
<span id="cb24-4"><a href="discussion.html#cb24-4" tabindex="-1"></a><span class="co"># Loading required package: nlme</span></span>
<span id="cb24-5"><a href="discussion.html#cb24-5" tabindex="-1"></a><span class="co"># This is mgcv 1.8-42. For overview type &#39;help(&quot;mgcv-package&quot;)&#39;.</span></span>
<span id="cb24-6"><a href="discussion.html#cb24-6" tabindex="-1"></a><span class="co"># Loading required package: genefilter</span></span>
<span id="cb24-7"><a href="discussion.html#cb24-7" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb24-8"><a href="discussion.html#cb24-8" tabindex="-1"></a><span class="co"># Attaching package: &#39;genefilter&#39;</span></span>
<span id="cb24-9"><a href="discussion.html#cb24-9" tabindex="-1"></a><span class="co"># The following object is masked from &#39;package:ComplexHeatmap&#39;:</span></span>
<span id="cb24-10"><a href="discussion.html#cb24-10" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb24-11"><a href="discussion.html#cb24-11" tabindex="-1"></a><span class="co">#     dist2</span></span>
<span id="cb24-12"><a href="discussion.html#cb24-12" tabindex="-1"></a><span class="co"># Loading required package: BiocParallel</span></span>
<span id="cb24-13"><a href="discussion.html#cb24-13" tabindex="-1"></a>batch <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_phenotype_part<span class="sc">$</span>Dataset</span>
<span id="cb24-14"><a href="discussion.html#cb24-14" tabindex="-1"></a>expr_pad2_rmbat <span class="ot">&lt;-</span> <span class="fu">ComBat</span>(</span>
<span id="cb24-15"><a href="discussion.html#cb24-15" tabindex="-1"></a>  <span class="at">dat=</span>expr_pad2,</span>
<span id="cb24-16"><a href="discussion.html#cb24-16" tabindex="-1"></a>  <span class="at">batch=</span>batch, </span>
<span id="cb24-17"><a href="discussion.html#cb24-17" tabindex="-1"></a>  <span class="at">mod=</span><span class="cn">NULL</span>)</span>
<span id="cb24-18"><a href="discussion.html#cb24-18" tabindex="-1"></a><span class="co"># Found14batches</span></span>
<span id="cb24-19"><a href="discussion.html#cb24-19" tabindex="-1"></a><span class="co"># Adjusting for0covariate(s) or covariate level(s)</span></span>
<span id="cb24-20"><a href="discussion.html#cb24-20" tabindex="-1"></a><span class="co"># Standardizing Data across genes</span></span>
<span id="cb24-21"><a href="discussion.html#cb24-21" tabindex="-1"></a><span class="co"># Fitting L/S model and finding priors</span></span>
<span id="cb24-22"><a href="discussion.html#cb24-22" tabindex="-1"></a><span class="co"># Finding parametric adjustments</span></span>
<span id="cb24-23"><a href="discussion.html#cb24-23" tabindex="-1"></a><span class="co"># Adjusting the Data</span></span></code></pre></div>
<p>Look at the alteration of batch effects in GC datasets:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="discussion.html#cb25-1" tabindex="-1"></a></span>
<span id="cb25-2"><a href="discussion.html#cb25-2" tabindex="-1"></a><span class="co"># Compare</span></span>
<span id="cb25-3"><a href="discussion.html#cb25-3" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(expr_pad2)</span>
<span id="cb25-4"><a href="discussion.html#cb25-4" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(expr_pad2_rmbat)</span>
<span id="cb25-5"><a href="discussion.html#cb25-5" tabindex="-1"></a>df3 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb25-6"><a href="discussion.html#cb25-6" tabindex="-1"></a>  <span class="fu">cbind</span>(df1, <span class="at">type =</span> <span class="st">&#39;After scale_before Combat&#39;</span>),</span>
<span id="cb25-7"><a href="discussion.html#cb25-7" tabindex="-1"></a>  <span class="fu">cbind</span>(df2, <span class="at">type =</span> <span class="st">&#39;After scale_after Combat&#39;</span>)</span>
<span id="cb25-8"><a href="discussion.html#cb25-8" tabindex="-1"></a>)</span>
<span id="cb25-9"><a href="discussion.html#cb25-9" tabindex="-1"></a>df3<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="fu">factor</span>(df3<span class="sc">$</span>type, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&#39;After scale_before Combat&#39;</span>, <span class="st">&#39;After scale_after Combat&#39;</span>))</span>
<span id="cb25-10"><a href="discussion.html#cb25-10" tabindex="-1"></a>df3<span class="sc">$</span>dataset <span class="ot">&lt;-</span> <span class="fu">convert</span>(df3<span class="sc">$</span>Var2,<span class="st">&#39;ID&#39;</span>,<span class="st">&#39;Dataset&#39;</span>, testData<span class="sc">$</span>PanSTAD_phenotype_part)</span>
<span id="cb25-11"><a href="discussion.html#cb25-11" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> <span class="fu">unique</span>(df3<span class="sc">$</span>dataset)</span>
<span id="cb25-12"><a href="discussion.html#cb25-12" tabindex="-1"></a>dataset_color <span class="ot">&lt;-</span> mycolor[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">10</span>,<span class="dv">21</span>,<span class="dv">22</span>,<span class="dv">23</span>,<span class="dv">24</span>,<span class="dv">25</span>,<span class="dv">54</span>,<span class="dv">56</span>,<span class="dv">60</span>)]</span>
<span id="cb25-13"><a href="discussion.html#cb25-13" tabindex="-1"></a></span>
<span id="cb25-14"><a href="discussion.html#cb25-14" tabindex="-1"></a><span class="co"># ggplot: Combat‰ºº‰πéÊúâ‰∏ÄÁÇπÁÇπÊïàÊûú</span></span>
<span id="cb25-15"><a href="discussion.html#cb25-15" tabindex="-1"></a><span class="cf">if</span>(T){</span>
<span id="cb25-16"><a href="discussion.html#cb25-16" tabindex="-1"></a>  </span>
<span id="cb25-17"><a href="discussion.html#cb25-17" tabindex="-1"></a>  size<span class="ot">=</span><span class="dv">6</span></span>
<span id="cb25-18"><a href="discussion.html#cb25-18" tabindex="-1"></a>  </span>
<span id="cb25-19"><a href="discussion.html#cb25-19" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df3,<span class="fu">aes</span>(<span class="at">x=</span>Var2,<span class="at">y=</span>value,<span class="at">fill=</span>dataset,<span class="at">color=</span>dataset)) <span class="sc">+</span> </span>
<span id="cb25-20"><a href="discussion.html#cb25-20" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">outlier.size =</span> <span class="sc">-</span><span class="fl">0.3</span>, <span class="at">size =</span> <span class="fl">0.02</span>) <span class="sc">+</span> </span>
<span id="cb25-21"><a href="discussion.html#cb25-21" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb25-22"><a href="discussion.html#cb25-22" tabindex="-1"></a>      <span class="at">breaks =</span> dataset,</span>
<span id="cb25-23"><a href="discussion.html#cb25-23" tabindex="-1"></a>      <span class="at">values =</span> dataset_color,</span>
<span id="cb25-24"><a href="discussion.html#cb25-24" tabindex="-1"></a>      <span class="at">labels =</span> dataset</span>
<span id="cb25-25"><a href="discussion.html#cb25-25" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb25-26"><a href="discussion.html#cb25-26" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(</span>
<span id="cb25-27"><a href="discussion.html#cb25-27" tabindex="-1"></a>      <span class="at">breaks =</span> dataset,</span>
<span id="cb25-28"><a href="discussion.html#cb25-28" tabindex="-1"></a>      <span class="at">values =</span> dataset_color,</span>
<span id="cb25-29"><a href="discussion.html#cb25-29" tabindex="-1"></a>      <span class="at">labels =</span> dataset</span>
<span id="cb25-30"><a href="discussion.html#cb25-30" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb25-31"><a href="discussion.html#cb25-31" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb25-32"><a href="discussion.html#cb25-32" tabindex="-1"></a>           <span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size=</span><span class="dv">1</span>))) <span class="sc">+</span> </span>
<span id="cb25-33"><a href="discussion.html#cb25-33" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(. <span class="sc">~</span> type, </span>
<span id="cb25-34"><a href="discussion.html#cb25-34" tabindex="-1"></a>               <span class="at">nrow =</span> <span class="fu">length</span>(<span class="fu">unique</span>(df3<span class="sc">$</span>type)),</span>
<span id="cb25-35"><a href="discussion.html#cb25-35" tabindex="-1"></a>               <span class="at">scales =</span> <span class="st">&#39;free&#39;</span>) <span class="sc">+</span></span>
<span id="cb25-36"><a href="discussion.html#cb25-36" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Samples in External cohorts of GC&#39;</span>,</span>
<span id="cb25-37"><a href="discussion.html#cb25-37" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">&#39;Gene expression&#39;</span>,</span>
<span id="cb25-38"><a href="discussion.html#cb25-38" tabindex="-1"></a>         <span class="at">fill =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb25-39"><a href="discussion.html#cb25-39" tabindex="-1"></a>    <span class="co"># coord_flip() + </span></span>
<span id="cb25-40"><a href="discussion.html#cb25-40" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-41"><a href="discussion.html#cb25-41" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb25-42"><a href="discussion.html#cb25-42" tabindex="-1"></a>      <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-43"><a href="discussion.html#cb25-43" tabindex="-1"></a>      <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> size<span class="sc">/</span><span class="dv">15</span><span class="sc">*</span><span class="dv">12</span>,<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>,<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>),</span>
<span id="cb25-44"><a href="discussion.html#cb25-44" tabindex="-1"></a>      <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> size<span class="sc">*</span><span class="fl">1.5</span>,<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>,<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>),</span>
<span id="cb25-45"><a href="discussion.html#cb25-45" tabindex="-1"></a>      <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> size<span class="sc">*</span><span class="fl">1.5</span>,<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>,<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>),</span>
<span id="cb25-46"><a href="discussion.html#cb25-46" tabindex="-1"></a>      <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> size<span class="sc">/</span><span class="dv">15</span><span class="sc">*</span><span class="dv">25</span>,<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>,<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>),</span>
<span id="cb25-47"><a href="discussion.html#cb25-47" tabindex="-1"></a>      <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> size<span class="sc">/</span><span class="dv">15</span><span class="sc">*</span><span class="dv">25</span>,<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>,<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>),</span>
<span id="cb25-48"><a href="discussion.html#cb25-48" tabindex="-1"></a>      <span class="at">legend.position=</span><span class="st">&#39;right&#39;</span>,</span>
<span id="cb25-49"><a href="discussion.html#cb25-49" tabindex="-1"></a>      <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">&quot;white&quot;</span>,<span class="at">size =</span> <span class="dv">2</span>),</span>
<span id="cb25-50"><a href="discussion.html#cb25-50" tabindex="-1"></a>      <span class="at">strip.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> size<span class="sc">*</span><span class="fl">1.5</span>,<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>,<span class="at">face =</span> <span class="st">&quot;bold&quot;</span>, <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="fl">0.5</span>, <span class="at">r =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="fl">0.5</span>, <span class="at">l =</span> <span class="dv">0</span>, <span class="at">unit =</span> <span class="st">&quot;cm&quot;</span>)),</span>
<span id="cb25-51"><a href="discussion.html#cb25-51" tabindex="-1"></a>      <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">&quot;black&quot;</span>,<span class="at">size=</span><span class="dv">2</span>),</span>
<span id="cb25-52"><a href="discussion.html#cb25-52" tabindex="-1"></a>      <span class="co"># panel.grid = element_blank(),</span></span>
<span id="cb25-53"><a href="discussion.html#cb25-53" tabindex="-1"></a>      <span class="co"># panel.border=element_rect(fill=&#39;transparent&#39;,color=&#39;transparent&#39;),</span></span>
<span id="cb25-54"><a href="discussion.html#cb25-54" tabindex="-1"></a>      <span class="co"># axis.ticks = element_line(colour = &quot;black&quot;,size = 2.5,linetype = 1,lineend = &#39;square&#39;),</span></span>
<span id="cb25-55"><a href="discussion.html#cb25-55" tabindex="-1"></a>      <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>()</span>
<span id="cb25-56"><a href="discussion.html#cb25-56" tabindex="-1"></a>      <span class="co"># axis.line = element_line(colour = &quot;black&quot;,size = 2.5,linetype = 1,lineend = &#39;square&#39;)</span></span>
<span id="cb25-57"><a href="discussion.html#cb25-57" tabindex="-1"></a>    ) </span>
<span id="cb25-58"><a href="discussion.html#cb25-58" tabindex="-1"></a>  <span class="fu">print</span>(p)</span>
<span id="cb25-59"><a href="discussion.html#cb25-59" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="Discussion_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>The rank differences after batch-effect control were as follows:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="discussion.html#cb26-1" tabindex="-1"></a><span class="co"># Time-consuming</span></span>
<span id="cb26-2"><a href="discussion.html#cb26-2" tabindex="-1"></a>rank_pad_rmbat <span class="ot">&lt;-</span> GSClassifier<span class="sc">:::</span><span class="fu">trainDataProc_X</span>(</span>
<span id="cb26-3"><a href="discussion.html#cb26-3" tabindex="-1"></a>  expr_pad2_rmbat, </span>
<span id="cb26-4"><a href="discussion.html#cb26-4" tabindex="-1"></a>  <span class="at">geneSet =</span> padi<span class="sc">$</span>geneSet,</span>
<span id="cb26-5"><a href="discussion.html#cb26-5" tabindex="-1"></a>  <span class="at">breakVec=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">1.0</span>)</span>
<span id="cb26-6"><a href="discussion.html#cb26-6" tabindex="-1"></a>)</span></code></pre></div>
<p>The comparison demonstrated that the adjustment for batch effects would change the rank differences:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="discussion.html#cb27-1" tabindex="-1"></a><span class="fu">mean</span>(rank_pad_rmbat<span class="sc">$</span>dat<span class="sc">$</span>Xbin <span class="sc">==</span> rank_pad<span class="sc">$</span>dat<span class="sc">$</span>Xbin)</span>
<span id="cb27-2"><a href="discussion.html#cb27-2" tabindex="-1"></a><span class="co"># [1] 0.7788832</span></span></code></pre></div>
<p>The rank differences, the base of TSP normalization, were critical for model training and subtype identification, so it‚Äôs not recommended to adjust for batch effects using <strong>sva::ComBat</strong> before model training in GSClassifier.</p>
<ol start="2" style="list-style-type: decimal">
<li><strong>Batch control would lead to an unbalanced subtype vectors</strong>.</li>
</ol>
<p>Here, a heatmap were used to show the self-clustering of the training cohort based on <strong>PIAM</strong> and <strong>PIDG</strong>:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="discussion.html#cb28-1" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">PAD</span>(</span>
<span id="cb28-2"><a href="discussion.html#cb28-2" tabindex="-1"></a>  <span class="at">expr =</span> expr_pad2_rmbat,</span>
<span id="cb28-3"><a href="discussion.html#cb28-3" tabindex="-1"></a>  <span class="at">cluster.method =</span> <span class="st">&quot;ward.D2&quot;</span>,</span>
<span id="cb28-4"><a href="discussion.html#cb28-4" tabindex="-1"></a>  <span class="at">subtype =</span> <span class="st">&quot;PAD.train_20220916&quot;</span>,</span>
<span id="cb28-5"><a href="discussion.html#cb28-5" tabindex="-1"></a>  <span class="at">verbose =</span> T</span>
<span id="cb28-6"><a href="discussion.html#cb28-6" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="Discussion_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>Look at the percentage of PAD subtypes:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="discussion.html#cb29-1" tabindex="-1"></a><span class="fu">table</span>(res<span class="sc">$</span>Data<span class="sc">$</span><span class="st">`</span><span class="at">PAD subtype</span><span class="st">`</span>)<span class="sc">/</span><span class="fu">ncol</span>(expr_pad2_rmbat)</span>
<span id="cb29-2"><a href="discussion.html#cb29-2" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb29-3"><a href="discussion.html#cb29-3" tabindex="-1"></a><span class="co">#      PAD-I     PAD-II    PAD-III     PAD-IV </span></span>
<span id="cb29-4"><a href="discussion.html#cb29-4" tabindex="-1"></a><span class="co"># 0.05169938 0.25801819 0.63140258 0.05887985</span></span></code></pre></div>
<p>This results demonstrated the unbalanced self-clustering of PAD subtypes based on the expression matrix after <strong>sva::ComBat</strong>, which would damage the performance of trained models.</p>
</div>
<div id="missing-value-imputation-mvi" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Missing value imputation (MVI)<a href="discussion.html#missing-value-imputation-mvi" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<!--
Missing Value Imputation (MVI); 
+ Missing Value type: Missing Completely at Random (MCAR), Missing at Random (MAR), and Not Missing at Random (NMAR).
+ The main reason of missing value in the scenario of **GSClassifier**?
+ Why is missing value imputation needed in **GSClassifier** modeling?
+ Routine algorithms of missing value imputation for DNA array (mRNA expression)?
+ Missing value imputation in subtype calling?
+ The flowchart of quantile algorithm?
+ zero == Cutout; https://arxiv.org/abs/1708.04552; In this paper, we show that the simple regularization technique of randomly masking out square regions of input during training, which we call cutout, can be used to improve the robustness and overall performance of convolutional neural networks. Not only is this method extremely easy to implement, but we also demonstrate that it can be used in conjunction with existing forms of data augmentation and other regularizers to further improve model performance. We evaluate this method by applying it to current state-of-the-art architectures on the CIFAR-10, CIFAR-100, and SVHN datasets, yielding new state-of-the-art results of 2.56%, 15.20%, and 1.30% test error respectively.
+ Experiments: leave genes out of validation cohorts and then test the performance of **GSClassifier**-PADi.
-->
<p>Due to reasons like weak signal, contamination of microarray surfaces, inappropriate manual operations, insufficient resolution, or systematic errors during the laboratory process <span class="citation">[<a href="#ref-RN387">8</a>‚Äì<a href="#ref-RN382">10</a>]</span>, <strong>missing value</strong> in high-input genetic data is common. Generally, tiny missing values could be just dealt with case deletion, while the biological discovery might be damaged when the missing rate tops 15% <span class="citation">[<a href="#ref-RN392">11</a>,<a href="#ref-RN386">12</a>]</span>. Currently, lots of methods, including statistic-based or machine learning-based methods (Figure <a href="discussion.html#fig:mvi01">2.1</a>), had been developed for <strong>missing value imputation (MVI)</strong> <span class="citation">[<a href="#ref-RN386">12</a>]</span>. Wang et al <span class="citation">[<a href="#ref-RN384">13</a>]</span> categorized MVI methods into simple (zeros or average), biology knowledge-, global learning-, local learning-, and hybrid-based methods. In order to satisfy the working conditions of <strong>xgboost</strong> <span class="citation">[<a href="#ref-xgboost">14</a>]</span> functions (<code>xgb.train</code>, <code>xgboost</code>, and <code>xgb.cv</code>) in GSClassifer, missing values in the expression matrix must be deleted or imputation.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:mvi01"></span>
<img src="fig/mvi-01.png" alt="Missing value imputation methods reviewed by Hasan et al." width="90%" />
<p class="caption">
Figure 2.1: Missing value imputation methods reviewed by Hasan et al.
</p>
</div>
<p>In <strong>PAD</strong> project, several strategies were applied to reduce the impact of missing values as possible. First, both <strong>PIAM</strong> and <strong>PIDG</strong> in <strong>PAD</strong> project were curated GEPs that were not missing in over 80% of gastric cancer datasets. Here we showed the actual distribution of missing values across samples in gastric cancer datasets we used.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="discussion.html#cb30-1" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb30-2"><a href="discussion.html#cb30-2" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(</span>
<span id="cb30-3"><a href="discussion.html#cb30-3" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, </span>
<span id="cb30-4"><a href="discussion.html#cb30-4" tabindex="-1"></a>              <span class="st">&quot;testData.rds&quot;</span>, </span>
<span id="cb30-5"><a href="discussion.html#cb30-5" tabindex="-1"></a>              <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)</span>
<span id="cb30-6"><a href="discussion.html#cb30-6" tabindex="-1"></a>  )</span>
<span id="cb30-7"><a href="discussion.html#cb30-7" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb30-8"><a href="discussion.html#cb30-8" tabindex="-1"></a></span>
<span id="cb30-9"><a href="discussion.html#cb30-9" tabindex="-1"></a><span class="co"># Missing value</span></span>
<span id="cb30-10"><a href="discussion.html#cb30-10" tabindex="-1"></a>expr_pad_na <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr_pad, <span class="dv">2</span>, </span>
<span id="cb30-11"><a href="discussion.html#cb30-11" tabindex="-1"></a>                     <span class="cf">function</span>(x) <span class="fu">sum</span>(<span class="fu">is.na</span>(x))<span class="sc">/</span><span class="fu">length</span>(x))</span>
<span id="cb30-12"><a href="discussion.html#cb30-12" tabindex="-1"></a>expr_pad_na_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb30-13"><a href="discussion.html#cb30-13" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">names</span>(expr_pad_na),</span>
<span id="cb30-14"><a href="discussion.html#cb30-14" tabindex="-1"></a>  <span class="at">prob =</span> <span class="fu">as.numeric</span>(expr_pad_na),</span>
<span id="cb30-15"><a href="discussion.html#cb30-15" tabindex="-1"></a>  <span class="at">stringsAsFactors =</span> F</span>
<span id="cb30-16"><a href="discussion.html#cb30-16" tabindex="-1"></a>)</span></code></pre></div>
<p>As shown in Figure <a href="discussion.html#fig:mvi02">2.2</a>, the percentage of all samples in gastric cancer datasets we used is lower than 8%.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="discussion.html#cb31-1" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb31-2"><a href="discussion.html#cb31-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> expr_pad_na_df, </span>
<span id="cb31-3"><a href="discussion.html#cb31-3" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> prob)) <span class="sc">+</span> </span>
<span id="cb31-4"><a href="discussion.html#cb31-4" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&#39;identity&#39;</span>, <span class="at">color =</span> mycolor[<span class="dv">3</span>]) <span class="sc">+</span> </span>
<span id="cb31-5"><a href="discussion.html#cb31-5" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">labels=</span>scales<span class="sc">::</span>percent) <span class="sc">+</span> </span>
<span id="cb31-6"><a href="discussion.html#cb31-6" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Samples in gastric cancer cohorts&#39;</span>, </span>
<span id="cb31-7"><a href="discussion.html#cb31-7" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Percentage of missing value&#39;</span>) <span class="sc">+</span> </span>
<span id="cb31-8"><a href="discussion.html#cb31-8" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb31-9"><a href="discussion.html#cb31-9" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb31-10"><a href="discussion.html#cb31-10" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb31-11"><a href="discussion.html#cb31-11" tabindex="-1"></a>    <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb31-12"><a href="discussion.html#cb31-12" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb31-13"><a href="discussion.html#cb31-13" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)</span>
<span id="cb31-14"><a href="discussion.html#cb31-14" tabindex="-1"></a>  )</span>
<span id="cb31-15"><a href="discussion.html#cb31-15" tabindex="-1"></a><span class="fu">print</span>(p1)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:mvi02"></span>
<img src="Discussion_files/figure-html/mvi02-1.png" alt="The distribution of missing value across gastric cancer samples." width="80%" />
<p class="caption">
Figure 2.2: The distribution of missing value across gastric cancer samples.
</p>
</div>
<p>Second, we did conduct some MVI strategies to deal with data before model training in <strong>GSClassifier</strong>. Due to the low missing rate of our experimental data, we just <strong>set missing values as zero</strong> during model training and subtype identification in the early version of PADi (<strong>PAD.train.v20200110</strong>). The model seemed to be robust in both the internal cohort and external cohorts, and greatly predicted the response to immune checkpoint inhibitors (ICIs) in advanced gastric cancer.</p>
<p>In the latest version of PADi (<strong>PAD.train.v20220916</strong>), we designed the so-called <strong>quantile</strong> algorithm for random MVI during <strong>PADi</strong> model training, which also seemed to work well for PADi model training.</p>
<!--

We supposed that ...
Add a flowchart for quantile algorithm

-->
<p>Here, we demonstrated the principle of <strong>quantile</strong> algorithm in the simulated dataset:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="discussion.html#cb32-1" tabindex="-1"></a><span class="co"># Simulated data</span></span>
<span id="cb32-2"><a href="discussion.html#cb32-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">&#39;./data/simulated-data.xlsx&#39;</span>, <span class="at">sheet =</span> <span class="st">&#39;RNA&#39;</span>)</span>
<span id="cb32-3"><a href="discussion.html#cb32-3" tabindex="-1"></a>expr0 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(x[,<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb32-4"><a href="discussion.html#cb32-4" tabindex="-1"></a><span class="fu">rownames</span>(expr0) <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">as.matrix</span>(x[,<span class="dv">1</span>])); <span class="fu">rm</span>(x)</span>
<span id="cb32-5"><a href="discussion.html#cb32-5" tabindex="-1"></a></span>
<span id="cb32-6"><a href="discussion.html#cb32-6" tabindex="-1"></a><span class="co"># MVI with Quantile algorithm</span></span>
<span id="cb32-7"><a href="discussion.html#cb32-7" tabindex="-1"></a>expr <span class="ot">&lt;-</span> expr0</span>
<span id="cb32-8"><a href="discussion.html#cb32-8" tabindex="-1"></a>na.pos <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr,<span class="dv">2</span>,is.one.na)</span>
<span id="cb32-9"><a href="discussion.html#cb32-9" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">478</span>); seeds <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(expr)<span class="sc">*</span><span class="dv">10</span>, <span class="fu">sum</span>(na.pos), <span class="at">replace =</span> F)</span>
<span id="cb32-10"><a href="discussion.html#cb32-10" tabindex="-1"></a>tSample <span class="ot">&lt;-</span> <span class="fu">names</span>(na.pos)[na.pos]</span>
<span id="cb32-11"><a href="discussion.html#cb32-11" tabindex="-1"></a>quantile_vector <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span>)<span class="sc">/</span><span class="dv">1000</span></span>
<span id="cb32-12"><a href="discussion.html#cb32-12" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tSample)){ <span class="co"># i=1</span></span>
<span id="cb32-13"><a href="discussion.html#cb32-13" tabindex="-1"></a>  </span>
<span id="cb32-14"><a href="discussion.html#cb32-14" tabindex="-1"></a>  sample.i <span class="ot">&lt;-</span> tSample[i]</span>
<span id="cb32-15"><a href="discussion.html#cb32-15" tabindex="-1"></a>  expr.i <span class="ot">&lt;-</span> expr[, sample.i]</span>
<span id="cb32-16"><a href="discussion.html#cb32-16" tabindex="-1"></a>  expr.i.max <span class="ot">&lt;-</span> <span class="fu">max</span>(expr.i, <span class="at">na.rm =</span> T)</span>
<span id="cb32-17"><a href="discussion.html#cb32-17" tabindex="-1"></a>  expr.i.min <span class="ot">&lt;-</span> <span class="fu">min</span>(expr.i, <span class="at">na.rm =</span> T)</span>
<span id="cb32-18"><a href="discussion.html#cb32-18" tabindex="-1"></a>  <span class="fu">set.seed</span>(seeds[i]);</span>
<span id="cb32-19"><a href="discussion.html#cb32-19" tabindex="-1"></a>  </span>
<span id="cb32-20"><a href="discussion.html#cb32-20" tabindex="-1"></a>  <span class="co"># Details of quantile algorithm</span></span>
<span id="cb32-21"><a href="discussion.html#cb32-21" tabindex="-1"></a>  expr.i[<span class="fu">is.na</span>(expr.i)] <span class="ot">&lt;-</span></span>
<span id="cb32-22"><a href="discussion.html#cb32-22" tabindex="-1"></a>    expr.i.min <span class="sc">+</span></span>
<span id="cb32-23"><a href="discussion.html#cb32-23" tabindex="-1"></a>    (expr.i.max<span class="sc">-</span>expr.i.min) <span class="sc">*</span> <span class="fu">sample</span>(quantile_vector,</span>
<span id="cb32-24"><a href="discussion.html#cb32-24" tabindex="-1"></a>                                     <span class="fu">sum</span>(<span class="fu">is.na</span>(expr.i)),</span>
<span id="cb32-25"><a href="discussion.html#cb32-25" tabindex="-1"></a>                                     <span class="at">replace =</span> T)</span>
<span id="cb32-26"><a href="discussion.html#cb32-26" tabindex="-1"></a>  expr[, sample.i] <span class="ot">&lt;-</span> expr.i</span>
<span id="cb32-27"><a href="discussion.html#cb32-27" tabindex="-1"></a>}</span>
<span id="cb32-28"><a href="discussion.html#cb32-28" tabindex="-1"></a>  </span>
<span id="cb32-29"><a href="discussion.html#cb32-29" tabindex="-1"></a></span>
<span id="cb32-30"><a href="discussion.html#cb32-30" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb32-31"><a href="discussion.html#cb32-31" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb32-32"><a href="discussion.html#cb32-32" tabindex="-1"></a><span class="fu">print</span>(expr0)</span>
<span id="cb32-33"><a href="discussion.html#cb32-33" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb32-34"><a href="discussion.html#cb32-34" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression without NA value:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb32-35"><a href="discussion.html#cb32-35" tabindex="-1"></a><span class="fu">print</span>(expr)</span>
<span id="cb32-36"><a href="discussion.html#cb32-36" tabindex="-1"></a><span class="co"># RNA expression: </span></span>
<span id="cb32-37"><a href="discussion.html#cb32-37" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb32-38"><a href="discussion.html#cb32-38" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60    0.21    0.30    0.40</span></span>
<span id="cb32-39"><a href="discussion.html#cb32-39" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58    0.22    0.31    0.35</span></span>
<span id="cb32-40"><a href="discussion.html#cb32-40" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61      NA    0.29    0.30</span></span>
<span id="cb32-41"><a href="discussion.html#cb32-41" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40    0.51    0.52    0.60</span></span>
<span id="cb32-42"><a href="discussion.html#cb32-42" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35    0.52    0.54    0.58</span></span>
<span id="cb32-43"><a href="discussion.html#cb32-43" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30    0.53      NA    0.61</span></span>
<span id="cb32-44"><a href="discussion.html#cb32-44" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09    0.11    0.12    0.14</span></span>
<span id="cb32-45"><a href="discussion.html#cb32-45" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb32-46"><a href="discussion.html#cb32-46" tabindex="-1"></a><span class="co"># RNA expression without NA value: </span></span>
<span id="cb32-47"><a href="discussion.html#cb32-47" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb32-48"><a href="discussion.html#cb32-48" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60 0.21000 0.30000    0.40</span></span>
<span id="cb32-49"><a href="discussion.html#cb32-49" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58 0.22000 0.31000    0.35</span></span>
<span id="cb32-50"><a href="discussion.html#cb32-50" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61 0.43256 0.29000    0.30</span></span>
<span id="cb32-51"><a href="discussion.html#cb32-51" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40 0.51000 0.52000    0.60</span></span>
<span id="cb32-52"><a href="discussion.html#cb32-52" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35 0.52000 0.54000    0.58</span></span>
<span id="cb32-53"><a href="discussion.html#cb32-53" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30 0.53000 0.32622    0.61</span></span>
<span id="cb32-54"><a href="discussion.html#cb32-54" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09 0.11000 0.12000    0.14</span></span></code></pre></div>
<p>Look at the new matrix via heatmap, where the clustering result is not significantly disturbed after MVI:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="discussion.html#cb33-1" tabindex="-1"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(expr))), <span class="at">name =</span> <span class="st">&quot;Z-score&quot;</span>, <span class="at">column_title =</span> <span class="st">&quot;After MVI&quot;</span>)</span></code></pre></div>
<p><img src="Discussion_files/figure-html/unnamed-chunk-14-1.png" width="60%" style="display: block; margin: auto;" /></p>
<p>Because missing values might damage the integrity of biological information, we explored <strong>how much the number of missing values in one sample impacts subtype identification via PADi</strong>. The steps are as follows: (i) we used the ‚Äúquantile‚Äù algorithm to do MVI in the internal validation cohort of gastric cancer; (ii) we randomly masked different proportions of genes as zero expression; (iii) we calculated the relative multi-ROC <span class="citation">[<a href="#ref-pROC">15</a>]</span> (masked data vs.¬†MVI data). In <strong>GSClassifier</strong>, we developed a function called <strong>mv_tolerance</strong> to complete the task.</p>
<ol style="list-style-type: lower-roman">
<li>Load the internal validation cohort:</li>
</ol>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="discussion.html#cb34-1" tabindex="-1"></a><span class="co"># Internal validation cohort</span></span>
<span id="cb34-2"><a href="discussion.html#cb34-2" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(</span>
<span id="cb34-3"><a href="discussion.html#cb34-3" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;testData.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)</span>
<span id="cb34-4"><a href="discussion.html#cb34-4" tabindex="-1"></a>  )</span>
<span id="cb34-5"><a href="discussion.html#cb34-5" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb34-6"><a href="discussion.html#cb34-6" tabindex="-1"></a>modelInfo <span class="ot">&lt;-</span> <span class="fu">modelData</span>(</span>
<span id="cb34-7"><a href="discussion.html#cb34-7" tabindex="-1"></a>  <span class="at">design =</span> testData<span class="sc">$</span>PanSTAD_phenotype_part,</span>
<span id="cb34-8"><a href="discussion.html#cb34-8" tabindex="-1"></a>  <span class="at">id.col =</span> <span class="st">&quot;ID&quot;</span>,</span>
<span id="cb34-9"><a href="discussion.html#cb34-9" tabindex="-1"></a>  <span class="at">variable =</span> <span class="fu">c</span>(<span class="st">&quot;platform&quot;</span>, <span class="st">&quot;PAD_subtype&quot;</span>),</span>
<span id="cb34-10"><a href="discussion.html#cb34-10" tabindex="-1"></a>  <span class="at">Prop =</span> <span class="fl">0.7</span>,</span>
<span id="cb34-11"><a href="discussion.html#cb34-11" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">19871</span></span>
<span id="cb34-12"><a href="discussion.html#cb34-12" tabindex="-1"></a>)</span>
<span id="cb34-13"><a href="discussion.html#cb34-13" tabindex="-1"></a>validInform <span class="ot">&lt;-</span> modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Valid</span>
<span id="cb34-14"><a href="discussion.html#cb34-14" tabindex="-1"></a>expr_pad_innervalid <span class="ot">&lt;-</span> expr_pad[,validInform<span class="sc">$</span>ID]</span></code></pre></div>
<ol start="2" style="list-style-type: lower-roman">
<li>Missing value tolerance analysis:</li>
</ol>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="discussion.html#cb35-1" tabindex="-1"></a><span class="co"># Time-consuming</span></span>
<span id="cb35-2"><a href="discussion.html#cb35-2" tabindex="-1"></a>mvt <span class="ot">&lt;-</span> <span class="fu">mv_tolerance</span>(</span>
<span id="cb35-3"><a href="discussion.html#cb35-3" tabindex="-1"></a>  <span class="at">X =</span> expr_pad_innervalid,</span>
<span id="cb35-4"><a href="discussion.html#cb35-4" tabindex="-1"></a>  <span class="at">gene.loss =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">12</span>),</span>
<span id="cb35-5"><a href="discussion.html#cb35-5" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>),</span>
<span id="cb35-6"><a href="discussion.html#cb35-6" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;PAD.train_20220916&quot;</span>,</span>
<span id="cb35-7"><a href="discussion.html#cb35-7" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">487</span>,</span>
<span id="cb35-8"><a href="discussion.html#cb35-8" tabindex="-1"></a>  <span class="at">verbose =</span> T</span>
<span id="cb35-9"><a href="discussion.html#cb35-9" tabindex="-1"></a>)</span></code></pre></div>
<ol start="3" style="list-style-type: lower-roman">
<li>multi-ROC analysis:</li>
</ol>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="discussion.html#cb36-1" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb36-2"><a href="discussion.html#cb36-2" tabindex="-1"></a>mvt_auc <span class="ot">&lt;-</span> mvt<span class="sc">$</span>multiAUC</span>
<span id="cb36-3"><a href="discussion.html#cb36-3" tabindex="-1"></a>mvt_auc_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb36-4"><a href="discussion.html#cb36-4" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(mvt_auc)){ <span class="co"># i=1</span></span>
<span id="cb36-5"><a href="discussion.html#cb36-5" tabindex="-1"></a>  df.i <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb36-6"><a href="discussion.html#cb36-6" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">as.integer</span>(<span class="fu">Fastextra</span>(<span class="fu">names</span>(mvt_auc)[i], <span class="st">&#39;=&#39;</span>, <span class="dv">2</span>)),</span>
<span id="cb36-7"><a href="discussion.html#cb36-7" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">as.numeric</span>(mvt_auc[[i]]<span class="sc">$</span>auc),</span>
<span id="cb36-8"><a href="discussion.html#cb36-8" tabindex="-1"></a>    <span class="at">stringsAsFactors =</span> F</span>
<span id="cb36-9"><a href="discussion.html#cb36-9" tabindex="-1"></a>  )</span>
<span id="cb36-10"><a href="discussion.html#cb36-10" tabindex="-1"></a>  mvt_auc_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(mvt_auc_df, df.i)</span>
<span id="cb36-11"><a href="discussion.html#cb36-11" tabindex="-1"></a>}</span>
<span id="cb36-12"><a href="discussion.html#cb36-12" tabindex="-1"></a>  </span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="discussion.html#cb37-1" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb37-2"><a href="discussion.html#cb37-2" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(mvt_auc_df, <span class="fu">aes</span>(x,y)) <span class="sc">+</span></span>
<span id="cb37-3"><a href="discussion.html#cb37-3" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb37-4"><a href="discussion.html#cb37-4" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">12</span>)) <span class="sc">+</span> </span>
<span id="cb37-5"><a href="discussion.html#cb37-5" tabindex="-1"></a>  <span class="fu">stat_smooth</span>(<span class="at">formula =</span> y <span class="sc">~</span> x,<span class="at">method =</span> <span class="st">&#39;glm&#39;</span>) <span class="sc">+</span></span>
<span id="cb37-6"><a href="discussion.html#cb37-6" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;No. of missing value&#39;</span>, </span>
<span id="cb37-7"><a href="discussion.html#cb37-7" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Relative AUC in multi-ROC analysis&#39;</span>) <span class="sc">+</span></span>
<span id="cb37-8"><a href="discussion.html#cb37-8" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb37-9"><a href="discussion.html#cb37-9" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb37-10"><a href="discussion.html#cb37-10" tabindex="-1"></a>    <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)</span>
<span id="cb37-11"><a href="discussion.html#cb37-11" tabindex="-1"></a>  )</span>
<span id="cb37-12"><a href="discussion.html#cb37-12" tabindex="-1"></a><span class="fu">print</span>(p2)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:mvi03"></span>
<img src="Discussion_files/figure-html/mvi03-1.png" alt="The association between the number of missing value and subtype identification performance." width="60%" />
<p class="caption">
Figure 2.3: The association between the number of missing value and subtype identification performance.
</p>
</div>
<p>As shown in Figure <a href="discussion.html#fig:mvi03">2.3</a>, there is a linear negative correlation between the number of missing values (missing rate ranges from 6.25% to 37.5%) and the subtype identification performance of <strong>PADi</strong> model. One of the reasons might be that PIAM/PIDG were small GEPs, so little gene loss might significantly impact the performance of <strong>PADi</strong>. By the way, there is no missing value in PIAM/PIDG of the ‚ÄòKim2018‚Äô cohort, an external validation cohort for ICIs therapy response prediction via <strong>PADi</strong>. Nonetheless, we still used the <strong>zero strategy</strong> during subtype identification of <strong>PADi</strong> if any missing values exist, because randomization might make the result unstable, which is not suitable for clinical decision.
<!--
Nonetheless, we still used **zero strategy** during subtype identification of **PADi** if any missing value exist, because randomization might make the result unstable, which is not suitable for clinical decision. 
Moreover, we explored **if rescue of missing value would help improving the performance of PADi**. 
--></p>
<p>In conclusion, the zero or ‚Äúquantile‚Äù strategy could be applied for MVI before <strong>GSClassifier</strong> model training. However, missing values should be avoided as possible in subtype identification for missing values could damage the performance of <strong>GSClassifier</strong> models. Nonetheless, due to low-input GEPs used in <strong>PADi</strong> model (No.¬†of Gene=32), it‚Äôs easy to avoid missing value in clinical practice.</p>
</div>
<div id="batch-effect" class="section level2 hasAnchor" number="2.4">
<h2><span class="header-section-number">2.4</span> Batch effect<a href="discussion.html#batch-effect" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<!--

gene-pair batch effect in Google Scholar

relative expression orderings (REO)

single-sample enrichment analysis 

Batch effect control & gene number

-->
<p><strong>TSP</strong> was widely applied to control batch effects in transcriptomic data <span class="citation">[<a href="#ref-RN369">16</a>‚Äì<a href="#ref-RN365">23</a>]</span>. Still, we tested whether <strong>TSP</strong> is a robust method for batch effect control in real-world data. As demonstrated in Figure <a href="discussion.html#fig:be01">2.4</a>, the obvious batch effects across gastric cancer datasets were significantly reduced after <strong>TSP</strong> normalization.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:be01"></span>
<img src="fig/bactch-effect-01.png" alt="Batch effects across gastric cancer cohorts. All gene pairs were used because subtype vectors were not specified. Top: Raw expression of all genes across samples. Middle: Raw expression of PIAM and PIDG across samples. Bottom: TSP of PIAM and PIDG across samples." width="90%" />
<p class="caption">
Figure 2.4: Batch effects across gastric cancer cohorts. All gene pairs were used because subtype vectors were not specified. Top: Raw expression of all genes across samples. Middle: Raw expression of PIAM and PIDG across samples. Bottom: TSP of PIAM and PIDG across samples.
</p>
</div>
<p>To confirm the association between <strong>gene counts</strong> in modeling and batch effect control via <strong>TSP</strong> normalization, we selected random genes with counts ranging from 4 to 80 for TSP matrix establishment. As shown in Figure <a href="discussion.html#fig:be02">2.5</a>, <strong>TSP</strong> normalization works greatly in different gene counts for batch effect control compared with the raw expression matrix.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:be02"></span>
<img src="fig/bactch-effect-02.png" alt="Batch effects of random genes across gastric cancer cohorts.  All gene pairs were used because subtype vectors were not specified. Gene counts 4, 8, 20, 40, and 80 were detected. Data of set difference were not available because only one gene set were applied." width="90%" />
<p class="caption">
Figure 2.5: Batch effects of random genes across gastric cancer cohorts. All gene pairs were used because subtype vectors were not specified. Gene counts 4, 8, 20, 40, and 80 were detected. Data of set difference were not available because only one gene set were applied.
</p>
</div>
<!--
Batch effect reduction of TSP could be explained in three parts:

+ **Binned expression**. Regardless of what raw distribution of a numeric vector is, it would be converted into a binned vector with exactly the same distribution under a fixed quantile vector.

+ **Pair/set difference**.
-->
</div>
<div id="subtype-number" class="section level2 hasAnchor" number="2.5">
<h2><span class="header-section-number">2.5</span> Subtype number<a href="discussion.html#subtype-number" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For PAD subtypes, it‚Äôs easy to determine the subtype number as 4. First, PAD subtypes are identified via 2 simple GEPs. The binary status (high/low expression) of 2 GEPs consists of a 2√ó2 matrix and hence the subtype number is exactly 4. Second, four PAD subtypes displayed different genetic/epigenetic alterations and clinical features (survival and ICI response), indicating that it‚Äôs biologically meaningful to distinguish gastric cancer into 4 immune subtypes. Third, clinicians are familiar with four subtypes, for the subtype number of the classical TNM stage in clinical oncology is 4 (Stage I to IV).</p>
<p>Also, there‚Äôs another more simple situation‚Äîdetermining the subtype number with only one GEP, where 2 (high/low) or 3 (low/moderate/high) deserve to be tried. However, with the number of GEPs increasing, the situation would become more complex. Regrettably, GSClassifier can not help determine what subtype number should be used; GSClassifier only promises a robust model no matter how many GEPs or subtypes. <strong>There‚Äôs no gold standard for subtype number selection, which is more like art instead of math</strong>.</p>
<p>Nevertheless, there‚Äôre several suggestions we can follow for best practice. First, the R package ‚Äú<strong>ConsensusClusterPlus</strong>‚Äù <span class="citation">[<a href="#ref-RN421">24</a>‚Äì<a href="#ref-RN424">27</a>]</span> can help figure out consensus clustering for transcriptomic data, which provides visualization (consensus matrices, consensus cumulative distribution function plot, delta area plot, tracking plot, and so on) for clustering quality control. Second, <strong>paying more attention to the biological problem usually gives extra or even crucial clues for the subtype number decision</strong>.</p>
<!--

## Hyperparameters

+ modelData(): Prop, seed

+ fitEnsembleModel():

  - n=100
  - sampSize=0.7
  - sampSeed = 2020
  - na.fill.seed = 443
  - breakVec=c(0, 0.25, 0.5, 0.75, 1.0)
  - ptail=0.8/2
  
  - https://www.youtube.com/watch?v=f3ryHJ05h5k
  
  - https://xgboost.readthedocs.io/en/latest/parameter.html
  - Parameters for Tree Booster
  - params = list(max_depth = 10,
                  eta = 0.5,
                  nrounds = 100,
                  nthread = 10,
                  nfold=5)
                  
  ÊâÄ‰ª•ÔºåÂú®Êó∂Èó¥ÈùûÂ∏∏ÂÆùË¥µÁöÑÊÉÖÂÜµ‰∏ãÔºå‰Ω†ÂøÖÈ°ªÂ∫îÁî®Ëá™ Â∑±ÁöÑÂà§Êñ≠ÂäõÔºåÂπ∂‰ΩøÁî®Â∞èÊï∞ÊçÆÊ†∑Êú¨Êù•ÊâæÂá∫ÂêàÈÄÇÁöÑË∞É‰ºòÂèÇÊï∞ÔºåÂê¶ÂàôÁ°¨ÁõòÁ©∫Èó¥ÂèØËÉΩ‰∏çË∂≥
  nrounds ÔºöÊúÄÂ§ßËø≠‰ª£Ê¨°Êï∞ÔºàÊúÄÁªàÊ®°Âûã‰∏≠Ê†ëÁöÑÊï∞ÈáèÔºâ„ÄÇ
  colsample_bytree ÔºöÂª∫Á´ãÊ†ëÊó∂ÈöèÊú∫ÊäΩÂèñÁöÑÁâπÂæÅÊï∞ÈáèÔºåÁî®‰∏Ä‰∏™ÊØîÁéáË°®Á§∫ÔºåÈªòËÆ§ÂÄº‰∏∫1Ôºà‰Ωø
Áî®100%ÁöÑÁâπÂæÅÔºâ„ÄÇ
  min_child_weight ÔºöÂØπÊ†ëËøõË°åÊèêÂçáÊó∂‰ΩøÁî®ÁöÑÊúÄÂ∞èÊùÉÈáçÔºåÈªòËÆ§‰∏∫1„ÄÇ
  eta ÔºöÂ≠¶‰π†ÁéáÔºåÊØèÊ£µÊ†ëÂú®ÊúÄÁªàËß£‰∏≠ÁöÑË¥°ÁåÆÔºåÈªòËÆ§‰∏∫0.3„ÄÇ
  gamma ÔºöÂú®Ê†ë‰∏≠Êñ∞Â¢û‰∏Ä‰∏™Âè∂Â≠êÂàÜÂå∫Êó∂ÊâÄÈúÄÁöÑÊúÄÂ∞èÂáèÊçü„ÄÇ
  subsample ÔºöÂ≠êÊ†∑Êú¨Êï∞ÊçÆÂç†Êï¥‰∏™ËßÇÊµãÁöÑÊØî‰æãÔºåÈªòËÆ§ÂÄº‰∏∫1Ôºà100%Ôºâ„ÄÇ
  max_depth ÔºöÂçï‰∏™Ê†ëÁöÑÊúÄÂ§ßÊ∑±Â∫¶„ÄÇ
-->

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body">
<div id="ref-RN387" class="csl-entry">
8. Troyanskaya O, Cantor M, Sherlock G, et al. <a href="https://doi.org/10.1093/bioinformatics/17.6.520">Missing value estimation methods for DNA microarrays</a>. Bioinformatics 2001; 17:520‚Äì5
</div>
<div id="ref-RN382" class="csl-entry">
10. Zhu X, Wang J, Sun B, et al. <a href="https://doi.org/10.1186/s12859-021-04109-4">An efficient ensemble method for missing value imputation in microarray gene expression data</a>. BMC Bioinformatics 2021; 22:188
</div>
<div id="ref-RN392" class="csl-entry">
11. Lin W-C, Tsai C-F. Missing value imputation: A review and analysis of the literature (2006‚Äì2017). 2020; 53:1487‚Äì1509
</div>
<div id="ref-RN386" class="csl-entry">
12. Hasan MK, Alam MA, Roy S, et al. Missing value imputation affects the performance of machine learning: A review and analysis of the literature (2010‚Äì2021). Informatics in Medicine Unlocked 2021; 27:100799
</div>
<div id="ref-RN384" class="csl-entry">
13. Wang A, Yang J, An N. Regularized sparse modelling for microarray missing value estimation. 2021; 9:16899‚Äì16913
</div>
<div id="ref-xgboost" class="csl-entry">
14. Chen T, He T, Benesty M, et al. <a href="https://CRAN.R-project.org/package=xgboost">Xgboost: Extreme gradient boosting</a>. 2022;
</div>
<div id="ref-pROC" class="csl-entry">
15. Robin X, Turck N, Hainard A, et al. pROC: An open-source package for r and s+ to analyze and compare ROC curves. BMC Bioinformatics 2011; 12:77
</div>
<div id="ref-RN369" class="csl-entry">
16. Zhu S, Kong W, Zhu J, et al. <a href="https://doi.org/10.1093/bib/bbac344">The genetic algorithm-aided three-stage ensemble learning method identified a robust survival risk score in patients with glioma</a>. Brief Bioinform 2022;
</div>
<div id="ref-RN365" class="csl-entry">
23. Zheng H, Xie J, Song K, et al. <a href="https://doi.org/10.1186/s13287-022-02803-5">StemSC: A cross-dataset human stemness index for single-cell samples</a>. Stem Cell Res Ther 2022; 13:115
</div>
<div id="ref-RN421" class="csl-entry">
24. Monti S, Tamayo P, Mesirov J, et al. Consensus clustering: A resampling-based method for class discovery and visualization of gene expression microarray data. 2003; 52:91‚Äì118
</div>
<div id="ref-RN424" class="csl-entry">
27. Verhaak RG, Hoadley KA, Purdom E, et al. Integrated genomic analysis identifies clinically relevant subtypes of glioblastoma characterized by abnormalities in PDGFRA, IDH1, EGFR, and NF1. 2010; 17:98‚Äì110
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="the-principle-of-gsclassifier.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="quick-start.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["GSClassifier.principle.pdf", "GSClassifier.principle.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
