<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>The Principle of R Package GSClassifier</title>
  <meta name="description" content="The Principle of GSClassifier" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="The Principle of R Package GSClassifier" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The Principle of GSClassifier" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="The Principle of R Package GSClassifier" />
  
  <meta name="twitter:description" content="The Principle of GSClassifier" />
  

<meta name="author" content="Weibin Huang" />


<meta name="date" content="2023-08-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="the-principle-of-gsclassifier.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">GSClassifier</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about"><i class="fa fa-check"></i>üí¶About</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>üíøLicense</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i>üëçInstallation</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#mirror"><i class="fa fa-check"></i>üëÄMirror</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#change-log"><i class="fa fa-check"></i>üìöChange log</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#todo"><i class="fa fa-check"></i>üìÜTODO</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#other-projects"><i class="fa fa-check"></i>üå¥Other Projects</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html"><i class="fa fa-check"></i><b>1</b> The Principle of GSClassifier</a>
<ul>
<li class="chapter" data-level="1.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#packages"><i class="fa fa-check"></i><b>1.1</b> Packages</a></li>
<li class="chapter" data-level="1.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#flowchart"><i class="fa fa-check"></i><b>1.2</b> Flowchart</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#data-processing"><i class="fa fa-check"></i><b>1.2.1</b> Data Processing</a></li>
<li class="chapter" data-level="1.2.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#model-establishment-and-validation"><i class="fa fa-check"></i><b>1.2.2</b> Model Establishment and Validation</a></li>
<li class="chapter" data-level="1.2.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#model-application"><i class="fa fa-check"></i><b>1.2.3</b> Model Application</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#topicTSP"><i class="fa fa-check"></i><b>1.3</b> Top scoring pairs</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#tsp-intro"><i class="fa fa-check"></i><b>1.3.1</b> Introduction</a></li>
<li class="chapter" data-level="1.3.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#simulated-dataset"><i class="fa fa-check"></i><b>1.3.2</b> Simulated Dataset</a></li>
<li class="chapter" data-level="1.3.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#binned-expression"><i class="fa fa-check"></i><b>1.3.3</b> Binned expression</a></li>
<li class="chapter" data-level="1.3.4" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#pair-difference"><i class="fa fa-check"></i><b>1.3.4</b> Pair difference</a></li>
<li class="chapter" data-level="1.3.5" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#set-difference"><i class="fa fa-check"></i><b>1.3.5</b> Set difference</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="discussion.html"><a href="discussion.html"><i class="fa fa-check"></i><b>2</b> Discussion</a>
<ul>
<li class="chapter" data-level="2.1" data-path="discussion.html"><a href="discussion.html#packages-1"><i class="fa fa-check"></i><b>2.1</b> Packages</a></li>
<li class="chapter" data-level="2.2" data-path="discussion.html"><a href="discussion.html#subtype-vector"><i class="fa fa-check"></i><b>2.2</b> Subtype Vector</a></li>
<li class="chapter" data-level="2.3" data-path="discussion.html"><a href="discussion.html#missing-value-imputation-mvi"><i class="fa fa-check"></i><b>2.3</b> Missing value imputation (MVI)</a></li>
<li class="chapter" data-level="2.4" data-path="discussion.html"><a href="discussion.html#batch-effect"><i class="fa fa-check"></i><b>2.4</b> Batch effect</a></li>
<li class="chapter" data-level="2.5" data-path="discussion.html"><a href="discussion.html#subtype-number"><i class="fa fa-check"></i><b>2.5</b> Subtype number</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>3</b> Quick start</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quick-start.html"><a href="quick-start.html#about-1"><i class="fa fa-check"></i><b>3.1</b> About</a></li>
<li class="chapter" data-level="3.2" data-path="quick-start.html"><a href="quick-start.html#package"><i class="fa fa-check"></i><b>3.2</b> Package</a></li>
<li class="chapter" data-level="3.3" data-path="quick-start.html"><a href="quick-start.html#data"><i class="fa fa-check"></i><b>3.3</b> Data</a></li>
<li class="chapter" data-level="3.4" data-path="quick-start.html"><a href="quick-start.html#pad"><i class="fa fa-check"></i><b>3.4</b> PAD</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="quick-start.html"><a href="quick-start.html#the-work-flow-of-pad-exploration"><i class="fa fa-check"></i><b>3.4.1</b> The work flow of PAD exploration</a></li>
<li class="chapter" data-level="3.4.2" data-path="quick-start.html"><a href="quick-start.html#preparation-of-the-test-data"><i class="fa fa-check"></i><b>3.4.2</b> Preparation of the test data</a></li>
<li class="chapter" data-level="3.4.3" data-path="quick-start.html"><a href="quick-start.html#unsupervised-clustering"><i class="fa fa-check"></i><b>3.4.3</b> Unsupervised clustering</a></li>
<li class="chapter" data-level="3.4.4" data-path="quick-start.html"><a href="quick-start.html#of-note"><i class="fa fa-check"></i><b>3.4.4</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="quick-start.html"><a href="quick-start.html#padi"><i class="fa fa-check"></i><b>3.5</b> PADi</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="quick-start.html"><a href="quick-start.html#preparation-of-the-test-data-1"><i class="fa fa-check"></i><b>3.5.1</b> Preparation of the test data</a></li>
<li class="chapter" data-level="3.5.2" data-path="quick-start.html"><a href="quick-start.html#use-a-specific-function-called-padi"><i class="fa fa-check"></i><b>3.5.2</b> Use a specific function called <code>PADi</code></a></li>
<li class="chapter" data-level="3.5.3" data-path="quick-start.html"><a href="quick-start.html#use-the-callensemble-function"><i class="fa fa-check"></i><b>3.5.3</b> Use the <code>callEnsemble</code> function</a></li>
<li class="chapter" data-level="3.5.4" data-path="quick-start.html"><a href="quick-start.html#parallel-strategy-for-padi"><i class="fa fa-check"></i><b>3.5.4</b> Parallel strategy for PADi</a></li>
<li class="chapter" data-level="3.5.5" data-path="quick-start.html"><a href="quick-start.html#single-sample-subtype-calling"><i class="fa fa-check"></i><b>3.5.5</b> Single sample subtype calling</a></li>
<li class="chapter" data-level="3.5.6" data-path="quick-start.html"><a href="quick-start.html#of-note-1"><i class="fa fa-check"></i><b>3.5.6</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="quick-start.html"><a href="quick-start.html#use-external-models-from-luckymodel-package"><i class="fa fa-check"></i><b>3.6</b> Use external models from luckyModel package</a></li>
<li class="chapter" data-level="3.7" data-path="quick-start.html"><a href="quick-start.html#pancancer-immune-subtypes"><i class="fa fa-check"></i><b>3.7</b> PanCancer immune subtypes</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html"><i class="fa fa-check"></i><b>4</b> Model establishment via GSClassifier</a>
<ul>
<li class="chapter" data-level="4.1" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#about-2"><i class="fa fa-check"></i><b>4.1</b> About</a></li>
<li class="chapter" data-level="4.2" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#data-preparation"><i class="fa fa-check"></i><b>4.2</b> Data preparation</a></li>
<li class="chapter" data-level="4.3" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#fitting-models"><i class="fa fa-check"></i><b>4.3</b> Fitting models</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#gsclassifier-model-training"><i class="fa fa-check"></i><b>4.3.1</b> GSClassifier model training</a></li>
<li class="chapter" data-level="4.3.2" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#scaller-for-the-best-call"><i class="fa fa-check"></i><b>4.3.2</b> Scaller for the best call</a></li>
<li class="chapter" data-level="4.3.3" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#assemble-your-model"><i class="fa fa-check"></i><b>4.3.3</b> Assemble your model</a></li>
<li class="chapter" data-level="4.3.4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#of-note-2"><i class="fa fa-check"></i><b>4.3.4</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#calling-subtypes"><i class="fa fa-check"></i><b>4.4</b> Calling subtypes</a></li>
<li class="chapter" data-level="4.5" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#topicSubmodel"><i class="fa fa-check"></i><b>4.5</b> Number of SubModel</a></li>
<li class="chapter" data-level="4.6" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#padiPara"><i class="fa fa-check"></i><b>4.6</b> Parameters for PADi training</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html"><i class="fa fa-check"></i><b>5</b> Suggestions for GSClassifier model developers</a>
<ul>
<li class="chapter" data-level="5.1" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#about-3"><i class="fa fa-check"></i><b>5.1</b> About</a></li>
<li class="chapter" data-level="5.2" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#available-models"><i class="fa fa-check"></i><b>5.2</b> Available models</a></li>
<li class="chapter" data-level="5.3" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#components-of-a-gsclassifier-model"><i class="fa fa-check"></i><b>5.3</b> Components of a GSClassifier model</a></li>
<li class="chapter" data-level="5.4" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#submit-models-to-luckymodel-package"><i class="fa fa-check"></i><b>5.4</b> Submit models to luckyModel package</a></li>
<li class="chapter" data-level="5.5" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#repeatablility-of-models"><i class="fa fa-check"></i><b>5.5</b> Repeatablility of models</a></li>
<li class="chapter" data-level="5.6" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#gene-annotation"><i class="fa fa-check"></i><b>5.6</b> Gene Annotation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://blognas.hwb0307.com/" target="blank">Welcome to My Blog!</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Principle of R Package GSClassifier</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">The Principle of R Package GSClassifier</h1>
<p class="author"><em>Weibin Huang</em></p>
<p class="date"><em>2023-08-04</em></p>
</div>
<div id="welcome" class="section level1 unnumbered hasAnchor">
<h1>Welcome<a href="index.html#welcome" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<!--

[GSClassifier](https://github.com/huangwb8/GSClassifier) is an R package for modeling and identification of Gene Expression Profiles (GEPs) subtypes. The detail of **GSClassifier** package usage had been demonstrated in [Github WiKi](https://github.com/huangwb8/GSClassifier/wiki). Here, we propose to introduce the principle of GSClassifier, including flowchart, **top scoring pairs (TSP)** algorithm, and batch effect control. 

emoji: https://github.com/rstudio/blogdown/issues/171

-->
<div id="about" class="section level2 unnumbered hasAnchor">
<h2>üí¶About<a href="#about" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>The Priciple of GSClassifier</strong> is a book for users of the R package <strong>GSClassifier</strong> who want to know the most details. If you‚Äôre looking for the PDF edition, you can find it <a href="https://github.com/huangwb8/GSClassifier.principle/blob/master/docs/GSClassifier.principle.pdf">here</a>.</p>
<p><a href="https://github.com/huangwb8/GSClassifier"><strong>GSClassifier</strong></a> is an R-based comprehensive classification tool for subtypes modeling and personalized calling based on pure transcriptomics. It could be used for precision medicine, such as cancer diagnosis. The inspiration for <strong>GSClassifier</strong> comes from <a href="https://github.com/CRI-iAtlas/ImmuneSubtypeClassifier">ImmuneSubtypeClassifier</a>, an R package for classification of PanCancer immune subtypes based on the work of Gibbs et al <span class="citation">[<a href="#ref-RN160">1</a>,<a href="#ref-RN315">2</a>]</span>.</p>
<p>Lots of surprising features in <strong>GSClassifier</strong> are as follows:</p>
<ul>
<li><p>Optimized for just <code>one sample</code></p></li>
<li><p>Available for modeling and calling of brand-new <code>GEPs-based subtypes</code> in any diseases (cancers)</p></li>
<li><p>No limitation of the number of <code>gene signatures</code>(‚â•1) or <code>subtypes</code>(‚â•2)</p></li>
<li><p><code>Insensitive normalization</code> due to the use of the individual <code>gene rank matrix</code></p></li>
<li><p>More ensemble and repeatable modeling process</p></li>
<li><p>More optimizations in the parallel computing</p></li>
<li><p>New useful functions as supplements</p></li>
</ul>
<p><strong>ATTENTION!</strong> In the future, there might be third-party contributors in <code>GSClassifier</code> platform, with some useful models for specific usages. If you use models provided by these people, <strong>you had better know more details as possible</strong>, including <strong>designs, data sources, destinations, training scripts, and limitations</strong> of models, especially those from studies under peer review.</p>
</div>
<div id="license" class="section level2 unnumbered hasAnchor">
<h2>üíøLicense<a href="#license" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><p><strong>GSClassifier</strong> is released under the Apache-2.0 license. See <a href="https://github.com/huangwb8/GSClassifier/blob/master/license.txt">LICENSE</a> for details.</p></li>
<li><p>The technical documentation, as a whole, is licensed under a <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-
NonCommercial-ShareAlike 4.0 International License</a>. The code contained in this book is simultaneously available under the <a href="https://opensource.org/licenses/MIT">MIT license</a>; this means that you are free to use it in your packages, as long as you cite the source.</p></li>
</ul>
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a></p>
</div>
<div id="installation" class="section level2 unnumbered hasAnchor">
<h2>üëçInstallation<a href="#installation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>RStudio/Posit</strong> is one of the best Integrated Development Environments (IDE) in R programming. If you‚Äôre struggling in R-GUI, it is recommended to turn to <a href="https://www.rstudio.com/">RStudio/Posit</a>.</p>
<p>For installation of <strong>GSClassifier</strong>, please run these commands in an R environment:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="index.html#cb1-1" tabindex="-1"></a><span class="co"># Install &quot;devtools&quot; package</span></span>
<span id="cb1-2"><a href="index.html#cb1-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-3"><a href="index.html#cb1-3" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-4"><a href="index.html#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="index.html#cb1-5" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb1-6"><a href="index.html#cb1-6" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;luckyBase&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-7"><a href="index.html#cb1-7" tabindex="-1"></a>    devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/luckyBase&quot;</span>)</span>
<span id="cb1-8"><a href="index.html#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="index.html#cb1-9" tabindex="-1"></a><span class="co"># Install the &quot;GSClassifier&quot; package</span></span>
<span id="cb1-10"><a href="index.html#cb1-10" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;GSClassifier&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-11"><a href="index.html#cb1-11" tabindex="-1"></a>    devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/GSClassifier&quot;</span>)</span></code></pre></div>
<p>In the future, a stable <code>GSClassifier</code> version might be sent to <a href="https://cran.r-project.org/"><code>CRAN</code></a>. Still beta.</p>
</div>
<div id="mirror" class="section level2 unnumbered hasAnchor">
<h2>üëÄ<a href="https://gitee.com/huangwb8/GSClassifier">Mirror</a><a href="#mirror" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For some special countries or regions, users could also try:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="index.html#cb2-1" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb2-2"><a href="index.html#cb2-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;https://gitee.com/huangwb8/luckyBase/repository/archive/Primary?format=tar.gz&quot;</span>, <span class="at">repos=</span><span class="cn">NULL</span>, <span class="at">method=</span><span class="st">&quot;libcurl&quot;</span>)</span>
<span id="cb2-3"><a href="index.html#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="index.html#cb2-4" tabindex="-1"></a><span class="co"># Install the &quot;GSClassifier&quot; package</span></span>
<span id="cb2-5"><a href="index.html#cb2-5" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;https://gitee.com/huangwb8/GSClassifier/repository/archive/Primary?format=tar.gz&quot;</span>, <span class="at">repos=</span><span class="cn">NULL</span>, <span class="at">method=</span><span class="st">&quot;libcurl&quot;</span>)</span></code></pre></div>
</div>
<div id="change-log" class="section level2 unnumbered hasAnchor">
<h2>üìöChange log<a href="#change-log" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><p>Version 0.1.27</p>
<ul>
<li><p>Enhaned <code>geneMatch</code> function</p></li>
<li><p>Repair some bugs</p></li>
</ul></li>
<li><p>Version 0.1.9</p>
<ul>
<li><p>Optimize function verbose</p></li>
<li><p>Optimize for a routine scenario: one gene set and two subtypes</p></li>
<li><p>Optimize the strategy of automatic parameters selection for modeling training with R package <code>caret</code></p></li>
<li><p>Interact with external models from the <a href="https://github.com/huangwb8/luckyModel">luckyModel</a> package</p></li>
</ul></li>
<li><p>Version 0.1.8</p>
<ul>
<li><p>Primary public version of <code>GSClassifier</code></p></li>
<li><p>Apache License, Version 2.0</p></li>
<li><p>Friendly wiki-based tutorial</p></li>
<li><p>Platform for developers</p></li>
</ul></li>
</ul>
</div>
<div id="todo" class="section level2 unnumbered hasAnchor">
<h2>üìÜTODO<a href="#todo" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><p>More medical fields included, such as in the pan-cancer utility</p></li>
<li><p>Advanced methods (such as artificial intelligence) for enhanced robustness</p></li>
<li><p>Unsupervised learning for de-novo classification based on intrinsic frames of omics instead of human knowledge</p></li>
<li><p>Multi-omics exploration and support</p></li>
<li><p>More friendly characteristics for developers and contributors</p></li>
<li><p>Web application for newbies to R programming</p></li>
</ul>
</div>
<div id="other-projects" class="section level2 unnumbered hasAnchor">
<h2>üå¥Other Projects<a href="#other-projects" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>You may also be interested in:</p>
<ul>
<li><p><strong>‚Äú<a href="https://github.com/huangwb8/luckyBase">luckyBase</a>‚Äù</strong> The base functions of lucky series.</p></li>
<li><p><strong>‚Äú<a href="https://github.com/huangwb8/luckyModel">luckyModel</a>‚Äù</strong> Model ensemble for third-party lucky series, such GSClassifier.</p></li>
</ul>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body">
<div id="ref-RN160" class="csl-entry">
1. Thorsson V, Gibbs DL, Brown SD, et al. <a href="https://doi.org/10.1016/j.immuni.2018.03.023">The immune landscape of cancer</a>. Immunity 2018; 48:812‚Äì830 e14
</div>
<div id="ref-RN315" class="csl-entry">
2. Gibbs DL. Robust classification of immune subtypes in cancer. 2020;
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="the-principle-of-gsclassifier.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["GSClassifier.principle.pdf", "GSClassifier.principle.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
