<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>References | The Principle of R Package GSClassifier</title>
  <meta name="description" content="The Principle of GSClassifier" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="References | The Principle of R Package GSClassifier" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The Principle of GSClassifier" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="References | The Principle of R Package GSClassifier" />
  
  <meta name="twitter:description" content="The Principle of GSClassifier" />
  

<meta name="author" content="Weibin Huang" />


<meta name="date" content="2025-05-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="suggestions-for-gsclassifier-model-developers.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">GSClassifier</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about"><i class="fa fa-check"></i>üí¶About</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>üíøLicense</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i>üëçInstallation</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#mirror"><i class="fa fa-check"></i>üëÄMirror</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#change-log"><i class="fa fa-check"></i>üìöChange log</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#todo"><i class="fa fa-check"></i>üìÜTODO</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#other-projects"><i class="fa fa-check"></i>üå¥Other Projects</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html"><i class="fa fa-check"></i><b>1</b> The Principle of GSClassifier</a>
<ul>
<li class="chapter" data-level="1.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#packages"><i class="fa fa-check"></i><b>1.1</b> Packages</a></li>
<li class="chapter" data-level="1.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#flowchart"><i class="fa fa-check"></i><b>1.2</b> Flowchart</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#data-processing"><i class="fa fa-check"></i><b>1.2.1</b> Data Processing</a></li>
<li class="chapter" data-level="1.2.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#model-establishment-and-validation"><i class="fa fa-check"></i><b>1.2.2</b> Model Establishment and Validation</a></li>
<li class="chapter" data-level="1.2.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#model-application"><i class="fa fa-check"></i><b>1.2.3</b> Model Application</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#topicTSP"><i class="fa fa-check"></i><b>1.3</b> Top scoring pairs</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#tsp-intro"><i class="fa fa-check"></i><b>1.3.1</b> Introduction</a></li>
<li class="chapter" data-level="1.3.2" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#simulated-dataset"><i class="fa fa-check"></i><b>1.3.2</b> Simulated Dataset</a></li>
<li class="chapter" data-level="1.3.3" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#binned-expression"><i class="fa fa-check"></i><b>1.3.3</b> Binned expression</a></li>
<li class="chapter" data-level="1.3.4" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#pair-difference"><i class="fa fa-check"></i><b>1.3.4</b> Pair difference</a></li>
<li class="chapter" data-level="1.3.5" data-path="the-principle-of-gsclassifier.html"><a href="the-principle-of-gsclassifier.html#set-difference"><i class="fa fa-check"></i><b>1.3.5</b> Set difference</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="discussion.html"><a href="discussion.html"><i class="fa fa-check"></i><b>2</b> Discussion</a>
<ul>
<li class="chapter" data-level="2.1" data-path="discussion.html"><a href="discussion.html#packages-1"><i class="fa fa-check"></i><b>2.1</b> Packages</a></li>
<li class="chapter" data-level="2.2" data-path="discussion.html"><a href="discussion.html#subtype-vector"><i class="fa fa-check"></i><b>2.2</b> Subtype Vector</a></li>
<li class="chapter" data-level="2.3" data-path="discussion.html"><a href="discussion.html#missing-value-imputation-mvi"><i class="fa fa-check"></i><b>2.3</b> Missing value imputation (MVI)</a></li>
<li class="chapter" data-level="2.4" data-path="discussion.html"><a href="discussion.html#batch-effect"><i class="fa fa-check"></i><b>2.4</b> Batch effect</a></li>
<li class="chapter" data-level="2.5" data-path="discussion.html"><a href="discussion.html#subtype-number"><i class="fa fa-check"></i><b>2.5</b> Subtype number</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>3</b> Quick start</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quick-start.html"><a href="quick-start.html#about-1"><i class="fa fa-check"></i><b>3.1</b> About</a></li>
<li class="chapter" data-level="3.2" data-path="quick-start.html"><a href="quick-start.html#package"><i class="fa fa-check"></i><b>3.2</b> Package</a></li>
<li class="chapter" data-level="3.3" data-path="quick-start.html"><a href="quick-start.html#data"><i class="fa fa-check"></i><b>3.3</b> Data</a></li>
<li class="chapter" data-level="3.4" data-path="quick-start.html"><a href="quick-start.html#pad"><i class="fa fa-check"></i><b>3.4</b> PAD</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="quick-start.html"><a href="quick-start.html#the-work-flow-of-pad-exploration"><i class="fa fa-check"></i><b>3.4.1</b> The work flow of PAD exploration</a></li>
<li class="chapter" data-level="3.4.2" data-path="quick-start.html"><a href="quick-start.html#preparation-of-the-test-data"><i class="fa fa-check"></i><b>3.4.2</b> Preparation of the test data</a></li>
<li class="chapter" data-level="3.4.3" data-path="quick-start.html"><a href="quick-start.html#unsupervised-clustering"><i class="fa fa-check"></i><b>3.4.3</b> Unsupervised clustering</a></li>
<li class="chapter" data-level="3.4.4" data-path="quick-start.html"><a href="quick-start.html#of-note"><i class="fa fa-check"></i><b>3.4.4</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="quick-start.html"><a href="quick-start.html#padi"><i class="fa fa-check"></i><b>3.5</b> PADi</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="quick-start.html"><a href="quick-start.html#preparation-of-the-test-data-1"><i class="fa fa-check"></i><b>3.5.1</b> Preparation of the test data</a></li>
<li class="chapter" data-level="3.5.2" data-path="quick-start.html"><a href="quick-start.html#use-a-specific-function-called-padi"><i class="fa fa-check"></i><b>3.5.2</b> Use a specific function called <code>PADi</code></a></li>
<li class="chapter" data-level="3.5.3" data-path="quick-start.html"><a href="quick-start.html#use-the-callensemble-function"><i class="fa fa-check"></i><b>3.5.3</b> Use the <code>callEnsemble</code> function</a></li>
<li class="chapter" data-level="3.5.4" data-path="quick-start.html"><a href="quick-start.html#parallel-strategy-for-padi"><i class="fa fa-check"></i><b>3.5.4</b> Parallel strategy for PADi</a></li>
<li class="chapter" data-level="3.5.5" data-path="quick-start.html"><a href="quick-start.html#single-sample-subtype-calling"><i class="fa fa-check"></i><b>3.5.5</b> Single sample subtype calling</a></li>
<li class="chapter" data-level="3.5.6" data-path="quick-start.html"><a href="quick-start.html#of-note-1"><i class="fa fa-check"></i><b>3.5.6</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="quick-start.html"><a href="quick-start.html#use-external-models-from-luckymodel-package"><i class="fa fa-check"></i><b>3.6</b> Use external models from luckyModel package</a></li>
<li class="chapter" data-level="3.7" data-path="quick-start.html"><a href="quick-start.html#pancancer-immune-subtypes"><i class="fa fa-check"></i><b>3.7</b> PanCancer immune subtypes</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html"><i class="fa fa-check"></i><b>4</b> Model establishment via GSClassifier</a>
<ul>
<li class="chapter" data-level="4.1" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#about-2"><i class="fa fa-check"></i><b>4.1</b> About</a></li>
<li class="chapter" data-level="4.2" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#data-preparation"><i class="fa fa-check"></i><b>4.2</b> Data preparation</a></li>
<li class="chapter" data-level="4.3" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#fitting-models"><i class="fa fa-check"></i><b>4.3</b> Fitting models</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#gsclassifier-model-training"><i class="fa fa-check"></i><b>4.3.1</b> GSClassifier model training</a></li>
<li class="chapter" data-level="4.3.2" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#scaller-for-the-best-call"><i class="fa fa-check"></i><b>4.3.2</b> Scaller for the best call</a></li>
<li class="chapter" data-level="4.3.3" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#assemble-your-model"><i class="fa fa-check"></i><b>4.3.3</b> Assemble your model</a></li>
<li class="chapter" data-level="4.3.4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#of-note-2"><i class="fa fa-check"></i><b>4.3.4</b> Of note</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#calling-subtypes"><i class="fa fa-check"></i><b>4.4</b> Calling subtypes</a></li>
<li class="chapter" data-level="4.5" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#topicSubmodel"><i class="fa fa-check"></i><b>4.5</b> Number of SubModel</a></li>
<li class="chapter" data-level="4.6" data-path="model-establishment-via-gsclassifier.html"><a href="model-establishment-via-gsclassifier.html#padiPara"><i class="fa fa-check"></i><b>4.6</b> Parameters for PADi training</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html"><i class="fa fa-check"></i><b>5</b> Suggestions for GSClassifier model developers</a>
<ul>
<li class="chapter" data-level="5.1" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#about-3"><i class="fa fa-check"></i><b>5.1</b> About</a></li>
<li class="chapter" data-level="5.2" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#available-models"><i class="fa fa-check"></i><b>5.2</b> Available models</a></li>
<li class="chapter" data-level="5.3" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#components-of-a-gsclassifier-model"><i class="fa fa-check"></i><b>5.3</b> Components of a GSClassifier model</a></li>
<li class="chapter" data-level="5.4" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#submit-models-to-luckymodel-package"><i class="fa fa-check"></i><b>5.4</b> Submit models to luckyModel package</a></li>
<li class="chapter" data-level="5.5" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#repeatablility-of-models"><i class="fa fa-check"></i><b>5.5</b> Repeatablility of models</a></li>
<li class="chapter" data-level="5.6" data-path="suggestions-for-gsclassifier-model-developers.html"><a href="suggestions-for-gsclassifier-model-developers.html#gene-annotation"><i class="fa fa-check"></i><b>5.6</b> Gene Annotation</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://blognas.hwb0307.com/" target="blank">Welcome to My Blog!</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Principle of R Package GSClassifier</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1 unnumbered hasAnchor">
<h1>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h1>

<div id="refs" class="references csl-bib-body">
<div class="csl-entry">
1. Thorsson V, Gibbs DL, Brown SD, et al. <a href="https://doi.org/10.1016/j.immuni.2018.03.023">The immune landscape of cancer</a>. Immunity 2018; 48:812‚Äì830 e14
</div>
<div class="csl-entry">
2. Gibbs DL. Robust classification of immune subtypes in cancer. 2020;
</div>
<div class="csl-entry">
3. Chen T, He T, Benesty M, et al. Xgboost: Extreme gradient boosting. 2015; 1:1‚Äì4
</div>
<div class="csl-entry">
4. Geman D, d‚ÄôAvignon C, Naiman DQ, et al. <a href="https://doi.org/10.2202/1544-6115.1071">Classifying gene expression profiles from pairwise mRNA comparisons</a>. Stat Appl Genet Mol Biol 2004; 3:Article19
</div>
<div class="csl-entry">
5. Zhao H, Logothetis CJ, Gorlov IP. <a href="https://doi.org/10.1038/pcan.2010.9">Usefulness of the top-scoring pairs of genes for prediction of prostate cancer progression</a>. Prostate Cancer Prostatic Dis 2010; 13:252‚Äì9
</div>
<div class="csl-entry">
6. Youssef YM, White NM, Grigull J, et al. <a href="https://doi.org/10.1016/j.eururo.2011.01.004">Accurate molecular classification of kidney cancer subtypes using microRNA signature</a>. Eur Urol 2011; 59:721‚Äì30
</div>
<div class="csl-entry">
7. Auslander N, Zhang G, Lee JS, et al. <a href="https://doi.org/10.1038/s41591-018-0157-9">Robust prediction of response to immune checkpoint blockade therapy in metastatic melanoma</a>. Nat Med 2018; 24:1545‚Äì1549
</div>
<div class="csl-entry">
8. Troyanskaya O, Cantor M, Sherlock G, et al. <a href="https://doi.org/10.1093/bioinformatics/17.6.520">Missing value estimation methods for DNA microarrays</a>. Bioinformatics 2001; 17:520‚Äì5
</div>
<div class="csl-entry">
9. Arbeitman MN, Furlong EE, Imam F, et al. <a href="https://doi.org/10.1126/science.1072152">Gene expression during the life cycle of drosophila melanogaster</a>. Science 2002; 297:2270‚Äì5
</div>
<div class="csl-entry">
10. Zhu X, Wang J, Sun B, et al. <a href="https://doi.org/10.1186/s12859-021-04109-4">An efficient ensemble method for missing value imputation in microarray gene expression data</a>. BMC Bioinformatics 2021; 22:188
</div>
<div class="csl-entry">
11. Lin W-C, Tsai C-F. Missing value imputation: A review and analysis of the literature (2006‚Äì2017). 2020; 53:1487‚Äì1509
</div>
<div class="csl-entry">
12. Hasan MK, Alam MA, Roy S, et al. Missing value imputation affects the performance of machine learning: A review and analysis of the literature (2010‚Äì2021). Informatics in Medicine Unlocked 2021; 27:100799
</div>
<div class="csl-entry">
13. Wang A, Yang J, An N. Regularized sparse modelling for microarray missing value estimation. 2021; 9:16899‚Äì16913
</div>
<div class="csl-entry">
14. Chen T, He T, Benesty M, et al. <a href="https://CRAN.R-project.org/package=xgboost">Xgboost: Extreme gradient boosting</a>. 2022;
</div>
<div class="csl-entry">
15. Robin X, Turck N, Hainard A, et al. pROC: An open-source package for r and s+ to analyze and compare ROC curves. BMC Bioinformatics 2011; 12:77
</div>
<div class="csl-entry">
16. Zhu S, Kong W, Zhu J, et al. <a href="https://doi.org/10.1093/bib/bbac344">The genetic algorithm-aided three-stage ensemble learning method identified a robust survival risk score in patients with glioma</a>. Brief Bioinform 2022;
</div>
<div class="csl-entry">
17. Tong M, Zheng W, Li H, et al. <a href="https://doi.org/10.1038/oncsis.2016.51">Multi-omics landscapes of colorectal cancer subtypes discriminated by an individualized prognostic signature for 5-fluorouracil-based chemotherapy</a>. Oncogenesis 2016; 5:e242
</div>
<div class="csl-entry">
18. Qi L, Li Y, Qin Y, et al. <a href="https://doi.org/10.1038/bjc.2016.370">An individualised signature for predicting response with concordant survival benefit for lung adenocarcinoma patients receiving platinum-based chemotherapy</a>. Br J Cancer 2016; 115:1513‚Äì1519
</div>
<div class="csl-entry">
19. Huang H, Zou Y, Zhang H, et al. <a href="https://doi.org/10.1016/j.trsl.2020.02.004">A qualitative transcriptional prognostic signature for patients with stage i-II pancreatic ductal adenocarcinoma</a>. Transl Res 2020; 219:30‚Äì44
</div>
<div class="csl-entry">
20. Zheng H, Song K, Fu Y, et al. <a href="https://doi.org/10.1093/bib/bbz174">An absolute human stemness index associated with oncogenic dedifferentiation</a>. Brief Bioinform 2021; 22:2151‚Äì2160
</div>
<div class="csl-entry">
21. Kong W, He L, Zhu J, et al. <a href="https://doi.org/10.1038/s41375-022-01662-6">An immunity and pyroptosis gene-pair signature predicts overall survival in acute myeloid leukemia</a>. Leukemia 2022;
</div>
<div class="csl-entry">
22. Liu K, Geng Y, Wang L, et al. <a href="https://doi.org/10.1002/1878-0261.13279">Systematic exploration of the underlying mechanism of gemcitabine resistance in pancreatic adenocarcinoma</a>. Mol Oncol 2022; 16:3034‚Äì3051
</div>
<div class="csl-entry">
23. Zheng H, Xie J, Song K, et al. <a href="https://doi.org/10.1186/s13287-022-02803-5">StemSC: A cross-dataset human stemness index for single-cell samples</a>. Stem Cell Res Ther 2022; 13:115
</div>
<div class="csl-entry">
24. Monti S, Tamayo P, Mesirov J, et al. Consensus clustering: A resampling-based method for class discovery and visualization of gene expression microarray data. 2003; 52:91‚Äì118
</div>
<div class="csl-entry">
25. Wilkerson MD, Hayes DN. ConsensusClusterPlus: A class discovery tool with confidence assessments and item tracking. 2010; 26:1572‚Äì1573
</div>
<div class="csl-entry">
26. Hayes DN, Monti S, Parmigiani G, et al. Gene expression profiling reveals reproducible human lung adenocarcinoma subtypes in multiple independent patient cohorts. 2006; 24:5079‚Äì5090
</div>
<div class="csl-entry">
27. Verhaak RG, Hoadley KA, Purdom E, et al. Integrated genomic analysis identifies clinically relevant subtypes of glioblastoma characterized by abnormalities in PDGFRA, IDH1, EGFR, and NF1. 2010; 17:98‚Äì110
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="suggestions-for-gsclassifier-model-developers.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["GSClassifier.principle.pdf", "GSClassifier.principle.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
