<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Suggestions for GSClassifier model developers | The Principle of R Package GSClassifier</title>
  <meta name="description" content="The Principle of GSClassifier" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Suggestions for GSClassifier model developers | The Principle of R Package GSClassifier" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The Principle of GSClassifier" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Suggestions for GSClassifier model developers | The Principle of R Package GSClassifier" />
  
  <meta name="twitter:description" content="The Principle of GSClassifier" />
  

<meta name="author" content="Weibin Huang" />


<meta name="date" content="2022-09-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="model-establishment-via-gsclassifier.html"/>
<link rel="next" href="references.html"/>
<script src="libs/header-attrs-2.14/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">GSClassifier</a></li>

<li class="divider"></li>
<li><a href="index.html#welcome" id="toc-welcome">Welcome</a>
<ul>
<li><a href="index.html#license" id="toc-license">License</a></li>
<li><a href="index.html#other-projects" id="toc-other-projects">Other Projects</a></li>
</ul></li>
<li><a href="the-principle-of-gsclassifier.html#the-principle-of-gsclassifier" id="toc-the-principle-of-gsclassifier"><span class="toc-section-number">1</span> The Principle of GSClassifier</a>
<ul>
<li><a href="the-principle-of-gsclassifier.html#introduction" id="toc-introduction"><span class="toc-section-number">1.1</span> Introduction</a></li>
<li><a href="the-principle-of-gsclassifier.html#packages" id="toc-packages"><span class="toc-section-number">1.2</span> Packages</a></li>
<li><a href="the-principle-of-gsclassifier.html#flowchart" id="toc-flowchart"><span class="toc-section-number">1.3</span> Flowchart</a>
<ul>
<li><a href="the-principle-of-gsclassifier.html#data-processing" id="toc-data-processing"><span class="toc-section-number">1.3.1</span> Data Processing</a></li>
<li><a href="the-principle-of-gsclassifier.html#model-establishment-and-validation" id="toc-model-establishment-and-validation"><span class="toc-section-number">1.3.2</span> Model Establishment and Validation</a></li>
<li><a href="the-principle-of-gsclassifier.html#model-application" id="toc-model-application"><span class="toc-section-number">1.3.3</span> Model Application</a></li>
</ul></li>
<li><a href="the-principle-of-gsclassifier.html#topicTSP" id="toc-topicTSP"><span class="toc-section-number">1.4</span> Top scoring pairs</a>
<ul>
<li><a href="the-principle-of-gsclassifier.html#tsp-intro" id="toc-tsp-intro"><span class="toc-section-number">1.4.1</span> Introduction</a></li>
<li><a href="the-principle-of-gsclassifier.html#simulated-dataset" id="toc-simulated-dataset"><span class="toc-section-number">1.4.2</span> Simulated Dataset</a></li>
<li><a href="the-principle-of-gsclassifier.html#binned-expression" id="toc-binned-expression"><span class="toc-section-number">1.4.3</span> Binned expression</a></li>
<li><a href="the-principle-of-gsclassifier.html#pair-difference" id="toc-pair-difference"><span class="toc-section-number">1.4.4</span> Pair difference</a></li>
<li><a href="the-principle-of-gsclassifier.html#set-difference" id="toc-set-difference"><span class="toc-section-number">1.4.5</span> Set difference</a></li>
</ul></li>
</ul></li>
<li><a href="discussion.html#discussion" id="toc-discussion"><span class="toc-section-number">2</span> Discussion</a>
<ul>
<li><a href="discussion.html#packages-1" id="toc-packages-1"><span class="toc-section-number">2.1</span> Packages</a></li>
<li><a href="discussion.html#missing-value-imputation-mvi" id="toc-missing-value-imputation-mvi"><span class="toc-section-number">2.2</span> Missing value imputation (MVI)</a></li>
<li><a href="discussion.html#batch-effect" id="toc-batch-effect"><span class="toc-section-number">2.3</span> Batch effect</a></li>
<li><a href="discussion.html#hyperparameters" id="toc-hyperparameters"><span class="toc-section-number">2.4</span> Hyperparameters</a></li>
</ul></li>
<li><a href="quick-start.html#quick-start" id="toc-quick-start"><span class="toc-section-number">3</span> Quick start</a>
<ul>
<li><a href="quick-start.html#about" id="toc-about"><span class="toc-section-number">3.1</span> About</a></li>
<li><a href="quick-start.html#data" id="toc-data"><span class="toc-section-number">3.2</span> Data</a></li>
<li><a href="quick-start.html#pad" id="toc-pad"><span class="toc-section-number">3.3</span> PAD</a>
<ul>
<li><a href="quick-start.html#preparation-of-the-test-data" id="toc-preparation-of-the-test-data"><span class="toc-section-number">3.3.1</span> Preparation of the test data</a></li>
<li><a href="quick-start.html#unsupervised-clustering" id="toc-unsupervised-clustering"><span class="toc-section-number">3.3.2</span> Unsupervised clustering</a></li>
<li><a href="quick-start.html#of-note" id="toc-of-note"><span class="toc-section-number">3.3.3</span> Of note</a></li>
</ul></li>
<li><a href="quick-start.html#padi" id="toc-padi"><span class="toc-section-number">3.4</span> PADi</a>
<ul>
<li><a href="quick-start.html#preparation-of-the-test-data-1" id="toc-preparation-of-the-test-data-1"><span class="toc-section-number">3.4.1</span> Preparation of the test data</a></li>
<li><a href="quick-start.html#use-a-specific-function-called-padi" id="toc-use-a-specific-function-called-padi"><span class="toc-section-number">3.4.2</span> Use a specific function called <code>PADi</code></a></li>
<li><a href="quick-start.html#use-the-callensemble-function" id="toc-use-the-callensemble-function"><span class="toc-section-number">3.4.3</span> Use the <code>callEnsemble</code> function</a></li>
<li><a href="quick-start.html#parallel-strategy-for-padi" id="toc-parallel-strategy-for-padi"><span class="toc-section-number">3.4.4</span> Parallel strategy for PADi</a></li>
<li><a href="quick-start.html#single-sample-subtype-calling" id="toc-single-sample-subtype-calling"><span class="toc-section-number">3.4.5</span> Single sample subtype calling</a></li>
<li><a href="quick-start.html#of-note-1" id="toc-of-note-1"><span class="toc-section-number">3.4.6</span> Of note</a></li>
</ul></li>
<li><a href="quick-start.html#use-external-models-from-luckymodel-package" id="toc-use-external-models-from-luckymodel-package"><span class="toc-section-number">3.5</span> Use external models from luckyModel package</a></li>
<li><a href="quick-start.html#pancancer-immune-subtypes" id="toc-pancancer-immune-subtypes"><span class="toc-section-number">3.6</span> PanCancer immune subtypes</a></li>
</ul></li>
<li><a href="model-establishment-via-gsclassifier.html#model-establishment-via-gsclassifier" id="toc-model-establishment-via-gsclassifier"><span class="toc-section-number">4</span> Model establishment via GSClassifier</a>
<ul>
<li><a href="model-establishment-via-gsclassifier.html#about-1" id="toc-about-1"><span class="toc-section-number">4.1</span> About</a></li>
<li><a href="model-establishment-via-gsclassifier.html#data-preparation" id="toc-data-preparation"><span class="toc-section-number">4.2</span> Data preparation</a></li>
<li><a href="model-establishment-via-gsclassifier.html#fitting-models" id="toc-fitting-models"><span class="toc-section-number">4.3</span> Fitting models</a>
<ul>
<li><a href="model-establishment-via-gsclassifier.html#gsclassifier-model-training" id="toc-gsclassifier-model-training"><span class="toc-section-number">4.3.1</span> GSClassifier model training</a></li>
<li><a href="model-establishment-via-gsclassifier.html#scaller-for-the-best-call" id="toc-scaller-for-the-best-call"><span class="toc-section-number">4.3.2</span> Scaller for the best call</a></li>
<li><a href="model-establishment-via-gsclassifier.html#assemble-your-model" id="toc-assemble-your-model"><span class="toc-section-number">4.3.3</span> Assemble your model</a></li>
<li><a href="model-establishment-via-gsclassifier.html#of-note-2" id="toc-of-note-2"><span class="toc-section-number">4.3.4</span> Of note</a></li>
</ul></li>
<li><a href="model-establishment-via-gsclassifier.html#calling-subtypes" id="toc-calling-subtypes"><span class="toc-section-number">4.4</span> Calling subtypes</a></li>
</ul></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#suggestions-for-gsclassifier-model-developers" id="toc-suggestions-for-gsclassifier-model-developers"><span class="toc-section-number">5</span> Suggestions for GSClassifier model developers</a>
<ul>
<li><a href="suggestions-for-gsclassifier-model-developers.html#about-2" id="toc-about-2"><span class="toc-section-number">5.1</span> About</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#available-models" id="toc-available-models"><span class="toc-section-number">5.2</span> Available models</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#components-of-a-gsclassifier-model" id="toc-components-of-a-gsclassifier-model"><span class="toc-section-number">5.3</span> Components of a GSClassifier model</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#submit-models-to-luckymodel-package" id="toc-submit-models-to-luckymodel-package"><span class="toc-section-number">5.4</span> Submit models to luckyModel package</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#repeatablility-of-models" id="toc-repeatablility-of-models"><span class="toc-section-number">5.5</span> Repeatablility of models</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#gene-annotation" id="toc-gene-annotation"><span class="toc-section-number">5.6</span> Gene Annotation</a></li>
</ul></li>
<li><a href="references.html#references" id="toc-references">References</a></li>
<li class="divider"></li>
<li><a href="https://blognas.hwb0307.com/" target="blank">Welcome to My Blog!</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Principle of R Package GSClassifier</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="suggestions-for-gsclassifier-model-developers" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Suggestions for GSClassifier model developers</h1>
<div id="about-2" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> About</h2>
<ul>
<li><p>The book <code>R packages</code> is a straightaway and useful reference book for R developers. The free-access website of <code>R packages</code> is <a href="https://r-pkgs.org/">https://r-pkgs.org/</a>. As a developer of R, if you haven’t hear about it, it’s strongly recommanded to just read it. Hadley Wickham, the main author of the book, is an active R developer and have led some master works like <code>ggplot2</code> and <code>plyr</code>.</p></li>
<li><p>With <code>GSClassifier</code> package, it could be easy for users to build a model only with certain gene sets and transcriptomics data. If you are interesting in sharing your model, <code>GSClassifier</code> also provides a simple methodology for this vision. In this section, let’s see how to achieve it!</p></li>
</ul>
<p>First, load the package</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GSClassifier)</span>
<span id="cb78-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="co"># 载入需要的程辑包：luckyBase</span></span></code></pre></div>
</div>
<div id="available-models" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Available models</h2>
<p>With <code>GSClassifier_Data()</code>, all models supported in the current <code>GSClassifier</code> package would showed.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">GSClassifier_Data</span>()</span>
<span id="cb79-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Available data:</span></span>
<span id="cb79-3"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage example:</span></span>
<span id="cb79-4"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype.rds </span></span>
<span id="cb79-5"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20200110.rds </span></span>
<span id="cb79-6"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20220916.rds </span></span>
<span id="cb79-7"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;PAD.train_20200110.rds&quot;, package = &quot;GSClassifier&quot;)) </span></span>
<span id="cb79-8"><a href="suggestions-for-gsclassifier-model-developers.html#cb79-8" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;ImmuneSubtype.rds&quot;, package = &quot;GSClassifier&quot;))</span></span></code></pre></div>
<p>For more details of <code>GSClassifier_Data()</code>, just:</p>
<pre><code>?GSClassifier_Data()</code></pre>
<p>Set <code>model=F</code>, all <code>.rds</code> data would be showed:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">GSClassifier_Data</span>(<span class="at">model =</span> F)</span>
<span id="cb81-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Available data:</span></span>
<span id="cb81-3"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage example:</span></span>
<span id="cb81-4"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   general-gene-annotation.rds </span></span>
<span id="cb81-5"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype.rds </span></span>
<span id="cb81-6"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20200110.rds </span></span>
<span id="cb81-7"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD.train_20220916.rds </span></span>
<span id="cb81-8"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="co">#   testData.rds </span></span>
<span id="cb81-9"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="co">#   PAD &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;PAD.train_20200110.rds&quot;, package = &quot;GSClassifier&quot;)) </span></span>
<span id="cb81-10"><a href="suggestions-for-gsclassifier-model-developers.html#cb81-10" aria-hidden="true" tabindex="-1"></a><span class="co">#   ImmuneSubtype &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;ImmuneSubtype.rds&quot;, package = &quot;GSClassifier&quot;))</span></span></code></pre></div>
</div>
<div id="components-of-a-gsclassifier-model" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Components of a GSClassifier model</h2>
<p>Currently, a GSClassifier model and related product environments is designed as a <code>list</code> object. Let’s take <code>PAD.train_20210110</code>(also called <code>PADi</code>) as an example.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb82-1" aria-hidden="true" tabindex="-1"></a>PADi <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;PAD.train_20220916.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>)) </span></code></pre></div>
<p>This picture shows the components of <code>PADi</code>:</p>
<div class="figure" style="text-align: center"><span id="fig:GSClassifierModel"></span>
<img src="fig/GSClassifier-model.jpg" alt="Details of a GSClassifier model" width="90%" />
<p class="caption">
Figure 5.1: Details of a GSClassifier model
</p>
</div>
<p>As shown, a typical <code>GSClassifier</code> model is consist of four parts (with different colors in the picture):</p>
<ul>
<li><code>1. ens</code>:
<ul>
<li><code>Repeat</code>: productive parameters of <code>GSClassifier</code> models</li>
<li><code>Model</code>: <code>GSClassifier</code> models. Here, <code>PADi</code> had 20 models from different subs of the training cohorts</li>
</ul></li>
<li><code>2. scaller</code>:
<ul>
<li><code>Repeat</code>: productive parameters of the <code>scaller</code> model, which was used for <code>BestCall</code> calling</li>
<li><code>Model</code>: the <code>scaller</code> model</li>
</ul></li>
<li><code>3. geneAnnotation</code>: a data frame containing gene annotation information</li>
<li><code>4. geneSet</code>: a list contains several gene sets</li>
</ul>
<p>Thus, you can assemble your model like:</p>
<pre><code>model &lt;- list()

# bootstrap models based on the training cohort
model[[&#39;ens&#39;]] &lt;- &lt;Your model for subtypes calling&gt;

# Scaller model
model[[&#39;scaller&#39;]] &lt;- &lt;Your scaller for BestCall calling&gt;

# a data frame contarining gene annotation for IDs convertion
model[[&#39;geneAnnotation&#39;]] &lt;- &lt;Your gene annotation&gt;

# Your gene sets
model[[&#39;geneSet&#39;]] &lt;- &lt;Your gene sets&gt;

saveRDS(model, &#39;your-model.rds&#39;)
</code></pre>
<p>More tutorials for model establishment, please go to <a href="https://github.com/huangwb8/GSClassifier/wiki/Model-establishment">markdown tutorial</a> or <a href="http://htmlpreview.github.io/?https://raw.githubusercontent.com/wiki/huangwb8/GSClassifier/Model-establishment.html">html tutorial</a>.</p>
</div>
<div id="submit-models-to-luckymodel-package" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Submit models to luckyModel package</h2>
<p>Considering most users of <code>GSClassifier</code> might have no need for lots of models, We divided the model storage feature into an new ensembl package called <a href="https://github.com/huangwb8/luckyModel"><strong>luckyModel</strong></a>. Don’t worry, the usage is very easy!</p>
<p>If you want to summit your model, you should apply for a contributor of <code>luckyModel</code> first. Then, just send the model (<code>.rds</code>) into the <code>inst/extdata/&lt;project&gt;</code> path of <code>luckyModel</code>. After audit, your branch would be accepted and available for the users.</p>
<p>The name of your model must be the format as following:</p>
<pre><code># &lt;project&gt;
GSClassifier

# &lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;:
HWB_PAD_v20211201.rds</code></pre>
</div>
<div id="repeatablility-of-models" class="section level2" number="5.5">
<h2><span class="header-section-number">5.5</span> Repeatablility of models</h2>
<p>For repeatablility, you had better submit a <code>.zip</code> or <code>.tar.gz</code> file that containing the information of your model. Here are some suggestions:</p>
<ul>
<li><p><code>&lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;.md</code></p>
<ul>
<li><p><strong>Destinations</strong>: Why you develop the model</p></li>
<li><p><strong>Design</strong>: The evidence for gene sigatures, et al</p></li>
<li><p><strong>Data sources</strong>: The data for model training and validating, et al</p></li>
<li><p><strong>Applications</strong>: Where to use your model</p></li>
<li><p><strong>Limintations</strong>: Limitation or improvement direction of your model</p></li>
</ul></li>
<li><p><code>&lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;.R</code>: The code you used for model training and validating.</p></li>
<li><p><code>Data-of-&lt;creator&gt;_&lt;model&gt;_v&lt;yyyymmdd&gt;.rds</code> (Optional): Due to huge size of omics data, it’s OK for you not to submit the raw data.</p></li>
</ul>
<p>:cupid: Welcome your contributions!</p>
</div>
<div id="gene-annotation" class="section level2" number="5.6">
<h2><span class="header-section-number">5.6</span> Gene Annotation</h2>
<p>For convenience, we provided a general gene annotation dataset for different genomics:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb85-1" aria-hidden="true" tabindex="-1"></a>gga <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;general-gene-annotation.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb85-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gga)</span>
<span id="cb85-3"><a href="suggestions-for-gsclassifier-model-developers.html#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;hg38&quot; &quot;hg19&quot; &quot;mm10&quot;</span></span></code></pre></div>
<p>I believe they’re enough for routine medicine studies.</p>
<p>Here, take a look at <code>hg38</code>:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-1" aria-hidden="true" tabindex="-1"></a>hg38 <span class="ot">&lt;-</span> gga<span class="sc">$</span>hg38</span>
<span id="cb86-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hg38)</span>
<span id="cb86-3"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-3" aria-hidden="true" tabindex="-1"></a><span class="co">#           ENSEMBL       SYMBOL  ENTREZID</span></span>
<span id="cb86-4"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1 ENSG00000223972      DDX11L1 100287102</span></span>
<span id="cb86-5"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 ENSG00000227232       WASH7P      &lt;NA&gt;</span></span>
<span id="cb86-6"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 4 ENSG00000278267    MIR6859-1 102466751</span></span>
<span id="cb86-7"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 5 ENSG00000243485 RP11-34P13.3      &lt;NA&gt;</span></span>
<span id="cb86-8"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 6 ENSG00000284332    MIR1302-2 100302278</span></span>
<span id="cb86-9"><a href="suggestions-for-gsclassifier-model-developers.html#cb86-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 7 ENSG00000237613      FAM138A    645520</span></span></code></pre></div>
<p>With this kind of data, it’s simple to customize your own gene annotation (take <code>PADi</code> as examples):</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb87-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb87-2" aria-hidden="true" tabindex="-1"></a>tGene <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">unlist</span>(PADi<span class="sc">$</span>geneSet))</span>
<span id="cb87-3"><a href="suggestions-for-gsclassifier-model-developers.html#cb87-3" aria-hidden="true" tabindex="-1"></a>geneAnnotation <span class="ot">&lt;-</span> hg38[hg38<span class="sc">$</span>ENSEMBL <span class="sc">%in%</span> tGene, ]</span>
<span id="cb87-4"><a href="suggestions-for-gsclassifier-model-developers.html#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(geneAnnotation)</span>
<span id="cb87-5"><a href="suggestions-for-gsclassifier-model-developers.html#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] 32  3</span></span></code></pre></div>
<p>Have a check:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(geneAnnotation)</span>
<span id="cb88-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co">#              ENSEMBL SYMBOL ENTREZID</span></span>
<span id="cb88-3"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 353  ENSG00000171608 PIK3CD     5293</span></span>
<span id="cb88-4"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 1169 ENSG00000134686   PHC2     1912</span></span>
<span id="cb88-5"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 2892 ENSG00000134247 PTGFRN     5738</span></span>
<span id="cb88-6"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 3855 ENSG00000117090 SLAMF1     6504</span></span>
<span id="cb88-7"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 3858 ENSG00000117091   CD48      962</span></span>
<span id="cb88-8"><a href="suggestions-for-gsclassifier-model-developers.html#cb88-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 4043 ENSG00000198821  CD247      919</span></span></code></pre></div>
<p>This <code>geneAnnotation</code> could be the <code>model[['geneAnnotation']]</code>.</p>
<p>Also, we use a function called <code>convert</code> to do gene ID convertion.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="suggestions-for-gsclassifier-model-developers.html#cb89-1" aria-hidden="true" tabindex="-1"></a>luckyBase<span class="sc">::</span><span class="fu">convert</span>(<span class="fu">c</span>(<span class="st">&#39;GAPDH&#39;</span>,<span class="st">&#39;TP53&#39;</span>), <span class="st">&#39;SYMBOL&#39;</span>, <span class="st">&#39;ENSEMBL&#39;</span>, hg38)</span>
<span id="cb89-2"><a href="suggestions-for-gsclassifier-model-developers.html#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;ENSG00000111640&quot; &quot;ENSG00000141510&quot;</span></span></code></pre></div>
<p>Note: the <code>luckyBase</code> package integrates lots of useful tiny functions, you could explore it sometimes.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="model-establishment-via-gsclassifier.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["GSClassifier.principle.pdf", "GSClassifier.principle.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
