<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 1 The Principle of GSClassifier | The Principle of R Package GSClassifier</title>
  <meta name="description" content="The Principle of GSClassifier" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 1 The Principle of GSClassifier | The Principle of R Package GSClassifier" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The Principle of GSClassifier" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 The Principle of GSClassifier | The Principle of R Package GSClassifier" />
  
  <meta name="twitter:description" content="The Principle of GSClassifier" />
  

<meta name="author" content="Weibin Huang" />


<meta name="date" content="2022-09-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="discussion.html"/>
<script src="libs/header-attrs-2.14/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">GSClassifier</a></li>

<li class="divider"></li>
<li><a href="index.html#welcome" id="toc-welcome">Welcome</a>
<ul>
<li><a href="index.html#license" id="toc-license">License</a></li>
<li><a href="index.html#other-projects" id="toc-other-projects">Other Projects</a></li>
</ul></li>
<li><a href="the-principle-of-gsclassifier.html#the-principle-of-gsclassifier" id="toc-the-principle-of-gsclassifier"><span class="toc-section-number">1</span> The Principle of GSClassifier</a>
<ul>
<li><a href="the-principle-of-gsclassifier.html#introduction" id="toc-introduction"><span class="toc-section-number">1.1</span> Introduction</a></li>
<li><a href="the-principle-of-gsclassifier.html#flowchart" id="toc-flowchart"><span class="toc-section-number">1.2</span> Flowchart</a>
<ul>
<li><a href="the-principle-of-gsclassifier.html#data-processing" id="toc-data-processing"><span class="toc-section-number">1.2.1</span> Data Processing</a></li>
<li><a href="the-principle-of-gsclassifier.html#model-establishment-and-validation" id="toc-model-establishment-and-validation"><span class="toc-section-number">1.2.2</span> Model Establishment and Validation</a></li>
<li><a href="the-principle-of-gsclassifier.html#model-application" id="toc-model-application"><span class="toc-section-number">1.2.3</span> Model Application</a></li>
</ul></li>
<li><a href="the-principle-of-gsclassifier.html#topicTSP" id="toc-topicTSP"><span class="toc-section-number">1.3</span> Top scoring pairs</a>
<ul>
<li><a href="the-principle-of-gsclassifier.html#tsp-intro" id="toc-tsp-intro"><span class="toc-section-number">1.3.1</span> Introduction</a></li>
<li><a href="the-principle-of-gsclassifier.html#packages" id="toc-packages"><span class="toc-section-number">1.3.2</span> Packages</a></li>
<li><a href="the-principle-of-gsclassifier.html#simulated-dataset" id="toc-simulated-dataset"><span class="toc-section-number">1.3.3</span> Simulated Dataset</a></li>
<li><a href="the-principle-of-gsclassifier.html#missing-value-imputation-mvi" id="toc-missing-value-imputation-mvi"><span class="toc-section-number">1.3.4</span> Missing value imputation (MVI)</a></li>
<li><a href="the-principle-of-gsclassifier.html#binned-expression" id="toc-binned-expression"><span class="toc-section-number">1.3.5</span> Binned expression</a></li>
<li><a href="the-principle-of-gsclassifier.html#pair-difference" id="toc-pair-difference"><span class="toc-section-number">1.3.6</span> Pair difference</a></li>
<li><a href="the-principle-of-gsclassifier.html#set-difference" id="toc-set-difference"><span class="toc-section-number">1.3.7</span> Set difference</a></li>
<li><a href="the-principle-of-gsclassifier.html#batch-effect" id="toc-batch-effect"><span class="toc-section-number">1.3.8</span> Batch effect</a></li>
</ul></li>
</ul></li>
<li><a href="discussion.html#discussion" id="toc-discussion"><span class="toc-section-number">2</span> Discussion</a>
<ul>
<li><a href="discussion.html#model-complexibility" id="toc-model-complexibility"><span class="toc-section-number">2.1</span> Model complexibility</a></li>
<li><a href="discussion.html#model-extensibility" id="toc-model-extensibility"><span class="toc-section-number">2.2</span> Model extensibility</a></li>
<li><a href="discussion.html#missing-value" id="toc-missing-value"><span class="toc-section-number">2.3</span> Missing value</a></li>
<li><a href="discussion.html#subtypevector" id="toc-subtypevector"><span class="toc-section-number">2.4</span> Subtype vector calling</a></li>
<li><a href="discussion.html#topicSubmodel" id="toc-topicSubmodel"><span class="toc-section-number">2.5</span> The number of submodels</a></li>
</ul></li>
<li><a href="quick-start.html#quick-start" id="toc-quick-start"><span class="toc-section-number">3</span> Quick start</a>
<ul>
<li><a href="quick-start.html#about" id="toc-about"><span class="toc-section-number">3.1</span> About</a></li>
<li><a href="quick-start.html#data" id="toc-data"><span class="toc-section-number">3.2</span> Data</a></li>
<li><a href="quick-start.html#pad" id="toc-pad"><span class="toc-section-number">3.3</span> PAD</a>
<ul>
<li><a href="quick-start.html#preparation-of-the-test-data" id="toc-preparation-of-the-test-data"><span class="toc-section-number">3.3.1</span> Preparation of the test data</a></li>
<li><a href="quick-start.html#unsupervised-clustering" id="toc-unsupervised-clustering"><span class="toc-section-number">3.3.2</span> Unsupervised clustering</a></li>
<li><a href="quick-start.html#of-note" id="toc-of-note"><span class="toc-section-number">3.3.3</span> Of note</a></li>
</ul></li>
<li><a href="quick-start.html#padi" id="toc-padi"><span class="toc-section-number">3.4</span> PADi</a>
<ul>
<li><a href="quick-start.html#preparation-of-the-test-data-1" id="toc-preparation-of-the-test-data-1"><span class="toc-section-number">3.4.1</span> Preparation of the test data</a></li>
<li><a href="quick-start.html#use-a-specific-function-called-padi" id="toc-use-a-specific-function-called-padi"><span class="toc-section-number">3.4.2</span> Use a specific function called <code>PADi</code></a></li>
<li><a href="quick-start.html#use-the-callensemble-function" id="toc-use-the-callensemble-function"><span class="toc-section-number">3.4.3</span> Use the <code>callEnsemble</code> function</a></li>
<li><a href="quick-start.html#parallel-strategy-for-padi" id="toc-parallel-strategy-for-padi"><span class="toc-section-number">3.4.4</span> Parallel strategy for PADi</a></li>
<li><a href="quick-start.html#single-sample-subtype-calling" id="toc-single-sample-subtype-calling"><span class="toc-section-number">3.4.5</span> Single sample subtype calling</a></li>
<li><a href="quick-start.html#of-note-1" id="toc-of-note-1"><span class="toc-section-number">3.4.6</span> Of note</a></li>
</ul></li>
<li><a href="quick-start.html#use-external-models-from-luckymodel-package" id="toc-use-external-models-from-luckymodel-package"><span class="toc-section-number">3.5</span> Use external models from luckyModel package</a></li>
<li><a href="quick-start.html#pancancer-immune-subtypes" id="toc-pancancer-immune-subtypes"><span class="toc-section-number">3.6</span> PanCancer immune subtypes</a></li>
</ul></li>
<li><a href="model-establishment-via-gsclassifier.html#model-establishment-via-gsclassifier" id="toc-model-establishment-via-gsclassifier"><span class="toc-section-number">4</span> Model establishment via GSClassifier</a>
<ul>
<li><a href="model-establishment-via-gsclassifier.html#about-1" id="toc-about-1"><span class="toc-section-number">4.1</span> About</a></li>
<li><a href="model-establishment-via-gsclassifier.html#data-preparation" id="toc-data-preparation"><span class="toc-section-number">4.2</span> Data preparation</a></li>
<li><a href="model-establishment-via-gsclassifier.html#fitting-models" id="toc-fitting-models"><span class="toc-section-number">4.3</span> Fitting models</a>
<ul>
<li><a href="model-establishment-via-gsclassifier.html#gsclassifier-model-training" id="toc-gsclassifier-model-training"><span class="toc-section-number">4.3.1</span> GSClassifier model training</a></li>
<li><a href="model-establishment-via-gsclassifier.html#scaller-for-the-best-call" id="toc-scaller-for-the-best-call"><span class="toc-section-number">4.3.2</span> Scaller for the best call</a></li>
<li><a href="model-establishment-via-gsclassifier.html#assemble-your-model" id="toc-assemble-your-model"><span class="toc-section-number">4.3.3</span> Assemble your model</a></li>
<li><a href="model-establishment-via-gsclassifier.html#of-note-2" id="toc-of-note-2"><span class="toc-section-number">4.3.4</span> Of note</a></li>
</ul></li>
<li><a href="model-establishment-via-gsclassifier.html#calling-subtypes" id="toc-calling-subtypes"><span class="toc-section-number">4.4</span> Calling subtypes</a></li>
</ul></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#suggestions-for-gsclassifier-model-developers" id="toc-suggestions-for-gsclassifier-model-developers"><span class="toc-section-number">5</span> Suggestions for GSClassifier model developers</a>
<ul>
<li><a href="suggestions-for-gsclassifier-model-developers.html#about-2" id="toc-about-2"><span class="toc-section-number">5.1</span> About</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#available-models" id="toc-available-models"><span class="toc-section-number">5.2</span> Available models</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#components-of-a-gsclassifier-model" id="toc-components-of-a-gsclassifier-model"><span class="toc-section-number">5.3</span> Components of a GSClassifier model</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#submit-models-to-luckymodel-package" id="toc-submit-models-to-luckymodel-package"><span class="toc-section-number">5.4</span> Submit models to luckyModel package</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#repeatablility-of-models" id="toc-repeatablility-of-models"><span class="toc-section-number">5.5</span> Repeatablility of models</a></li>
<li><a href="suggestions-for-gsclassifier-model-developers.html#gene-annotation" id="toc-gene-annotation"><span class="toc-section-number">5.6</span> Gene Annotation</a></li>
</ul></li>
<li><a href="references.html#references" id="toc-references">References</a></li>
<li class="divider"></li>
<li><a href="https://blognas.hwb0307.com/" target="blank">Welcome to My Blog!</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Principle of R Package GSClassifier</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="the-principle-of-gsclassifier" class="section level1" number="1">
<h1><span class="header-section-number">Chapter 1</span> The Principle of GSClassifier</h1>
<div id="introduction" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Introduction</h2>
<!--
描述GSClassifier的Flowchart，关注每一步的具体细节；
演示TSP的计算过程
-->
<p><a href="https://github.com/huangwb8/GSClassifier">GSClassifier</a> is an R package for modeling and identification of Gene Expression Profiles (GEPs) subtypes. The detail of <strong>GSClassifier</strong> package usage had been demonstrated in <a href="https://github.com/huangwb8/GSClassifier/wiki">Github WiKi</a>. Here, we propose to introduce the principle of GSClassifier, including flowchart, <strong>top scoring pairs (TSP)</strong> algorithm, and batch effect control.</p>
</div>
<div id="flowchart" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Flowchart</h2>
<p>The flowchart of <strong>GSClassifier</strong> is showed in Figure <a href="the-principle-of-gsclassifier.html#fig:flowchart">1.1</a>.</p>
<div class="figure" style="text-align: center"><span id="fig:flowchart"></span>
<img src="fig/flowchart.png" alt="The flow chart of GSClassifier" width="90%" />
<p class="caption">
Figure 1.1: The flow chart of GSClassifier
</p>
</div>
<div id="data-processing" class="section level3" number="1.2.1">
<h3><span class="header-section-number">1.2.1</span> Data Processing</h3>
<p>For each dataset, RNA expression matrix would be normalized internally (<strong>Raw Matrix</strong>) so that the expression data of the samples in the dataset were comparable and suitable for subtype identification. As demonstrated in Figure <a href="the-principle-of-gsclassifier.html#fig:flowchart">1.1</a>, the <strong>Subtype vector</strong> is identified based on independent cohorts instead of a merged matrix with batch effect control technologies. More details about batch effect control are discussed in <a href="discussion.html#subtypevector">2.4</a>.</p>
<p>There is no standard method to figure out subtype vectors. It depends on the Gene Expression Profiles (GEPs) used, the biological problems or ideas of researchers. For <strong>Pan-immune Activation and Dysfunction (PAD)</strong> subtypes, the GEPs, <strong>Pan-Immune Activation Module (PIAM)</strong> and <strong>Pan-Immune Dysfunction Genes (PIDG)</strong>, are biologically associated and suitable for calling four sutbypes (PIAM<sup>high</sup>PIDG<sup>high</sup>, PIAM<sup>high</sup>PIDG<sup>low</sup>, PIAM<sup>low</sup>PIDG<sup>high</sup>, and PIAM<sup>low</sup>PIDG<sup>low</sup>).<!--这里描述有待改进--> Theoretically, we can also use a category strategy like low/medium/high, but more evidences or motivations are lacked for chasing such a complex model.</p>
<p>With subtype vectors and raw matrices, <strong>Top Scoring Pairs (TSP)</strong>, the core data format for model training and application in GSClassifier, would be calculated for the following process. The details of TSP are summarized in <a href="the-principle-of-gsclassifier.html#topicTSP">1.3</a>.</p>
</div>
<div id="model-establishment-and-validation" class="section level3" number="1.2.2">
<h3><span class="header-section-number">1.2.2</span> Model Establishment and Validation</h3>
<p>The TSP matrix would be devided into training cohort and internal validation cohort. In PAD project, the rate of samples (training vs. test) is <strong>7:3</strong>. Next, each <strong>Subset</strong> (70% of the training cohort in PAD project) would be further selected randomly to build a <strong>Submodel</strong> via cross-validation Extreme Gradient Boosting algorithm (<strong>xgboost::xgb.cv</strong> function) <span class="citation">[<a href="references.html#ref-RN345" role="doc-biblioref">1</a>]</span>. The number of submodels is suggested over 20 (more details in <a href="discussion.html#topicSubmodel">2.5</a>).</p>
<p>The internal validation cohort and external validation cohort (if any) would be used to test the performace of the trained model. By the way, <strong>the data of both internal and external validation cohort would not be used during model training</strong> so as to avoid over-fitting.</p>
</div>
<div id="model-application" class="section level3" number="1.2.3">
<h3><span class="header-section-number">1.2.3</span> Model Application</h3>
<p>In PAD project, <strong>Model for individual</strong>, the ensemble of submodels, is called “<strong>PAD for individual</strong>” (<strong>PADi</strong>). Supposed raw RNA expression of a sample was given. As showed in <a href="the-principle-of-gsclassifier.html#fig:flowchart">1.1</a> and <a href="the-principle-of-gsclassifier.html#fig:tsp">1.2</a>, <strong>GSClassifier</strong> would turn raw RNA expression into a TSP vector, which would be as a input to <strong>Model for individual</strong>. Then, <strong>GSClassifier</strong> would output the possibility matrix and the subtype for this sample. No extra data (RNA expression of others, follow-up data, etc) would be needed but RNA expression of the patient for subtype identification, so we suggest <strong>Model for individual</strong> (<strong>PADi</strong>, etc) as personalized model.</p>
</div>
</div>
<div id="topicTSP" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Top scoring pairs</h2>
<div id="tsp-intro" class="section level3" number="1.3.1">
<h3><span class="header-section-number">1.3.1</span> Introduction</h3>
<p>Genes expression of an individual is normalized during the model training and the subtype identification via <strong>Top Scoring Pairs</strong> (<strong>TSP</strong>, also called <strong>Relative Expression Orderings</strong> (<strong>REOs</strong>)) algorithm, which was previously described by Geman et al <span class="citation">[<a href="references.html#ref-RN267" role="doc-biblioref">2</a>]</span>. <strong>TSP</strong> normalization for an individual depends on its transcript data, implying that subtype calling would not be perturbed by data from other individuals or other extra information like follow-up data. <strong>TSP</strong> had been used in cancer research and effectively predicts cancer progression and ICIs response <span class="citation">[<a href="references.html#ref-RN265" role="doc-biblioref">3</a>–<a href="references.html#ref-RN261" role="doc-biblioref">5</a>]</span>.</p>
<p>As show in Figure <a href="the-principle-of-gsclassifier.html#fig:tsp">1.2</a>, The TSP matrix in GSClassifier consists of three parts: <strong>binned expression</strong>, <strong>pair difference</strong>, and <strong>set difference</strong>. In this section, we would conduct some experiments to demonstrate the potential of TSP normalization for development of cross-dataset/platform GEP-based models.</p>
<div class="figure" style="text-align: center"><span id="fig:tsp"></span>
<img src="fig/TSP.png" alt="The components of TSP (2 gene sets)" width="85%" />
<p class="caption">
Figure 1.2: The components of TSP (2 gene sets)
</p>
</div>
</div>
<div id="packages" class="section level3" number="1.3.2">
<h3><span class="header-section-number">1.3.2</span> Packages</h3>
<p>First, load needed packages:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="the-principle-of-gsclassifier.html#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="the-principle-of-gsclassifier.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Install &quot;devtools&quot; package</span></span>
<span id="cb1-3"><a href="the-principle-of-gsclassifier.html#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-4"><a href="the-principle-of-gsclassifier.html#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-5"><a href="the-principle-of-gsclassifier.html#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="the-principle-of-gsclassifier.html#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Install dependencies</span></span>
<span id="cb1-7"><a href="the-principle-of-gsclassifier.html#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;luckyBase&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-8"><a href="the-principle-of-gsclassifier.html#cb1-8" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/luckyBase&quot;</span>)</span>
<span id="cb1-9"><a href="the-principle-of-gsclassifier.html#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="the-principle-of-gsclassifier.html#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;GSClassifier&quot; package</span></span>
<span id="cb1-11"><a href="the-principle-of-gsclassifier.html#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;GSClassifier&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-12"><a href="the-principle-of-gsclassifier.html#cb1-12" aria-hidden="true" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;huangwb8/GSClassifier&quot;</span>)</span>
<span id="cb1-13"><a href="the-principle-of-gsclassifier.html#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="the-principle-of-gsclassifier.html#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the &quot;pacman&quot; package</span></span>
<span id="cb1-15"><a href="the-principle-of-gsclassifier.html#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;pacman&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb1-16"><a href="the-principle-of-gsclassifier.html#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;pacman&quot;</span>)</span>
<span id="cb1-17"><a href="the-principle-of-gsclassifier.html#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb1-18"><a href="the-principle-of-gsclassifier.html#cb1-18" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb1-19"><a href="the-principle-of-gsclassifier.html#cb1-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(pacman)</span>
<span id="cb1-20"><a href="the-principle-of-gsclassifier.html#cb1-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-21"><a href="the-principle-of-gsclassifier.html#cb1-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-22"><a href="the-principle-of-gsclassifier.html#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Load needed packages</span></span>
<span id="cb1-23"><a href="the-principle-of-gsclassifier.html#cb1-23" aria-hidden="true" tabindex="-1"></a>packages_needed <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb1-24"><a href="the-principle-of-gsclassifier.html#cb1-24" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;readxl&quot;</span>,</span>
<span id="cb1-25"><a href="the-principle-of-gsclassifier.html#cb1-25" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;ComplexHeatmap&quot;</span>,</span>
<span id="cb1-26"><a href="the-principle-of-gsclassifier.html#cb1-26" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;GSClassifier&quot;</span>,</span>
<span id="cb1-27"><a href="the-principle-of-gsclassifier.html#cb1-27" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;rpart&quot;</span>,</span>
<span id="cb1-28"><a href="the-principle-of-gsclassifier.html#cb1-28" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;tidyr&quot;</span>,</span>
<span id="cb1-29"><a href="the-principle-of-gsclassifier.html#cb1-29" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;reshape2&quot;</span>,</span>
<span id="cb1-30"><a href="the-principle-of-gsclassifier.html#cb1-30" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb1-31"><a href="the-principle-of-gsclassifier.html#cb1-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> packages_needed){<span class="fu">p_load</span>(<span class="at">char=</span>i)}</span></code></pre></div>
<p>Here is the environment of R programming:</p>
<pre><code># R version 4.0.3 (2020-10-10)
# Platform: x86_64-w64-mingw32/x64 (64-bit)
# Running under: Windows 10 x64 (build 18363)
# 
# Matrix products: default
# 
# locale:
# [1] LC_COLLATE=Chinese (Simplified)_China.936 
# [2] LC_CTYPE=Chinese (Simplified)_China.936   
# [3] LC_MONETARY=Chinese (Simplified)_China.936
# [4] LC_NUMERIC=C                              
# [5] LC_TIME=Chinese (Simplified)_China.936    
# 
# attached base packages:
# [1] grid      stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
# [1] ggplot2_3.3.6        reshape2_1.4.4       tidyr_1.2.0         
# [4] rpart_4.1.16         GSClassifier_0.1.22  luckyBase_0.1.0     
# [7] ComplexHeatmap_2.4.3 readxl_1.4.0         pacman_0.5.1        
# 
# loaded via a namespace (and not attached):
#   [1] colorspace_2.0-3     ggsignif_0.6.3       rjson_0.2.21        
#   [4] ellipsis_0.3.2       class_7.3-20         rprojroot_2.0.3     
#   [7] circlize_0.4.15      GlobalOptions_0.1.2  fs_1.5.2            
#  [10] clue_0.3-57          rstudioapi_0.13      listenv_0.8.0       
#  [13] ggpubr_0.4.0         remotes_2.4.2        lubridate_1.8.0     
#  [16] prodlim_2019.11.13   fansi_1.0.3          codetools_0.2-18    
#  [19] splines_4.0.3        doParallel_1.0.17    cachem_1.0.6        
#  [22] knitr_1.30           pkgload_1.2.4        jsonlite_1.8.0      
#  [25] pROC_1.18.0          caret_6.0-92         broom_1.0.0         
#  [28] cluster_2.1.3        png_0.1-7            compiler_4.0.3      
#  [31] backports_1.4.1      assertthat_0.2.1     Matrix_1.2-18       
#  [34] fastmap_1.1.0        cli_3.3.0            htmltools_0.5.2     
#  [37] prettyunits_1.1.1    tools_4.0.3          gtable_0.3.0        
#  [40] glue_1.6.2           dplyr_1.0.9          Rcpp_1.0.8.3        
#  [43] carData_3.0-5        cellranger_1.1.0     jquerylib_0.1.4     
#  [46] vctrs_0.4.1          nlme_3.1-149         iterators_1.0.14    
#  [49] timeDate_3043.102    xfun_0.33            gower_1.0.0         
#  [52] stringr_1.4.0        globals_0.15.1       ps_1.4.0            
#  [55] testthat_3.1.0       lifecycle_1.0.1      devtools_2.4.3      
#  [58] rstatix_0.7.0        future_1.26.1        MASS_7.3-53         
#  [61] scales_1.2.0         ipred_0.9-12         parallel_4.0.3      
#  [64] RColorBrewer_1.1-3   yaml_2.3.5           memoise_2.0.1       
#  [67] sass_0.4.1           stringi_1.7.6        highr_0.9           
#  [70] desc_1.4.1           randomForest_4.6-14  foreach_1.5.2       
#  [73] hardhat_1.1.0        pkgbuild_1.3.1       lava_1.6.10         
#  [76] shape_1.4.6          tuneR_1.4.0          rlang_1.0.2         
#  [79] pkgconfig_2.0.3      evaluate_0.15        lattice_0.20-41     
#  [82] purrr_0.3.4          recipes_0.2.0        processx_3.7.0      
#  [85] tidyselect_1.1.2     parallelly_1.32.0    plyr_1.8.7          
#  [88] magrittr_2.0.3       bookdown_0.21        R6_2.5.1            
#  [91] generics_0.1.2       DBI_1.1.3            pillar_1.7.0        
#  [94] withr_2.5.0          survival_3.3-1       abind_1.4-5         
#  [97] nnet_7.3-17          tibble_3.1.7         future.apply_1.9.0  
# [100] crayon_1.5.1         car_3.1-0            xgboost_1.6.0.1     
# [103] utf8_1.2.2           rmarkdown_2.14       GetoptLong_1.0.5    
# [106] usethis_2.1.3        data.table_1.14.2    callr_3.7.0         
# [109] ModelMetrics_1.2.2.2 digest_0.6.29        stats4_4.0.3        
# [112] signal_0.7-7         munsell_0.5.0        bslib_0.3.1         
# [115] sessioninfo_1.2.2</code></pre>
</div>
<div id="simulated-dataset" class="section level3" number="1.3.3">
<h3><span class="header-section-number">1.3.3</span> Simulated Dataset</h3>
<p>We simulated a dataset to demonstrate the principle of TSP normalization in GSClassifier:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="the-principle-of-gsclassifier.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Geneset</span></span>
<span id="cb3-2"><a href="the-principle-of-gsclassifier.html#cb3-2" aria-hidden="true" tabindex="-1"></a>geneSet <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb3-3"><a href="the-principle-of-gsclassifier.html#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Set1 =</span> <span class="fu">paste</span>(<span class="st">&#39;Gene&#39;</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="at">sep =</span> <span class="st">&#39;&#39;</span>),</span>
<span id="cb3-4"><a href="the-principle-of-gsclassifier.html#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Set2 =</span> <span class="fu">paste</span>(<span class="st">&#39;Gene&#39;</span>,<span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>,<span class="at">sep =</span> <span class="st">&#39;&#39;</span>)</span>
<span id="cb3-5"><a href="the-principle-of-gsclassifier.html#cb3-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-6"><a href="the-principle-of-gsclassifier.html#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="the-principle-of-gsclassifier.html#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression</span></span>
<span id="cb3-8"><a href="the-principle-of-gsclassifier.html#cb3-8" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">&#39;./data/simulated-data.xlsx&#39;</span>, <span class="at">sheet =</span> <span class="st">&#39;RNA&#39;</span>)</span>
<span id="cb3-9"><a href="the-principle-of-gsclassifier.html#cb3-9" aria-hidden="true" tabindex="-1"></a>expr0 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(x[,<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb3-10"><a href="the-principle-of-gsclassifier.html#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr0) <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">as.matrix</span>(x[,<span class="dv">1</span>])); <span class="fu">rm</span>(x)</span>
<span id="cb3-11"><a href="the-principle-of-gsclassifier.html#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="the-principle-of-gsclassifier.html#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Subtype information</span></span>
<span id="cb3-13"><a href="the-principle-of-gsclassifier.html#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co"># It depends on the application scenarios of GEPs</span></span>
<span id="cb3-14"><a href="the-principle-of-gsclassifier.html#cb3-14" aria-hidden="true" tabindex="-1"></a>subtype_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>)</span>
<span id="cb3-15"><a href="the-principle-of-gsclassifier.html#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Binned data for subtype 1</span></span>
<span id="cb3-16"><a href="the-principle-of-gsclassifier.html#cb3-16" aria-hidden="true" tabindex="-1"></a>Ybin <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(subtype_vector <span class="sc">==</span> <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb3-17"><a href="the-principle-of-gsclassifier.html#cb3-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-18"><a href="the-principle-of-gsclassifier.html#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Parameters</span></span>
<span id="cb3-19"><a href="the-principle-of-gsclassifier.html#cb3-19" aria-hidden="true" tabindex="-1"></a>breakVec <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="fl">1.0</span>)</span>
<span id="cb3-20"><a href="the-principle-of-gsclassifier.html#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="the-principle-of-gsclassifier.html#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb3-22"><a href="the-principle-of-gsclassifier.html#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">c</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>, <span class="st">&#39;Gene sets:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>))</span>
<span id="cb3-23"><a href="the-principle-of-gsclassifier.html#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(geneSet)</span>
<span id="cb3-24"><a href="the-principle-of-gsclassifier.html#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb3-25"><a href="the-principle-of-gsclassifier.html#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr0)</span>
<span id="cb3-26"><a href="the-principle-of-gsclassifier.html#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb3-27"><a href="the-principle-of-gsclassifier.html#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="co">#  Gene sets: </span></span>
<span id="cb3-28"><a href="the-principle-of-gsclassifier.html#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="co"># $Set1</span></span>
<span id="cb3-29"><a href="the-principle-of-gsclassifier.html#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot;</span></span>
<span id="cb3-30"><a href="the-principle-of-gsclassifier.html#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb3-31"><a href="the-principle-of-gsclassifier.html#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="co"># $Set2</span></span>
<span id="cb3-32"><a href="the-principle-of-gsclassifier.html#cb3-32" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span>
<span id="cb3-33"><a href="the-principle-of-gsclassifier.html#cb3-33" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb3-34"><a href="the-principle-of-gsclassifier.html#cb3-34" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression: </span></span>
<span id="cb3-35"><a href="the-principle-of-gsclassifier.html#cb3-35" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb3-36"><a href="the-principle-of-gsclassifier.html#cb3-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60    0.21    0.30    0.40</span></span>
<span id="cb3-37"><a href="the-principle-of-gsclassifier.html#cb3-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58    0.22    0.31    0.35</span></span>
<span id="cb3-38"><a href="the-principle-of-gsclassifier.html#cb3-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61      NA    0.29    0.30</span></span>
<span id="cb3-39"><a href="the-principle-of-gsclassifier.html#cb3-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40    0.51    0.52    0.60</span></span>
<span id="cb3-40"><a href="the-principle-of-gsclassifier.html#cb3-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35    0.52    0.54    0.58</span></span>
<span id="cb3-41"><a href="the-principle-of-gsclassifier.html#cb3-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30    0.53      NA    0.61</span></span>
<span id="cb3-42"><a href="the-principle-of-gsclassifier.html#cb3-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09    0.11    0.12    0.14</span></span></code></pre></div>
<p>Look at the matrix via heatmap:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="the-principle-of-gsclassifier.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(expr0))), <span class="at">name =</span> <span class="st">&quot;Z-score&quot;</span>)</span></code></pre></div>
<p><img src="Flowchart_files/figure-html/unnamed-chunk-4-1.png" width="60%" style="display: block; margin: auto;" /></p>
<p>This is an intersting dataset with features as following:</p>
<ul>
<li><p><strong>Distinguished gene sets</strong>: The expression profile between <strong>Gene 1-3</strong> and <strong>Gene 4-6</strong> is obviously different arross samples. Thus, these gene sets might represent different biology meaning.</p></li>
<li><p><strong>Stable gene</strong>: The expression level and rank of <strong>Gene 7</strong> seemed to be similar across samples. Thus, <strong>Gene 7</strong> might not be a robust marker for subtype modeling. Thus, it could help us to understand how filtering of <strong>GSClassifier</strong> works.</p></li>
<li><p><strong>Expression heterogeneity &amp; rank homogeneity</strong>: Take <strong>Sample1</strong> and <strong>Sample3</strong> as examples. The expression of <strong>Gene 1-6</strong> in <strong>Sample3</strong> seemed to be higher than those of <strong>Sample1</strong>. However, the expression of <strong>Gene 1-3</strong> is higher than <strong>Gene 4-6</strong> in both <strong>Sample1</strong> and <strong>Sample3</strong>, indicating similar bioprocess in these samples exists so that they should be classified as the same subtype.</p></li>
</ul>
</div>
<div id="missing-value-imputation-mvi" class="section level3" number="1.3.4">
<h3><span class="header-section-number">1.3.4</span> Missing value imputation (MVI)</h3>
<!--
Missing Value Imputation (MVI); 
+ Missing Value type: Missing Completely at Random (MCAR), Missing at Random (MAR), and Not Missing at Random (NMAR).
+ The main reason of missing value in the scenario of GSClassifier?
+ Why is missing value imputation needed in GSClassifier modeling?
+ Routine algorithms of missing value imputation for DNA array (mRNA expression)?
+ Missing value imputation in subtype calling?
+ The flowchart of quantile algorithm?
+ zero == Cutout; https://arxiv.org/abs/1708.04552; In this paper, we show that the simple regularization technique of randomly masking out square regions of input during training, which we call cutout, can be used to improve the robustness and overall performance of convolutional neural networks. Not only is this method extremely easy to implement, but we also demonstrate that it can be used in conjunction with existing forms of data augmentation and other regularizers to further improve model performance. We evaluate this method by applying it to current state-of-the-art architectures on the CIFAR-10, CIFAR-100, and SVHN datasets, yielding new state-of-the-art results of 2.56%, 15.20%, and 1.30% test error respectively.
+ Experiments: leave genes out of validation cohorts and then test the performance of GSClassifier-PADi.
-->
<p>Due to reasons like low expression/weak signal, contamination of microarray surfaces, inappropriate manual operations, insufficient resolution or systematic errors during the laboratory process <span class="citation">[<a href="references.html#ref-RN387" role="doc-biblioref">6</a>–<a href="references.html#ref-RN382" role="doc-biblioref">8</a>]</span>, <strong>missing value</strong> in high-imput genetic data is common. Generally, tiny missing value could be just dealed with case deletion, while the biological discovery might be damaged when the missing rate tops 15% <span class="citation">[<a href="references.html#ref-RN392" role="doc-biblioref">9</a>,<a href="references.html#ref-RN386" role="doc-biblioref">10</a>]</span>. Currently, lots of methods, including statistic-based or machine learning-based methods (Figure <a href="the-principle-of-gsclassifier.html#fig:mvi01">1.3</a>), had been developed for <strong>missing value imputation (MVI)</strong> <span class="citation">[<a href="references.html#ref-RN386" role="doc-biblioref">10</a>]</span>. Wang et al <span class="citation">[<a href="references.html#ref-RN384" role="doc-biblioref">11</a>]</span> categorized MVI methods into simple (zeros or average),biology knowledge-, global learning-, local learning-, hybrid-based methods. In order to satisfy the working conditions of xgboost <span class="citation">[<a href="references.html#ref-xgboost" role="doc-biblioref">12</a>]</span> functions (<code>xgb.train</code>, <code>xgboost</code>, and <code>xgb.cv</code>) in GSClassifer, the missing value in expression matrix must be deleted or imputation. In <strong>PAD</strong> project, several strategies were applied to reduce the impact of missing values as possible.</p>
<div class="figure" style="text-align: center"><span id="fig:mvi01"></span>
<img src="fig/mvi-01.png" alt="Missing value imputation methods reviewed by Hasan et al." width="90%" />
<p class="caption">
Figure 1.3: Missing value imputation methods reviewed by Hasan et al.
</p>
</div>
<p>First, both <strong>PIAM</strong> and <strong>PIDG</strong> in <strong>PAD</strong> project were curated GEPs that were not be missing in over 80% gastric cancer datasets. Here we showed the actual distribution of missing value across samples in gastric cancer datasets we used.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="the-principle-of-gsclassifier.html#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="the-principle-of-gsclassifier.html#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb5-3"><a href="the-principle-of-gsclassifier.html#cb5-3" aria-hidden="true" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;testData.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb5-4"><a href="the-principle-of-gsclassifier.html#cb5-4" aria-hidden="true" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb5-5"><a href="the-principle-of-gsclassifier.html#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="the-principle-of-gsclassifier.html#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Missing value</span></span>
<span id="cb5-7"><a href="the-principle-of-gsclassifier.html#cb5-7" aria-hidden="true" tabindex="-1"></a>expr_pad_na <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr_pad, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">sum</span>(<span class="fu">is.na</span>(x))<span class="sc">/</span><span class="fu">length</span>(x))</span>
<span id="cb5-8"><a href="the-principle-of-gsclassifier.html#cb5-8" aria-hidden="true" tabindex="-1"></a>expr_pad_na_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb5-9"><a href="the-principle-of-gsclassifier.html#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">names</span>(expr_pad_na),</span>
<span id="cb5-10"><a href="the-principle-of-gsclassifier.html#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">prob =</span> <span class="fu">as.numeric</span>(expr_pad_na),</span>
<span id="cb5-11"><a href="the-principle-of-gsclassifier.html#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">stringsAsFactors =</span> F</span>
<span id="cb5-12"><a href="the-principle-of-gsclassifier.html#cb5-12" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>As shown in Figure <a href="the-principle-of-gsclassifier.html#fig:mvi02">1.4</a>, the percentage of all samples in gastric cancer datasets we used is lower than 8%.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="the-principle-of-gsclassifier.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb6-2"><a href="the-principle-of-gsclassifier.html#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> expr_pad_na_df, </span>
<span id="cb6-3"><a href="the-principle-of-gsclassifier.html#cb6-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> prob)) <span class="sc">+</span> </span>
<span id="cb6-4"><a href="the-principle-of-gsclassifier.html#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&#39;identity&#39;</span>, <span class="at">color =</span> mycolor[<span class="dv">3</span>]) <span class="sc">+</span> </span>
<span id="cb6-5"><a href="the-principle-of-gsclassifier.html#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">labels=</span>scales<span class="sc">::</span>percent) <span class="sc">+</span> </span>
<span id="cb6-6"><a href="the-principle-of-gsclassifier.html#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Samples in gastric cancer cohorts&#39;</span>, </span>
<span id="cb6-7"><a href="the-principle-of-gsclassifier.html#cb6-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&#39;Percentage of missing value&#39;</span>) <span class="sc">+</span> </span>
<span id="cb6-8"><a href="the-principle-of-gsclassifier.html#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb6-9"><a href="the-principle-of-gsclassifier.html#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb6-10"><a href="the-principle-of-gsclassifier.html#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()</span>
<span id="cb6-11"><a href="the-principle-of-gsclassifier.html#cb6-11" aria-hidden="true" tabindex="-1"></a>  )</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:mvi02"></span>
<img src="Flowchart_files/figure-html/mvi02-1.png" alt="The distribution of missing value across gastric cancer samples." width="60%" />
<p class="caption">
Figure 1.4: The distribution of missing value across gastric cancer samples.
</p>
</div>
<p>Due to the high quality of our data, we just set missing value as zero during model training and subtype identification in the early version of PADi (<strong>PAD.train_v20200110</strong>). The model seemed to be robust in both the internal cohort and external cohorts, and greatly predicted the response to immune checkpoint inhibitors (ICIs) in advanced gastric cancer.</p>
<p>In the new version of PADi (<strong>PAD.train_v20220917</strong>), we designed the so-called <strong>quantile</strong> algorithm for random MVI during <strong>PADi</strong> model training, which also seemed to work well for PADi model training. Here, we demonstrated the use of <strong>quantile</strong> algorithm in the simulated dataset.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="the-principle-of-gsclassifier.html#cb7-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-2"><a href="the-principle-of-gsclassifier.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantile algorithm</span></span>
<span id="cb7-3"><a href="the-principle-of-gsclassifier.html#cb7-3" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> expr0</span>
<span id="cb7-4"><a href="the-principle-of-gsclassifier.html#cb7-4" aria-hidden="true" tabindex="-1"></a>na.pos <span class="ot">&lt;-</span> <span class="fu">apply</span>(expr,<span class="dv">2</span>,is.one.na)</span>
<span id="cb7-5"><a href="the-principle-of-gsclassifier.html#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">478</span>); seeds <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(expr)<span class="sc">*</span><span class="dv">10</span>, <span class="fu">sum</span>(na.pos), <span class="at">replace =</span> F)</span>
<span id="cb7-6"><a href="the-principle-of-gsclassifier.html#cb7-6" aria-hidden="true" tabindex="-1"></a>tSample <span class="ot">&lt;-</span> <span class="fu">names</span>(na.pos)[na.pos]</span>
<span id="cb7-7"><a href="the-principle-of-gsclassifier.html#cb7-7" aria-hidden="true" tabindex="-1"></a>quantile_vector <span class="ot">&lt;-</span> (<span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span>)<span class="sc">/</span><span class="dv">1000</span></span>
<span id="cb7-8"><a href="the-principle-of-gsclassifier.html#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(tSample)){ <span class="co"># i=1</span></span>
<span id="cb7-9"><a href="the-principle-of-gsclassifier.html#cb7-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb7-10"><a href="the-principle-of-gsclassifier.html#cb7-10" aria-hidden="true" tabindex="-1"></a>  sample.i <span class="ot">&lt;-</span> tSample[i]</span>
<span id="cb7-11"><a href="the-principle-of-gsclassifier.html#cb7-11" aria-hidden="true" tabindex="-1"></a>  expr.i <span class="ot">&lt;-</span> expr[, sample.i]</span>
<span id="cb7-12"><a href="the-principle-of-gsclassifier.html#cb7-12" aria-hidden="true" tabindex="-1"></a>  expr.i.max <span class="ot">&lt;-</span> <span class="fu">max</span>(expr.i, <span class="at">na.rm =</span> T)</span>
<span id="cb7-13"><a href="the-principle-of-gsclassifier.html#cb7-13" aria-hidden="true" tabindex="-1"></a>  expr.i.min <span class="ot">&lt;-</span> <span class="fu">min</span>(expr.i, <span class="at">na.rm =</span> T)</span>
<span id="cb7-14"><a href="the-principle-of-gsclassifier.html#cb7-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(seeds[i]);</span>
<span id="cb7-15"><a href="the-principle-of-gsclassifier.html#cb7-15" aria-hidden="true" tabindex="-1"></a>  expr.i[<span class="fu">is.na</span>(expr.i)] <span class="ot">&lt;-</span></span>
<span id="cb7-16"><a href="the-principle-of-gsclassifier.html#cb7-16" aria-hidden="true" tabindex="-1"></a>    expr.i.min <span class="sc">+</span></span>
<span id="cb7-17"><a href="the-principle-of-gsclassifier.html#cb7-17" aria-hidden="true" tabindex="-1"></a>    (expr.i.max<span class="sc">-</span>expr.i.min) <span class="sc">*</span> <span class="fu">sample</span>(quantile_vector,</span>
<span id="cb7-18"><a href="the-principle-of-gsclassifier.html#cb7-18" aria-hidden="true" tabindex="-1"></a>                                     <span class="fu">sum</span>(<span class="fu">is.na</span>(expr.i)),</span>
<span id="cb7-19"><a href="the-principle-of-gsclassifier.html#cb7-19" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">replace =</span> T)</span>
<span id="cb7-20"><a href="the-principle-of-gsclassifier.html#cb7-20" aria-hidden="true" tabindex="-1"></a>  expr[, sample.i] <span class="ot">&lt;-</span> expr.i</span>
<span id="cb7-21"><a href="the-principle-of-gsclassifier.html#cb7-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-22"><a href="the-principle-of-gsclassifier.html#cb7-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb7-23"><a href="the-principle-of-gsclassifier.html#cb7-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-24"><a href="the-principle-of-gsclassifier.html#cb7-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb7-25"><a href="the-principle-of-gsclassifier.html#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb7-26"><a href="the-principle-of-gsclassifier.html#cb7-26" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr0)</span>
<span id="cb7-27"><a href="the-principle-of-gsclassifier.html#cb7-27" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb7-28"><a href="the-principle-of-gsclassifier.html#cb7-28" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;RNA expression without NA value:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb7-29"><a href="the-principle-of-gsclassifier.html#cb7-29" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr)</span>
<span id="cb7-30"><a href="the-principle-of-gsclassifier.html#cb7-30" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression: </span></span>
<span id="cb7-31"><a href="the-principle-of-gsclassifier.html#cb7-31" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb7-32"><a href="the-principle-of-gsclassifier.html#cb7-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60    0.21    0.30    0.40</span></span>
<span id="cb7-33"><a href="the-principle-of-gsclassifier.html#cb7-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58    0.22    0.31    0.35</span></span>
<span id="cb7-34"><a href="the-principle-of-gsclassifier.html#cb7-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61      NA    0.29    0.30</span></span>
<span id="cb7-35"><a href="the-principle-of-gsclassifier.html#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40    0.51    0.52    0.60</span></span>
<span id="cb7-36"><a href="the-principle-of-gsclassifier.html#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35    0.52    0.54    0.58</span></span>
<span id="cb7-37"><a href="the-principle-of-gsclassifier.html#cb7-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30    0.53      NA    0.61</span></span>
<span id="cb7-38"><a href="the-principle-of-gsclassifier.html#cb7-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09    0.11    0.12    0.14</span></span>
<span id="cb7-39"><a href="the-principle-of-gsclassifier.html#cb7-39" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb7-40"><a href="the-principle-of-gsclassifier.html#cb7-40" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA expression without NA value: </span></span>
<span id="cb7-41"><a href="the-principle-of-gsclassifier.html#cb7-41" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb7-42"><a href="the-principle-of-gsclassifier.html#cb7-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60 0.21000 0.30000    0.40</span></span>
<span id="cb7-43"><a href="the-principle-of-gsclassifier.html#cb7-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58 0.22000 0.31000    0.35</span></span>
<span id="cb7-44"><a href="the-principle-of-gsclassifier.html#cb7-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61 0.43256 0.29000    0.30</span></span>
<span id="cb7-45"><a href="the-principle-of-gsclassifier.html#cb7-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40 0.51000 0.52000    0.60</span></span>
<span id="cb7-46"><a href="the-principle-of-gsclassifier.html#cb7-46" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35 0.52000 0.54000    0.58</span></span>
<span id="cb7-47"><a href="the-principle-of-gsclassifier.html#cb7-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30 0.53000 0.32622    0.61</span></span>
<span id="cb7-48"><a href="the-principle-of-gsclassifier.html#cb7-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7    0.10    0.12    0.09 0.11000 0.12000    0.14</span></span></code></pre></div>
<p>Look at the new matrix via heatmap, where the clustering result is not obviously disturbed after MVI:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="the-principle-of-gsclassifier.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(<span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(expr))), <span class="at">name =</span> <span class="st">&quot;Z-score&quot;</span>)</span></code></pre></div>
<p><img src="Flowchart_files/figure-html/unnamed-chunk-7-1.png" width="60%" style="display: block; margin: auto;" /></p>
<p>Next, we explored how the number of missing value impacts subtype identification by <strong>PADi</strong>. First, we used quantile algorithm to do MVI in the internal validation cohort of gastric cancer. Second, we randomly masked different proportion of genes as zero expression. Third, we calculated the relative multi-ROC (masked data vs. MVI data). We had developed a function called <strong>mv_tolerance</strong> in <strong>GSClassifier</strong> to do this task.</p>
<p>Load the internal validation cohort:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="the-principle-of-gsclassifier.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Internal validation cohort</span></span>
<span id="cb9-2"><a href="the-principle-of-gsclassifier.html#cb9-2" aria-hidden="true" tabindex="-1"></a>testData <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;testData.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;GSClassifier&quot;</span>))</span>
<span id="cb9-3"><a href="the-principle-of-gsclassifier.html#cb9-3" aria-hidden="true" tabindex="-1"></a>expr_pad <span class="ot">&lt;-</span> testData<span class="sc">$</span>PanSTAD_expr_part</span>
<span id="cb9-4"><a href="the-principle-of-gsclassifier.html#cb9-4" aria-hidden="true" tabindex="-1"></a>modelInfo <span class="ot">&lt;-</span> <span class="fu">modelData</span>(</span>
<span id="cb9-5"><a href="the-principle-of-gsclassifier.html#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> testData<span class="sc">$</span>PanSTAD_phenotype_part,</span>
<span id="cb9-6"><a href="the-principle-of-gsclassifier.html#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">id.col =</span> <span class="st">&quot;ID&quot;</span>,</span>
<span id="cb9-7"><a href="the-principle-of-gsclassifier.html#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">variable =</span> <span class="fu">c</span>(<span class="st">&quot;platform&quot;</span>, <span class="st">&quot;PAD_subtype&quot;</span>),</span>
<span id="cb9-8"><a href="the-principle-of-gsclassifier.html#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">Prop =</span> <span class="fl">0.7</span>,</span>
<span id="cb9-9"><a href="the-principle-of-gsclassifier.html#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">19871</span></span>
<span id="cb9-10"><a href="the-principle-of-gsclassifier.html#cb9-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-11"><a href="the-principle-of-gsclassifier.html#cb9-11" aria-hidden="true" tabindex="-1"></a>validInform <span class="ot">&lt;-</span> modelInfo<span class="sc">$</span>Data<span class="sc">$</span>Valid</span>
<span id="cb9-12"><a href="the-principle-of-gsclassifier.html#cb9-12" aria-hidden="true" tabindex="-1"></a>expr_pad_innervalid <span class="ot">&lt;-</span> expr_pad[,validInform<span class="sc">$</span>ID]</span></code></pre></div>
<p>Missing value tolerance analysis:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="the-principle-of-gsclassifier.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Time-consuming</span></span>
<span id="cb10-2"><a href="the-principle-of-gsclassifier.html#cb10-2" aria-hidden="true" tabindex="-1"></a>mvt <span class="ot">&lt;-</span> <span class="fu">mv_tolerance</span>(</span>
<span id="cb10-3"><a href="the-principle-of-gsclassifier.html#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">X =</span> expr_pad_innervalid,</span>
<span id="cb10-4"><a href="the-principle-of-gsclassifier.html#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.loss =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">12</span>),</span>
<span id="cb10-5"><a href="the-principle-of-gsclassifier.html#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>),</span>
<span id="cb10-6"><a href="the-principle-of-gsclassifier.html#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;PAD.train_20220916&quot;</span>,</span>
<span id="cb10-7"><a href="the-principle-of-gsclassifier.html#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">487</span>,</span>
<span id="cb10-8"><a href="the-principle-of-gsclassifier.html#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> T</span>
<span id="cb10-9"><a href="the-principle-of-gsclassifier.html#cb10-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Prepare the data of multi-ROC analysis:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="the-principle-of-gsclassifier.html#cb11-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-2"><a href="the-principle-of-gsclassifier.html#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb11-3"><a href="the-principle-of-gsclassifier.html#cb11-3" aria-hidden="true" tabindex="-1"></a>mvt_auc <span class="ot">&lt;-</span> mvt<span class="sc">$</span>multiAUC</span>
<span id="cb11-4"><a href="the-principle-of-gsclassifier.html#cb11-4" aria-hidden="true" tabindex="-1"></a>mvt_auc_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb11-5"><a href="the-principle-of-gsclassifier.html#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(mvt_auc)){ <span class="co"># i=1</span></span>
<span id="cb11-6"><a href="the-principle-of-gsclassifier.html#cb11-6" aria-hidden="true" tabindex="-1"></a>  df.i <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-7"><a href="the-principle-of-gsclassifier.html#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">as.integer</span>(<span class="fu">Fastextra</span>(<span class="fu">names</span>(mvt_auc)[i], <span class="st">&#39;=&#39;</span>, <span class="dv">2</span>)),</span>
<span id="cb11-8"><a href="the-principle-of-gsclassifier.html#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">as.numeric</span>(mvt_auc[[i]]<span class="sc">$</span>auc),</span>
<span id="cb11-9"><a href="the-principle-of-gsclassifier.html#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">stringsAsFactors =</span> F</span>
<span id="cb11-10"><a href="the-principle-of-gsclassifier.html#cb11-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb11-11"><a href="the-principle-of-gsclassifier.html#cb11-11" aria-hidden="true" tabindex="-1"></a>  mvt_auc_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(mvt_auc_df, df.i)</span>
<span id="cb11-12"><a href="the-principle-of-gsclassifier.html#cb11-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-13"><a href="the-principle-of-gsclassifier.html#cb11-13" aria-hidden="true" tabindex="-1"></a>  </span></code></pre></div>
<p>Plot the missing value tolerance:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="the-principle-of-gsclassifier.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mvt_auc_df, <span class="fu">aes</span>(x,y)) <span class="sc">+</span></span>
<span id="cb12-2"><a href="the-principle-of-gsclassifier.html#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb12-3"><a href="the-principle-of-gsclassifier.html#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">12</span>)) <span class="sc">+</span> </span>
<span id="cb12-4"><a href="the-principle-of-gsclassifier.html#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_smooth</span>(<span class="at">formula =</span> y <span class="sc">~</span> x,<span class="at">method =</span> <span class="st">&#39;glm&#39;</span>) <span class="sc">+</span></span>
<span id="cb12-5"><a href="the-principle-of-gsclassifier.html#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;No. of missing value&#39;</span>, </span>
<span id="cb12-6"><a href="the-principle-of-gsclassifier.html#cb12-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span><span class="st">&#39;Relative AUC in multi-ROC analysis&#39;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:mvi03"></span>
<img src="Flowchart_files/figure-html/mvi03-1.png" alt="The association between the number of missing value and subtype identification performance." width="60%" />
<p class="caption">
Figure 1.5: The association between the number of missing value and subtype identification performance.
</p>
</div>
<p>As showed in Figure <a href="the-principle-of-gsclassifier.html#fig:mvi03">1.5</a>, there is linear negative correlation between the number of missing value (mising rate ranges from 6.25% to 37.5%) and subtype identification performance of <strong>PADi</strong> model. One of the reasons might be that PIAM/PIDG were small gene sets, so little gene loss might significantly impact the performance of <strong>PADi</strong>. By the way, there is no mising value for <strong>PADi</strong> in the ‘Kim2018’ cohort, an external validation cohort for ICIs therapy response. Nonetheless, we still used <strong>zero strategy</strong> during subtype identification of <strong>PADi</strong> if any missing value exist, because randomization might make the result unstable, which is not suitable for clinical decision.</p>
<p>In conclusion, zero or quantile strategy could be applied for MVI before GSClassifier model training. However, missing value should be avoid as possible in subtype identification for missing value really damage the performance of GSClassifier. Nonetheless, due to low-input gene sets used in <strong>PADi</strong> model (No. of Gene=32), it’s easy to avoid missing value in clinical practice.</p>
</div>
<div id="binned-expression" class="section level3" number="1.3.5">
<h3><span class="header-section-number">1.3.5</span> Binned expression</h3>
<p>First, we binned genes with diffrent quantile intervals so that the distribution of rank information could be more consistent across samples.</p>
<p>Take <strong>Sample4</strong> as an example:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="the-principle-of-gsclassifier.html#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="the-principle-of-gsclassifier.html#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Data of Sample4</span></span>
<span id="cb13-3"><a href="the-principle-of-gsclassifier.html#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="the-principle-of-gsclassifier.html#cb13-4" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> expr[,<span class="dv">4</span>]</span>
<span id="cb13-5"><a href="the-principle-of-gsclassifier.html#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="the-principle-of-gsclassifier.html#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create quantiles  </span></span>
<span id="cb13-7"><a href="the-principle-of-gsclassifier.html#cb13-7" aria-hidden="true" tabindex="-1"></a>brks <span class="ot">&lt;-</span> <span class="fu">quantile</span>(<span class="fu">as.numeric</span>(x), </span>
<span id="cb13-8"><a href="the-principle-of-gsclassifier.html#cb13-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">probs=</span>breakVec, </span>
<span id="cb13-9"><a href="the-principle-of-gsclassifier.html#cb13-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">na.rm =</span> T)</span>
<span id="cb13-10"><a href="the-principle-of-gsclassifier.html#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="the-principle-of-gsclassifier.html#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Get interval orders</span></span>
<span id="cb13-12"><a href="the-principle-of-gsclassifier.html#cb13-12" aria-hidden="true" tabindex="-1"></a>xbin <span class="ot">&lt;-</span> <span class="fu">.bincode</span>(<span class="at">x =</span> x, </span>
<span id="cb13-13"><a href="the-principle-of-gsclassifier.html#cb13-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">breaks =</span> brks, </span>
<span id="cb13-14"><a href="the-principle-of-gsclassifier.html#cb13-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">include.lowest =</span> T)</span>
<span id="cb13-15"><a href="the-principle-of-gsclassifier.html#cb13-15" aria-hidden="true" tabindex="-1"></a>xbin <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(xbin)</span>
<span id="cb13-16"><a href="the-principle-of-gsclassifier.html#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(xbin) <span class="ot">&lt;-</span> <span class="fu">names</span>(x)</span>
<span id="cb13-17"><a href="the-principle-of-gsclassifier.html#cb13-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-18"><a href="the-principle-of-gsclassifier.html#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb13-19"><a href="the-principle-of-gsclassifier.html#cb13-19" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Quantiles:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>); <span class="fu">print</span>(brks)</span>
<span id="cb13-20"><a href="the-principle-of-gsclassifier.html#cb13-20" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb13-21"><a href="the-principle-of-gsclassifier.html#cb13-21" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Raw expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>);<span class="fu">print</span>(x)</span>
<span id="cb13-22"><a href="the-principle-of-gsclassifier.html#cb13-22" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb13-23"><a href="the-principle-of-gsclassifier.html#cb13-23" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Binned expression:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>); <span class="fu">print</span>(xbin)</span>
<span id="cb13-24"><a href="the-principle-of-gsclassifier.html#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Quantiles: </span></span>
<span id="cb13-25"><a href="the-principle-of-gsclassifier.html#cb13-25" aria-hidden="true" tabindex="-1"></a><span class="co">#      0%     25%     50%     75%    100% </span></span>
<span id="cb13-26"><a href="the-principle-of-gsclassifier.html#cb13-26" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.11000 0.21500 0.43256 0.51500 0.53000 </span></span>
<span id="cb13-27"><a href="the-principle-of-gsclassifier.html#cb13-27" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb13-28"><a href="the-principle-of-gsclassifier.html#cb13-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Raw expression: </span></span>
<span id="cb13-29"><a href="the-principle-of-gsclassifier.html#cb13-29" aria-hidden="true" tabindex="-1"></a><span class="co">#   Gene1   Gene2   Gene3   Gene4   Gene5   Gene6   Gene7 </span></span>
<span id="cb13-30"><a href="the-principle-of-gsclassifier.html#cb13-30" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.21000 0.22000 0.43256 0.51000 0.52000 0.53000 0.11000 </span></span>
<span id="cb13-31"><a href="the-principle-of-gsclassifier.html#cb13-31" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb13-32"><a href="the-principle-of-gsclassifier.html#cb13-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Binned expression: </span></span>
<span id="cb13-33"><a href="the-principle-of-gsclassifier.html#cb13-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1 Gene2 Gene3 Gene4 Gene5 Gene6 Gene7 </span></span>
<span id="cb13-34"><a href="the-principle-of-gsclassifier.html#cb13-34" aria-hidden="true" tabindex="-1"></a><span class="co">#     1     2     2     3     4     4     1</span></span></code></pre></div>
<p>For example, <strong>0.110</strong> is the minimun of the raw expression vector, so its binned expression is <strong>1</strong>. Similarly, the binned expression of maximum <strong>0.530</strong> is <strong>4</strong>.</p>
<p>Generally, we calculate binned expression via function <strong>breakBin</strong> of <strong>GSClassifier</strong>:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="the-principle-of-gsclassifier.html#cb14-1" aria-hidden="true" tabindex="-1"></a>expr_binned <span class="ot">&lt;-</span> <span class="fu">apply</span>(</span>
<span id="cb14-2"><a href="the-principle-of-gsclassifier.html#cb14-2" aria-hidden="true" tabindex="-1"></a>  expr, <span class="dv">2</span>, </span>
<span id="cb14-3"><a href="the-principle-of-gsclassifier.html#cb14-3" aria-hidden="true" tabindex="-1"></a>  GSClassifier<span class="sc">:::</span>breakBin,</span>
<span id="cb14-4"><a href="the-principle-of-gsclassifier.html#cb14-4" aria-hidden="true" tabindex="-1"></a>  breakVec)</span>
<span id="cb14-5"><a href="the-principle-of-gsclassifier.html#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr_binned) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(expr)</span>
<span id="cb14-6"><a href="the-principle-of-gsclassifier.html#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr_binned)</span>
<span id="cb14-7"><a href="the-principle-of-gsclassifier.html#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb14-8"><a href="the-principle-of-gsclassifier.html#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1       3       3       4       1       2       2</span></span>
<span id="cb14-9"><a href="the-principle-of-gsclassifier.html#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2       4       4       3       2       2       2</span></span>
<span id="cb14-10"><a href="the-principle-of-gsclassifier.html#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3       4       4       4       2       1       1</span></span>
<span id="cb14-11"><a href="the-principle-of-gsclassifier.html#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4       1       2       2       3       4       4</span></span>
<span id="cb14-12"><a href="the-principle-of-gsclassifier.html#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5       2       2       2       4       4       3</span></span>
<span id="cb14-13"><a href="the-principle-of-gsclassifier.html#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6       2       1       1       4       3       4</span></span>
<span id="cb14-14"><a href="the-principle-of-gsclassifier.html#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7       1       1       1       1       1       1</span></span></code></pre></div>
<p>In this simulated dataset, <strong>Gene7</strong> is a gene whose expression is always the lowest across all samples. In other words, the rank of <strong>Gene7</strong> is stable or invariable across samples so that it’s not robust for identification of differentail subtypes.</p>
<p>Except binned expression, we also calculated pair difference later. Due to the number of gene pair is <span class="math inline">\(C_{2 \atop n}\)</span>, the removement of genes like <strong>Gene7</strong> before modeling could really reduce the complexibility and save computing resources. In all, genes with low rank difference should be dropped out in some extent in <strong>GSClassifier</strong>.</p>
<p>First, We use <strong>base::rank</strong> to return the sample ranks of the values in a vector:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="the-principle-of-gsclassifier.html#cb15-1" aria-hidden="true" tabindex="-1"></a>expr_binned_rank <span class="ot">&lt;-</span> <span class="fu">apply</span>(</span>
<span id="cb15-2"><a href="the-principle-of-gsclassifier.html#cb15-2" aria-hidden="true" tabindex="-1"></a>  expr_binned, <span class="dv">2</span>, </span>
<span id="cb15-3"><a href="the-principle-of-gsclassifier.html#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x)<span class="fu">rank</span>(x, <span class="at">na.last =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-4"><a href="the-principle-of-gsclassifier.html#cb15-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-5"><a href="the-principle-of-gsclassifier.html#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr_binned_rank)</span>
<span id="cb15-6"><a href="the-principle-of-gsclassifier.html#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb15-7"><a href="the-principle-of-gsclassifier.html#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1     5.0     5.0     6.5     1.5     3.5     3.5</span></span>
<span id="cb15-8"><a href="the-principle-of-gsclassifier.html#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2     6.5     6.5     5.0     3.5     3.5     3.5</span></span>
<span id="cb15-9"><a href="the-principle-of-gsclassifier.html#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3     6.5     6.5     6.5     3.5     1.5     1.5</span></span>
<span id="cb15-10"><a href="the-principle-of-gsclassifier.html#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4     1.5     3.5     3.5     5.0     6.5     6.5</span></span>
<span id="cb15-11"><a href="the-principle-of-gsclassifier.html#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5     3.5     3.5     3.5     6.5     6.5     5.0</span></span>
<span id="cb15-12"><a href="the-principle-of-gsclassifier.html#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6     3.5     1.5     1.5     6.5     5.0     6.5</span></span>
<span id="cb15-13"><a href="the-principle-of-gsclassifier.html#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene7     1.5     1.5     1.5     1.5     1.5     1.5</span></span></code></pre></div>
<p>Then, get weighted average rank difference of each gene based on specified subtype distribution (<strong>Ybin</strong>):</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="the-principle-of-gsclassifier.html#cb16-1" aria-hidden="true" tabindex="-1"></a>testRes <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb16-2"><a href="the-principle-of-gsclassifier.html#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(expr_binned_rank), </span>
<span id="cb16-3"><a href="the-principle-of-gsclassifier.html#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(gi){</span>
<span id="cb16-4"><a href="the-principle-of-gsclassifier.html#cb16-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-5"><a href="the-principle-of-gsclassifier.html#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Rank vector of each gene</span></span>
<span id="cb16-6"><a href="the-principle-of-gsclassifier.html#cb16-6" aria-hidden="true" tabindex="-1"></a>    rankg <span class="ot">=</span> expr_binned_rank[gi,];</span>
<span id="cb16-7"><a href="the-principle-of-gsclassifier.html#cb16-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-8"><a href="the-principle-of-gsclassifier.html#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Weighted average rank difference of a gene for specified subtype </span></span>
<span id="cb16-9"><a href="the-principle-of-gsclassifier.html#cb16-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Here is subtype 1 vs. others</span></span>
<span id="cb16-10"><a href="the-principle-of-gsclassifier.html#cb16-10" aria-hidden="true" tabindex="-1"></a>    (<span class="fu">sum</span>(rankg[Ybin <span class="sc">==</span> <span class="dv">0</span>], <span class="at">na.rm =</span> T) <span class="sc">/</span> <span class="fu">sum</span>(Ybin <span class="sc">==</span> <span class="dv">0</span>, <span class="at">na.rm =</span> T)) <span class="sc">-</span> </span>
<span id="cb16-11"><a href="the-principle-of-gsclassifier.html#cb16-11" aria-hidden="true" tabindex="-1"></a>    (<span class="fu">sum</span>(rankg[Ybin <span class="sc">==</span> <span class="dv">1</span>], <span class="at">na.rm =</span> T) <span class="sc">/</span> <span class="fu">sum</span>(Ybin <span class="sc">==</span> <span class="dv">1</span>, <span class="at">na.rm =</span> T))</span>
<span id="cb16-12"><a href="the-principle-of-gsclassifier.html#cb16-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-13"><a href="the-principle-of-gsclassifier.html#cb16-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-14"><a href="the-principle-of-gsclassifier.html#cb16-14" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(testRes) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(expr_binned_rank)</span>
<span id="cb16-15"><a href="the-principle-of-gsclassifier.html#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(testRes)</span>
<span id="cb16-16"><a href="the-principle-of-gsclassifier.html#cb16-16" aria-hidden="true" tabindex="-1"></a><span class="co">#     Gene1     Gene2     Gene3     Gene4     Gene5     Gene6     Gene7 </span></span>
<span id="cb16-17"><a href="the-principle-of-gsclassifier.html#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="co"># -2.666667 -2.500000 -4.333333  3.166667  2.500000  3.833333  0.000000</span></span></code></pre></div>
<p><strong>Gene7</strong> is the one with the lowest absolute value (0) of rank diffrence. By the way, <strong>Gene 1-3</strong> have the same direction (&lt;0), so do <strong>Gene 4-6</strong> (&gt;0), which indicates the nature of clustering based on these two gene sets.</p>
<p>In practice, we use <strong>ptail</strong> to select differential genes based on rank diffrences. <strong>Smaller ptail is, less gene kept</strong>. Here, we just set <strong>ptail=0.4</strong>:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="the-principle-of-gsclassifier.html#cb17-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-2"><a href="the-principle-of-gsclassifier.html#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ptail is a numeber ranging (0,0.5].</span></span>
<span id="cb17-3"><a href="the-principle-of-gsclassifier.html#cb17-3" aria-hidden="true" tabindex="-1"></a>ptail <span class="ot">=</span> <span class="fl">0.4</span></span>
<span id="cb17-4"><a href="the-principle-of-gsclassifier.html#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="the-principle-of-gsclassifier.html#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Index of target genes with big rank differences</span></span>
<span id="cb17-6"><a href="the-principle-of-gsclassifier.html#cb17-6" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">which</span>((testRes <span class="sc">&lt;</span> <span class="fu">quantile</span>(testRes, ptail, <span class="at">na.rm =</span> T)) <span class="sc">|</span> </span>
<span id="cb17-7"><a href="the-principle-of-gsclassifier.html#cb17-7" aria-hidden="true" tabindex="-1"></a>             (testRes <span class="sc">&gt;</span> <span class="fu">quantile</span>(testRes, <span class="fl">1.0</span><span class="sc">-</span>ptail, <span class="at">na.rm =</span> T)))</span>
<span id="cb17-8"><a href="the-principle-of-gsclassifier.html#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="the-principle-of-gsclassifier.html#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Target genes</span></span>
<span id="cb17-10"><a href="the-principle-of-gsclassifier.html#cb17-10" aria-hidden="true" tabindex="-1"></a>gene_bigRank <span class="ot">&lt;-</span> <span class="fu">names</span>(testRes)[idx]</span>
<span id="cb17-11"><a href="the-principle-of-gsclassifier.html#cb17-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-12"><a href="the-principle-of-gsclassifier.html#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb17-13"><a href="the-principle-of-gsclassifier.html#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Index of target genes: &#39;</span>,<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>);<span class="fu">print</span>(idx); <span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb17-14"><a href="the-principle-of-gsclassifier.html#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Target genes:&#39;</span>,<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>);<span class="fu">print</span>(gene_bigRank)</span>
<span id="cb17-15"><a href="the-principle-of-gsclassifier.html#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Index of target genes:  </span></span>
<span id="cb17-16"><a href="the-principle-of-gsclassifier.html#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1 Gene2 Gene3 Gene4 Gene5 Gene6 </span></span>
<span id="cb17-17"><a href="the-principle-of-gsclassifier.html#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="co">#     1     2     3     4     5     6 </span></span>
<span id="cb17-18"><a href="the-principle-of-gsclassifier.html#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb17-19"><a href="the-principle-of-gsclassifier.html#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Target genes: </span></span>
<span id="cb17-20"><a href="the-principle-of-gsclassifier.html#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot; &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span></code></pre></div>
<p>Hence, <strong>Gene7</strong> was filtered and excluded in the following analysis. By the way, both <strong>ptail</strong> and <strong>breakVec</strong> are hyperparameters in GSClassifier modeling.</p>
</div>
<div id="pair-difference" class="section level3" number="1.3.6">
<h3><span class="header-section-number">1.3.6</span> Pair difference</h3>
<p>In GSClassifier, we use a ensemble function <strong>featureSelection</strong> to select data for pair difference scoring.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="the-principle-of-gsclassifier.html#cb18-1" aria-hidden="true" tabindex="-1"></a>expr_feat <span class="ot">&lt;-</span> <span class="fu">featureSelection</span>(expr, Ybin,</span>
<span id="cb18-2"><a href="the-principle-of-gsclassifier.html#cb18-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">testRes =</span> testRes,</span>
<span id="cb18-3"><a href="the-principle-of-gsclassifier.html#cb18-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ptail =</span> <span class="fl">0.4</span>)</span>
<span id="cb18-4"><a href="the-principle-of-gsclassifier.html#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="the-principle-of-gsclassifier.html#cb18-5" aria-hidden="true" tabindex="-1"></a>expr_sub <span class="ot">&lt;-</span> expr_feat<span class="sc">$</span>Xsub</span>
<span id="cb18-6"><a href="the-principle-of-gsclassifier.html#cb18-6" aria-hidden="true" tabindex="-1"></a>gene_bigRank <span class="ot">&lt;-</span> expr_feat<span class="sc">$</span>Genes</span>
<span id="cb18-7"><a href="the-principle-of-gsclassifier.html#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="the-principle-of-gsclassifier.html#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb18-9"><a href="the-principle-of-gsclassifier.html#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Raw xpression without NA:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb18-10"><a href="the-principle-of-gsclassifier.html#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(expr_sub)</span>
<span id="cb18-11"><a href="the-principle-of-gsclassifier.html#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb18-12"><a href="the-principle-of-gsclassifier.html#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Genes with large rank diff:&#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb18-13"><a href="the-principle-of-gsclassifier.html#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene_bigRank)</span>
<span id="cb18-14"><a href="the-principle-of-gsclassifier.html#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Raw xpression without NA: </span></span>
<span id="cb18-15"><a href="the-principle-of-gsclassifier.html#cb18-15" aria-hidden="true" tabindex="-1"></a><span class="co">#       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb18-16"><a href="the-principle-of-gsclassifier.html#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1    0.51    0.52    0.60 0.21000 0.30000    0.40</span></span>
<span id="cb18-17"><a href="the-principle-of-gsclassifier.html#cb18-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2    0.52    0.54    0.58 0.22000 0.31000    0.35</span></span>
<span id="cb18-18"><a href="the-principle-of-gsclassifier.html#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3    0.53    0.60    0.61 0.43256 0.29000    0.30</span></span>
<span id="cb18-19"><a href="the-principle-of-gsclassifier.html#cb18-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4    0.21    0.30    0.40 0.51000 0.52000    0.60</span></span>
<span id="cb18-20"><a href="the-principle-of-gsclassifier.html#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5    0.22    0.31    0.35 0.52000 0.54000    0.58</span></span>
<span id="cb18-21"><a href="the-principle-of-gsclassifier.html#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6    0.23    0.29    0.30 0.53000 0.32622    0.61</span></span>
<span id="cb18-22"><a href="the-principle-of-gsclassifier.html#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb18-23"><a href="the-principle-of-gsclassifier.html#cb18-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Genes with large rank diff: </span></span>
<span id="cb18-24"><a href="the-principle-of-gsclassifier.html#cb18-24" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot; &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span></code></pre></div>
<p>In GSClassifier, we use function <strong>makeGenePairs</strong> to calculate s</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="the-principle-of-gsclassifier.html#cb19-1" aria-hidden="true" tabindex="-1"></a>gene_bigRank_pairs <span class="ot">&lt;-</span> GSClassifier<span class="sc">:::</span><span class="fu">makeGenePairs</span>(</span>
<span id="cb19-2"><a href="the-principle-of-gsclassifier.html#cb19-2" aria-hidden="true" tabindex="-1"></a>  gene_bigRank, </span>
<span id="cb19-3"><a href="the-principle-of-gsclassifier.html#cb19-3" aria-hidden="true" tabindex="-1"></a>  expr[gene_bigRank,])</span>
<span id="cb19-4"><a href="the-principle-of-gsclassifier.html#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene_bigRank_pairs)</span>
<span id="cb19-5"><a href="the-principle-of-gsclassifier.html#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">#             Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb19-6"><a href="the-principle-of-gsclassifier.html#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene2       0       0       1       0       0       1</span></span>
<span id="cb19-7"><a href="the-principle-of-gsclassifier.html#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene3       0       0       0       0       1       1</span></span>
<span id="cb19-8"><a href="the-principle-of-gsclassifier.html#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene4       1       1       1       0       0       0</span></span>
<span id="cb19-9"><a href="the-principle-of-gsclassifier.html#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene5       1       1       1       0       0       0</span></span>
<span id="cb19-10"><a href="the-principle-of-gsclassifier.html#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene6       1       1       1       0       0       0</span></span>
<span id="cb19-11"><a href="the-principle-of-gsclassifier.html#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene3       0       0       0       0       1       1</span></span>
<span id="cb19-12"><a href="the-principle-of-gsclassifier.html#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene4       1       1       1       0       0       0</span></span>
<span id="cb19-13"><a href="the-principle-of-gsclassifier.html#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene5       1       1       1       0       0       0</span></span>
<span id="cb19-14"><a href="the-principle-of-gsclassifier.html#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene6       1       1       1       0       0       0</span></span>
<span id="cb19-15"><a href="the-principle-of-gsclassifier.html#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene4       1       1       1       0       0       0</span></span>
<span id="cb19-16"><a href="the-principle-of-gsclassifier.html#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene5       1       1       1       0       0       0</span></span>
<span id="cb19-17"><a href="the-principle-of-gsclassifier.html#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene6       1       1       1       0       0       0</span></span>
<span id="cb19-18"><a href="the-principle-of-gsclassifier.html#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene5       0       0       1       0       0       1</span></span>
<span id="cb19-19"><a href="the-principle-of-gsclassifier.html#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene6       0       1       1       0       1       0</span></span>
<span id="cb19-20"><a href="the-principle-of-gsclassifier.html#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5:Gene6       0       1       1       0       1       0</span></span></code></pre></div>
<p>Take <strong>Gene1:Gene4</strong> of <strong>Sample1</strong> as an example. <span class="math inline">\(Expression_{Gene1} - Expression_{Gene4} = 0.51-0.21 = 0.3 &gt; 0\)</span>, so the pair score is 1. If the difference is less than or equal to 0, the pair score is 0. In addition, the difference of gene pair scoring between <strong>Sample 1-3</strong> and <strong>Sample 4-6</strong> is obivous, revealing the robustness of pair difference for subtype identification.</p>
</div>
<div id="set-difference" class="section level3" number="1.3.7">
<h3><span class="header-section-number">1.3.7</span> Set difference</h3>
<p>In <strong>GSClassifier</strong>, <strong>Set difference</strong> is defined as a weight average of gene-geneset rank difference.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="the-principle-of-gsclassifier.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># No. of gene sets</span></span>
<span id="cb20-2"><a href="the-principle-of-gsclassifier.html#cb20-2" aria-hidden="true" tabindex="-1"></a>nGS <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb20-3"><a href="the-principle-of-gsclassifier.html#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="the-principle-of-gsclassifier.html#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Name of gene set comparision, which is like s1s2, s1s3 and so on.</span></span>
<span id="cb20-5"><a href="the-principle-of-gsclassifier.html#cb20-5" aria-hidden="true" tabindex="-1"></a>featureNames <span class="ot">&lt;-</span> <span class="st">&#39;s1s2&#39;</span></span>
<span id="cb20-6"><a href="the-principle-of-gsclassifier.html#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="the-principle-of-gsclassifier.html#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene set difference across samples</span></span>
<span id="cb20-8"><a href="the-principle-of-gsclassifier.html#cb20-8" aria-hidden="true" tabindex="-1"></a>resultList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb20-9"><a href="the-principle-of-gsclassifier.html#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(expr_sub)) { <span class="co"># i=1</span></span>
<span id="cb20-10"><a href="the-principle-of-gsclassifier.html#cb20-10" aria-hidden="true" tabindex="-1"></a>  res0 <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="at">length=</span><span class="fu">length</span>(featureNames))</span>
<span id="cb20-11"><a href="the-principle-of-gsclassifier.html#cb20-11" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb20-12"><a href="the-principle-of-gsclassifier.html#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j1 <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(nGS<span class="dv">-1</span>)) { <span class="co"># j1=1</span></span>
<span id="cb20-13"><a href="the-principle-of-gsclassifier.html#cb20-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j2 <span class="cf">in</span> (j1<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span>nGS) { <span class="co"># j2=2</span></span>
<span id="cb20-14"><a href="the-principle-of-gsclassifier.html#cb20-14" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb20-15"><a href="the-principle-of-gsclassifier.html#cb20-15" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If j1=1 and j2=2, gene sets s1/s2 would be selected</span></span>
<span id="cb20-16"><a href="the-principle-of-gsclassifier.html#cb20-16" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb20-17"><a href="the-principle-of-gsclassifier.html#cb20-17" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Genes of different gene sets</span></span>
<span id="cb20-18"><a href="the-principle-of-gsclassifier.html#cb20-18" aria-hidden="true" tabindex="-1"></a>      set1 <span class="ot">&lt;-</span> geneSet[[j1]] <span class="co"># &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot;</span></span>
<span id="cb20-19"><a href="the-principle-of-gsclassifier.html#cb20-19" aria-hidden="true" tabindex="-1"></a>      set2 <span class="ot">&lt;-</span> geneSet[[j2]] <span class="co"># &quot;Gene4&quot; &quot;Gene5&quot; &quot;Gene6&quot;</span></span>
<span id="cb20-20"><a href="the-principle-of-gsclassifier.html#cb20-20" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb20-21"><a href="the-principle-of-gsclassifier.html#cb20-21" aria-hidden="true" tabindex="-1"></a>      <span class="co"># RNA expression of Genes by different gene sets</span></span>
<span id="cb20-22"><a href="the-principle-of-gsclassifier.html#cb20-22" aria-hidden="true" tabindex="-1"></a>      vals1 <span class="ot">&lt;-</span> expr_sub[<span class="fu">rownames</span>(expr_sub) <span class="sc">%in%</span> set1,i]</span>
<span id="cb20-23"><a href="the-principle-of-gsclassifier.html#cb20-23" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Gene1 Gene2 Gene3</span></span>
<span id="cb20-24"><a href="the-principle-of-gsclassifier.html#cb20-24" aria-hidden="true" tabindex="-1"></a>      <span class="co"># 0.51  0.52  0.53</span></span>
<span id="cb20-25"><a href="the-principle-of-gsclassifier.html#cb20-25" aria-hidden="true" tabindex="-1"></a>      vals2 <span class="ot">&lt;-</span> expr_sub[<span class="fu">rownames</span>(expr_sub) <span class="sc">%in%</span> set2,i]</span>
<span id="cb20-26"><a href="the-principle-of-gsclassifier.html#cb20-26" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Gene4 Gene5 Gene6</span></span>
<span id="cb20-27"><a href="the-principle-of-gsclassifier.html#cb20-27" aria-hidden="true" tabindex="-1"></a>      <span class="co"># 0.21  0.22  0.23</span></span>
<span id="cb20-28"><a href="the-principle-of-gsclassifier.html#cb20-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-29"><a href="the-principle-of-gsclassifier.html#cb20-29" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Differences between one gene and gene sets</span></span>
<span id="cb20-30"><a href="the-principle-of-gsclassifier.html#cb20-30" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Compare expression of each gene in Set1 with all genes in Set2. </span></span>
<span id="cb20-31"><a href="the-principle-of-gsclassifier.html#cb20-31" aria-hidden="true" tabindex="-1"></a>      <span class="co"># For example, 0.51&gt;0.21/0.22/0.23, so the value of Gene1:s2 is 3.</span></span>
<span id="cb20-32"><a href="the-principle-of-gsclassifier.html#cb20-32" aria-hidden="true" tabindex="-1"></a>      res1 <span class="ot">&lt;-</span> <span class="fu">sapply</span>(vals1, <span class="cf">function</span>(v1) <span class="fu">sum</span>(v1 <span class="sc">&gt;</span> vals2, <span class="at">na.rm=</span>T))</span>
<span id="cb20-33"><a href="the-principle-of-gsclassifier.html#cb20-33" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Gene1:s2   Gene2:s2   Gene3:s2</span></span>
<span id="cb20-34"><a href="the-principle-of-gsclassifier.html#cb20-34" aria-hidden="true" tabindex="-1"></a>      <span class="co"># 3          3           3</span></span>
<span id="cb20-35"><a href="the-principle-of-gsclassifier.html#cb20-35" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb20-36"><a href="the-principle-of-gsclassifier.html#cb20-36" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Weight average of gene-geneset rank difference</span></span>
<span id="cb20-37"><a href="the-principle-of-gsclassifier.html#cb20-37" aria-hidden="true" tabindex="-1"></a>      res0[idx] <span class="ot">&lt;-</span> <span class="fu">sum</span>(res1, <span class="at">na.rm =</span> T) <span class="sc">/</span> (<span class="fu">length</span>(vals1) <span class="sc">*</span> <span class="fu">length</span>(vals2))</span>
<span id="cb20-38"><a href="the-principle-of-gsclassifier.html#cb20-38" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb20-39"><a href="the-principle-of-gsclassifier.html#cb20-39" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Next gene set pair</span></span>
<span id="cb20-40"><a href="the-principle-of-gsclassifier.html#cb20-40" aria-hidden="true" tabindex="-1"></a>      idx <span class="ot">&lt;-</span> idx <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb20-41"><a href="the-principle-of-gsclassifier.html#cb20-41" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb20-42"><a href="the-principle-of-gsclassifier.html#cb20-42" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb20-43"><a href="the-principle-of-gsclassifier.html#cb20-43" aria-hidden="true" tabindex="-1"></a>  resultList[[i]] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(res0)</span>
<span id="cb20-44"><a href="the-principle-of-gsclassifier.html#cb20-44" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-45"><a href="the-principle-of-gsclassifier.html#cb20-45" aria-hidden="true" tabindex="-1"></a>resMat <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, resultList)</span>
<span id="cb20-46"><a href="the-principle-of-gsclassifier.html#cb20-46" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(resMat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(expr_sub)</span>
<span id="cb20-47"><a href="the-principle-of-gsclassifier.html#cb20-47" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(resMat) <span class="ot">&lt;-</span> featureNames</span>
<span id="cb20-48"><a href="the-principle-of-gsclassifier.html#cb20-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-49"><a href="the-principle-of-gsclassifier.html#cb20-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Report </span></span>
<span id="cb20-50"><a href="the-principle-of-gsclassifier.html#cb20-50" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Set difference across samples: &#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb20-51"><a href="the-principle-of-gsclassifier.html#cb20-51" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(resMat)</span>
<span id="cb20-52"><a href="the-principle-of-gsclassifier.html#cb20-52" aria-hidden="true" tabindex="-1"></a><span class="co"># Set difference across samples:  </span></span>
<span id="cb20-53"><a href="the-principle-of-gsclassifier.html#cb20-53" aria-hidden="true" tabindex="-1"></a><span class="co">#      Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb20-54"><a href="the-principle-of-gsclassifier.html#cb20-54" aria-hidden="true" tabindex="-1"></a><span class="co"># s1s2       1       1       1       0       0       0</span></span></code></pre></div>
<p>In <strong>GSClassifier</strong>, we established <strong>makeSetData</strong> to evaluate set difference across samples:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="the-principle-of-gsclassifier.html#cb21-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-2"><a href="the-principle-of-gsclassifier.html#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene set difference across samples</span></span>
<span id="cb21-3"><a href="the-principle-of-gsclassifier.html#cb21-3" aria-hidden="true" tabindex="-1"></a>geneset_interaction <span class="ot">&lt;-</span> GSClassifier<span class="sc">:::</span><span class="fu">makeSetData</span>(expr_sub, geneSet)</span>
<span id="cb21-4"><a href="the-principle-of-gsclassifier.html#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="the-principle-of-gsclassifier.html#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Report </span></span>
<span id="cb21-6"><a href="the-principle-of-gsclassifier.html#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;Set difference across samples: &#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb21-7"><a href="the-principle-of-gsclassifier.html#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(resMat)</span>
<span id="cb21-8"><a href="the-principle-of-gsclassifier.html#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Set difference across samples:  </span></span>
<span id="cb21-9"><a href="the-principle-of-gsclassifier.html#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co">#      Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb21-10"><a href="the-principle-of-gsclassifier.html#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co"># s1s2       1       1       1       0       0       0</span></span></code></pre></div>
<p>We have known that the subtype of <strong>Sample 1-3</strong> differs from that of <strong>Sample 4-6</strong>, which revealed the robustness of set difference for subtype indentification.</p>
<p>Based on the structure of TSP in Figure <a href="the-principle-of-gsclassifier.html#fig:tsp">1.2</a>, TSP matrix of the simulated dataset should be :</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="the-principle-of-gsclassifier.html#cb22-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-2"><a href="the-principle-of-gsclassifier.html#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># TSP matrix</span></span>
<span id="cb22-3"><a href="the-principle-of-gsclassifier.html#cb22-3" aria-hidden="true" tabindex="-1"></a>tsp <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb22-4"><a href="the-principle-of-gsclassifier.html#cb22-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-5"><a href="the-principle-of-gsclassifier.html#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Binned expression</span></span>
<span id="cb22-6"><a href="the-principle-of-gsclassifier.html#cb22-6" aria-hidden="true" tabindex="-1"></a>  expr_binned[gene_bigRank,],</span>
<span id="cb22-7"><a href="the-principle-of-gsclassifier.html#cb22-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-8"><a href="the-principle-of-gsclassifier.html#cb22-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Pair difference</span></span>
<span id="cb22-9"><a href="the-principle-of-gsclassifier.html#cb22-9" aria-hidden="true" tabindex="-1"></a>  gene_bigRank_pairs,</span>
<span id="cb22-10"><a href="the-principle-of-gsclassifier.html#cb22-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-11"><a href="the-principle-of-gsclassifier.html#cb22-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Set difference</span></span>
<span id="cb22-12"><a href="the-principle-of-gsclassifier.html#cb22-12" aria-hidden="true" tabindex="-1"></a>  resMat</span>
<span id="cb22-13"><a href="the-principle-of-gsclassifier.html#cb22-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-14"><a href="the-principle-of-gsclassifier.html#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="the-principle-of-gsclassifier.html#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Report </span></span>
<span id="cb22-16"><a href="the-principle-of-gsclassifier.html#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&#39;TSP matrix: &#39;</span>, <span class="st">&#39;</span><span class="sc">\n</span><span class="st">&#39;</span>)</span>
<span id="cb22-17"><a href="the-principle-of-gsclassifier.html#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(tsp)</span>
<span id="cb22-18"><a href="the-principle-of-gsclassifier.html#cb22-18" aria-hidden="true" tabindex="-1"></a><span class="co"># TSP matrix:  </span></span>
<span id="cb22-19"><a href="the-principle-of-gsclassifier.html#cb22-19" aria-hidden="true" tabindex="-1"></a><span class="co">#             Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span id="cb22-20"><a href="the-principle-of-gsclassifier.html#cb22-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1             3       3       4       1       2       2</span></span>
<span id="cb22-21"><a href="the-principle-of-gsclassifier.html#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2             4       4       3       2       2       2</span></span>
<span id="cb22-22"><a href="the-principle-of-gsclassifier.html#cb22-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3             4       4       4       2       1       1</span></span>
<span id="cb22-23"><a href="the-principle-of-gsclassifier.html#cb22-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4             1       2       2       3       4       4</span></span>
<span id="cb22-24"><a href="the-principle-of-gsclassifier.html#cb22-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5             2       2       2       4       4       3</span></span>
<span id="cb22-25"><a href="the-principle-of-gsclassifier.html#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene6             2       1       1       4       3       4</span></span>
<span id="cb22-26"><a href="the-principle-of-gsclassifier.html#cb22-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene2       0       0       1       0       0       1</span></span>
<span id="cb22-27"><a href="the-principle-of-gsclassifier.html#cb22-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene3       0       0       0       0       1       1</span></span>
<span id="cb22-28"><a href="the-principle-of-gsclassifier.html#cb22-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene4       1       1       1       0       0       0</span></span>
<span id="cb22-29"><a href="the-principle-of-gsclassifier.html#cb22-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene5       1       1       1       0       0       0</span></span>
<span id="cb22-30"><a href="the-principle-of-gsclassifier.html#cb22-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene1:Gene6       1       1       1       0       0       0</span></span>
<span id="cb22-31"><a href="the-principle-of-gsclassifier.html#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene3       0       0       0       0       1       1</span></span>
<span id="cb22-32"><a href="the-principle-of-gsclassifier.html#cb22-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene4       1       1       1       0       0       0</span></span>
<span id="cb22-33"><a href="the-principle-of-gsclassifier.html#cb22-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene5       1       1       1       0       0       0</span></span>
<span id="cb22-34"><a href="the-principle-of-gsclassifier.html#cb22-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene2:Gene6       1       1       1       0       0       0</span></span>
<span id="cb22-35"><a href="the-principle-of-gsclassifier.html#cb22-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene4       1       1       1       0       0       0</span></span>
<span id="cb22-36"><a href="the-principle-of-gsclassifier.html#cb22-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene5       1       1       1       0       0       0</span></span>
<span id="cb22-37"><a href="the-principle-of-gsclassifier.html#cb22-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene3:Gene6       1       1       1       0       0       0</span></span>
<span id="cb22-38"><a href="the-principle-of-gsclassifier.html#cb22-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene5       0       0       1       0       0       1</span></span>
<span id="cb22-39"><a href="the-principle-of-gsclassifier.html#cb22-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene4:Gene6       0       1       1       0       1       0</span></span>
<span id="cb22-40"><a href="the-principle-of-gsclassifier.html#cb22-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene5:Gene6       0       1       1       0       1       0</span></span>
<span id="cb22-41"><a href="the-principle-of-gsclassifier.html#cb22-41" aria-hidden="true" tabindex="-1"></a><span class="co"># s1s2              1       1       1       0       0       0</span></span></code></pre></div>
<p>Have a look at the distribution:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="the-principle-of-gsclassifier.html#cb23-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-2"><a href="the-principle-of-gsclassifier.html#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Data</span></span>
<span id="cb23-3"><a href="the-principle-of-gsclassifier.html#cb23-3" aria-hidden="true" tabindex="-1"></a>tsp_df <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(tsp)</span>
<span id="cb23-4"><a href="the-principle-of-gsclassifier.html#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="the-principle-of-gsclassifier.html#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(tsp_df,<span class="fu">aes</span>(<span class="at">x=</span>Var2,<span class="at">y=</span>value,<span class="at">fill=</span>Var2)) <span class="sc">+</span> </span>
<span id="cb23-6"><a href="the-principle-of-gsclassifier.html#cb23-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_boxplot</span>(<span class="at">outlier.size =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb23-7"><a href="the-principle-of-gsclassifier.html#cb23-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Samples&#39;</span>,</span>
<span id="cb23-8"><a href="the-principle-of-gsclassifier.html#cb23-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">y =</span> <span class="st">&#39;Epression&#39;</span>,</span>
<span id="cb23-9"><a href="the-principle-of-gsclassifier.html#cb23-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="cn">NULL</span>) </span></code></pre></div>
<p><img src="Flowchart_files/figure-html/unnamed-chunk-22-1.png" width="60%" style="display: block; margin: auto;" /></p>
</div>
<div id="batch-effect" class="section level3" number="1.3.8">
<h3><span class="header-section-number">1.3.8</span> Batch effect</h3>
<!--

gene-pair batch effect in Google Scholar

relative expression orderings (REO)

single-sample enrichment analysis 

Batch effect control & gene number

-->
<p><strong>TSP</strong> was widely applied to control batch effects in transciptomic data <span class="citation">[<a href="references.html#ref-RN369" role="doc-biblioref">13</a>–<a href="references.html#ref-RN365" role="doc-biblioref">20</a>]</span>. Still, we tested whether <strong>TSP</strong> is a robust method for batch effect control in real-world data. As demonstrated in Figure <a href="the-principle-of-gsclassifier.html#fig:be01">1.6</a>, the obvious batch effects across gastric cancer datsets were significantly reduced after <strong>TSP</strong> normalization.</p>
<div class="figure" style="text-align: center"><span id="fig:be01"></span>
<img src="fig/bactch-effect-01.png" alt="Batch effects across gastric cancer cohorts. All gene pairs were used because subtype vectors were not specified. Top: Raw expression of all genes across samples. Middle: Raw expression of PIAM and PIDG across samples. Bottom: TSP of PIAM and PIDG across samples." width="90%" />
<p class="caption">
Figure 1.6: Batch effects across gastric cancer cohorts. All gene pairs were used because subtype vectors were not specified. Top: Raw expression of all genes across samples. Middle: Raw expression of PIAM and PIDG across samples. Bottom: TSP of PIAM and PIDG across samples.
</p>
</div>
<p>In order to confirmed the association between <strong>gene counts</strong> in modeling and batch effect control via <strong>TSP</strong> normalization, we selected random genes with counts ranging 4, 8, 20, 40, and 80 for TSP matrix establishment. As shown in Figure <a href="the-principle-of-gsclassifier.html#fig:be02">1.7</a>, <strong>TSP</strong> normalization works greatly in different gene counts for batch effect control compared with raw expression matrix.</p>
<div class="figure" style="text-align: center"><span id="fig:be02"></span>
<img src="fig/bactch-effect-02.png" alt="Batch effects of random genes across gastric cancer cohorts.  All gene pairs were used because subtype vectors were not specified. Gene counts 4, 8, 20, 40, and 80 were detected. Data of set difference were not available because only one gene set were applied." width="90%" />
<p class="caption">
Figure 1.7: Batch effects of random genes across gastric cancer cohorts. All gene pairs were used because subtype vectors were not specified. Gene counts 4, 8, 20, 40, and 80 were detected. Data of set difference were not available because only one gene set were applied.
</p>
</div>
<!--
Batch effect reduction of TSP could be explained in three parts:

+ **Binned expression**. Regardless of what raw distribution of a numeric vector is, it would be converted into a binned vector with exactly the same distribution under a fixed quantile vector.

+ **Pair/set difference**.
-->

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="discussion.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["GSClassifier.principle.pdf", "GSClassifier.principle.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
